diff -urN cyrus-imapd-2.4.17/imap/index.c cyrus-imapd-2.4.17.vanishfix/imap/index.c
--- cyrus-imapd-2.4.17/imap/index.c	2012-12-01 20:57:54.000000000 +0100
+++ cyrus-imapd-2.4.17.vanishfix/imap/index.c	2014-06-27 16:50:49.000000000 +0200
@@ -648,6 +648,7 @@
 	    syslog(LOG_WARNING,
 		   "Mailbox %s has been (re)moved out from under client",
 		   mailbox->name);
+	    mailbox_close(&mailbox);
 	    fatal("Mailbox has been (re)moved", EC_IOERR);
 	}
 
