Index: client.c
===================================================================
RCS file: /cvs/src/sasl/lib/client.c,v
retrieving revision 1.73
diff -u -r1.73 client.c
--- client.c	20 Feb 2009 23:10:53 -0000	1.73
+++ client.c	21 Feb 2009 17:29:46 -0000
@@ -65,9 +65,6 @@
 
 static int init_mechlist()
 {
-  cmechlist->mutex = sasl_MUTEX_ALLOC();
-  if(!cmechlist->mutex) return SASL_FAIL;
-  
   cmechlist->utils=_sasl_alloc_utils(NULL, &global_callbacks_client);
   if (cmechlist->utils==NULL)
     return SASL_NOMEM;
@@ -134,7 +131,6 @@
     sasl_FREE(cprevm->m.plugname);
     sasl_FREE(cprevm);    
   }
-  sasl_MUTEX_FREE(cmechlist->mutex);
   _sasl_free_utils(&cmechlist->utils);
   sasl_FREE(cmechlist);
 
Index: server.c
===================================================================
RCS file: /cvs/src/sasl/lib/server.c,v
retrieving revision 1.159
diff -u -r1.159 server.c
--- server.c	20 Feb 2009 23:10:53 -0000	1.159
+++ server.c	21 Feb 2009 17:29:56 -0000
@@ -355,9 +355,6 @@
 {
     sasl_utils_t *newutils = NULL;
 
-    mechlist->mutex = sasl_MUTEX_ALLOC();
-    if(!mechlist->mutex) return SASL_FAIL;
-
     /* set util functions - need to do rest */
     newutils = _sasl_alloc_utils(NULL, &global_callbacks);
     if (newutils == NULL)
@@ -537,7 +534,6 @@
 	  sasl_FREE(prevm);    
       }
       _sasl_free_utils(&mechlist->utils);
-      sasl_MUTEX_FREE(mechlist->mutex);
       sasl_FREE(mechlist);
       mechlist = NULL;
   }
