*** cyrus-sasl-2.1.15/plugins/gssapi.c	2003-08-06 14:58:12.000000000 +0100
--- /tmp/gssapi.c.mine	2003-08-06 14:59:52.000000000 +0100
***************
*** 1129,1135 ****
  
  	if (oparams->mech_ssf) {
  	    /* xxx this is probably too big */
! 	    oparams->maxoutbuf -= 50;
  	}
  	
  	GSS_LOCK_MUTEX(params->utils);
--- 1129,1135 ----
  
  	if (oparams->mech_ssf) {
  	    /* xxx this is probably too big */
! 	    oparams->maxoutbuf -= 128;
  	}
  	
  	GSS_LOCK_MUTEX(params->utils);
***************
*** 1583,1589 ****
  
  	if(oparams->mech_ssf) {
  	    /* xxx probably too large */
! 	    oparams->maxoutbuf -= 50;
  	}
  	
  	GSS_LOCK_MUTEX(params->utils);
--- 1583,1589 ----
  
  	if(oparams->mech_ssf) {
  	    /* xxx probably too large */
! 	    oparams->maxoutbuf -= 128;
  	}
  	
  	GSS_LOCK_MUTEX(params->utils);
