From 32d6472b7dcd553e76cee06ea99317b27c7ecedd Mon Sep 17 00:00:00 2001
From: "Jeroen van Meeuwen (Kolab Systems)" <vanmeeuwen@kolabsys.com>
Date: Sun, 15 Apr 2012 14:56:11 +0100
Subject: [PATCH 6/6] Include sieve/sieve.c in imap_lmtpd_SOURCES

---
 Makefile.am |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index 745a3c6..f32f430 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -303,7 +303,7 @@ imap_lmtpd_SOURCES = imap/lmtpd.c imap/lmtpd.h imap/lmtpengine.c \
 	imap/mutex_fake.c imap/proxy.c imap/spool.c master/service.c
 imap_lmtpd_LDADD = $(imap_fud_LDADD)
 if SIEVE
-imap_lmtpd_SOURCES += imap/lmtp_sieve.c imap/smtpclient.c
+imap_lmtpd_SOURCES += imap/lmtp_sieve.c imap/smtpclient.c sieve/sieve.c
 imap_lmtpd_LDADD += sieve/libsieve.a
 endif
 
-- 
1.7.7.6

