--- imapoptions.orig	2010-08-20 13:46:20.000000000 +1200
+++ imapoptions	2010-08-20 13:50:05.000000000 +1200
@@ -824,14 +824,14 @@
    't', and 's' for the owner of the mailbox.  The 'r' right allows the
    user to open the mailbox and list/retrieve messages.  The 't' right
    allows the user to delete messages.  The 's' right allows messages
-   retrieved by the user to have the \Seen flag set (only if
+   retrieved by the user to have the \fIseen\fR flag set (only if
    \fIpopuseimapflags\fR is also enabled). */
 
 { "popuseimapflags", 0, SWITCH }
 /* If enabled, the pop server will set and obey IMAP flags.  Messages
-   having the \Deleted flag are ignored as if they do not exist.
-   Messages that are retrieved by the client will have the \Seen flag
-   set.  All messages will have the \Recent flag unset. */
+   having the \fIdeleted\fR flag are ignored as if they do not exist.
+   Messages that are retrieved by the client will have the \fIseen\fR flag
+   set.  All messages will have the \fIrecent\fR flag unset. */
 
 { "postmaster", "postmaster", STRING }
 /* Username that is used as the 'From' address in rejection MDNs produced
