--- ChangeLog~	2004-03-10 10:51:35.000000000 -0500
+++ ChangeLog	2004-06-01 21:40:18.402001000 -0400
@@ -1828,7 +1828,7 @@
 	* plugins/plugin_common.c (_plug_ipfromstring): add call to
 	  freeaddrinfo()
 
-	* lib/saslutil.c (sasl_randseed): actually initilize the randpool
+	* lib/saslutil.c (sasl_randseed): actually initialize the randpool
 
 	* saslauthd/auth_getpwent.c (auth_getpwent): clear a warning
 	* saslauthd/auth_shadow.c (auth_shadow): clear a similar warning
--- doc/sysadmin.html~	2004-01-07 11:35:07.000000000 -0500
+++ doc/sysadmin.html	2004-06-01 21:39:55.132003000 -0400
@@ -395,7 +395,7 @@
 
 <p>
 <dt><b>Q:</b> I'm having performance problems on each authentication, there is
-a noticeable slowdown when sasl initilizes, what can I do?
+a noticeable slowdown when sasl initializes, what can I do?
 <dd>
 <p><b>A:</b>libsasl reads from <tt>/dev/random</tt> as part of its
 initialization. <tt>/dev/random</tt> is a "secure" source of entropy,
--- saslauthd/saslauthd-main.c~	2003-07-23 13:05:33.000000000 -0400
+++ saslauthd/saslauthd-main.c	2004-06-01 21:38:55.001000000 -0400
@@ -423,7 +423,7 @@
 
 	if (auth_mech->initialize) {
 		if(auth_mech->initialize() != 0) {
-		    logger(L_ERR, L_FUNC, "failed to initilize mechanism %s",
+		    logger(L_ERR, L_FUNC, "failed to initialize mechanism %s",
 			   auth_mech->name);
 		    exit(1);
 		}
