--- cyrus-imapd-2.3.8/imap/ctl_cyrusdb.c	2006-11-30 12:11:17.000000000 -0500
+++ cyrus-imapd-2.3.8p1/imap/ctl_cyrusdb.c	2007-03-15 17:46:11.000000000 -0400
@@ -133,7 +133,7 @@
     /* if it is MBTYPE_RESERVED, unset it & call mboxlist_delete */
     if(!r && (mbtype & MBTYPE_RESERVE)) {
 	if(!r) {
-	    r = mboxlist_deletemailbox(name, 1, NULL, NULL, 0, 0, 1);
+	    r = mboxlist_deletemailbox(name, 1, NULL, NULL, 0, 0, 1, 1);
 	    if(r) {
 		/* log the error */
 		syslog(LOG_ERR,
--- cyrus-imapd-2.3.8/imap/ctl_mboxlist.c	2007-02-05 13:41:46.000000000 -0500
+++ cyrus-imapd-2.3.8p1/imap/ctl_mboxlist.c	2007-03-15 17:49:18.000000000 -0400
@@ -461,7 +461,7 @@
 	    
 	    wipe_head = wipe_head->next;
 	    
-	    ret = mboxlist_deletemailbox(me->mailbox, 1, "", NULL, 0, 1, 1);
+	    ret = mboxlist_deletemailbox(me->mailbox, 1, "", NULL, 0, 1, 1, 1);
 	    if(!ret) sync_log_mailbox(me->mailbox);
 	    if(ret) {
 		fprintf(stderr, "couldn't delete defunct mailbox %s\n",
--- cyrus-imapd-2.3.8/imap/imapd.c	2007-02-05 13:49:55.000000000 -0500
+++ cyrus-imapd-2.3.8p1/imap/imapd.c	2007-03-15 17:46:11.000000000 -0400
@@ -4983,7 +4985,7 @@
 
     r = mboxlist_deletemailbox(name, imapd_userisadmin,
 			       imapd_userid, imapd_authstate,
-			       0, 0, 0);
+			       0, 0, 0, 0);
     
     if (!r) sync_log_mailbox(name);
 
@@ -5065,7 +5067,7 @@
 
 	r = mboxlist_deletemailbox(mailboxname, imapd_userisadmin,
 				   imapd_userid, imapd_authstate, 1-force,
-				   localonly, 0);
+				   localonly, 0, localonly);
     }
 
     /* was it a top-level user mailbox? */
@@ -8221,7 +8223,7 @@
 	/* note also that we need to remember to let proxyadmins do this */
 	r = mboxlist_deletemailbox(mailboxname,
 				   imapd_userisadmin || imapd_userisproxyadmin,
-				   imapd_userid, imapd_authstate, 0, 1, 0);
+				   imapd_userid, imapd_authstate, 0, 1, 0, 1);
 	if(r) syslog(LOG_ERR,
 		     "Could not delete local mailbox during move of %s",
 		     mailboxname);
--- cyrus-imapd-2.3.8/imap/mailbox.c	2007-02-05 13:41:47.000000000 -0500
+++ cyrus-imapd-2.3.8p1/imap/mailbox.c	2007-03-16 16:34:28.000000000 -0400
@@ -2674,12 +2674,13 @@
 /*
  * Remove all files in directory
  */
-static void mailbox_delete_files(char *path)
+static void mailbox_delete_files(char *path, char *deleteddir)
 {
     DIR *dirp;
     struct dirent *f;
     char buf[MAX_MAILBOX_PATH+1];
     char *tail;
+    char dbuf[MAX_MAILBOX_PATH+1], *dtail;
 
     strlcpy(buf, path, sizeof(buf));
 
@@ -2691,6 +2692,19 @@
     tail = buf + strlen(buf);
     *tail++ = '/';
     *tail = '\0';
+
+    if ( deleteddir ) {
+	strlcpy( dbuf, deleteddir, sizeof( dbuf ));
+	if ( strlen( dbuf ) >= sizeof( dbuf ) - 2 ) {
+	    syslog( LOG_ERR, "IOERROR: Path too long (%s)", dbuf );
+	    fatal( "path too long", EC_OSFILE );
+	}
+
+	dtail = dbuf + strlen( dbuf );
+	*dtail++ = '/';
+	*dtail = '\0';
+    }
+
     dirp = opendir(path);
     if (dirp) {
 	while ((f = readdir(dirp))!=NULL) {
@@ -2710,7 +2724,34 @@
 		fatal("Path too long", EC_OSFILE);
 	    }
 	    strcpy(tail, f->d_name);
-	    unlink(buf);
+	    if ( deleteddir ) {
+		struct stat	st;
+
+		if ( stat( buf, &st ) != 0 ) {
+		    syslog( LOG_ERR, "IOERROR: Can't stat (%s) %m", buf );
+		    fatal( "Stat error", EC_OSFILE );
+		}
+
+		if ( !S_ISDIR( st.st_mode )) {
+		    if ( strlen( dbuf )+strlen( f->d_name ) >= sizeof( dbuf )) {
+			syslog( LOG_ERR, "IOERROR: Path too long (%s + %s)",
+				dbuf, f->d_name );
+			fatal( "Path too long", EC_OSFILE );
+		    }
+		    strcpy( dtail, f->d_name );
+		    if ( rename( buf, dbuf ) != 0 ) {
+			syslog( LOG_ERR, "IOERROR: Can't rename (%s %s) %m",
+				buf, dbuf );
+			fatal( "Rename error", EC_OSFILE );
+		    }
+		    *dtail = '\0';
+		}
+	    } else {
+		/*
+		 * This unlink() fails silently on directories.
+		 */
+		(void)unlink(buf);
+	    }
 	    *tail = '\0';
 	}
 	closedir(dirp);
@@ -2730,7 +2771,7 @@
  * Delete and close the mailbox 'mailbox'.  Closes 'mailbox' whether
  * or not the deletion was successful.  Requires a locked mailbox.
  */
-int mailbox_delete(struct mailbox *mailbox, int delete_quota_root)
+int mailbox_delete(struct mailbox *mailbox, int delete_quota_root, char *deleteddir)
 {
     int r, rquota = 0;
     char nbuf[MAX_MAILBOX_NAME+1];
@@ -2767,14 +2808,21 @@
 	    quota_commit(&tid);
     }
 
+    if ( deleteddir ) {
+	if ( cyrus_mkdir( deleteddir, 0755 ) != 0 ) return IMAP_IOERROR;
+	if ( mkdir( deleteddir, 0755 ) == -1 ) return IMAP_IOERROR;
+	syslog( LOG_NOTICE, "Saving deleted mailbox %s in %s",
+		mailbox->name, deleteddir );
+    }
+
     /* Flush data (message file) directory */
-    mailbox_delete_files(mailbox->path);
+    mailbox_delete_files(mailbox->path, deleteddir);
     strlcpy(pbuf, mailbox->path, sizeof(pbuf));
     ptail = pbuf + strlen(pbuf);
 
     /* Flush metadata directory */
     if (mailbox->mpath) {
-	mailbox_delete_files(mailbox->mpath);
+	mailbox_delete_files(mailbox->mpath, deleteddir);
 	strlcpy(mbuf, mailbox->mpath, sizeof(mbuf));
 	mtail = mbuf + strlen(mbuf);
     }
@@ -3009,7 +3057,7 @@
 	/* Expunge old mailbox */
 	r = mailbox_expunge(oldmailbox, expungeall, (char *)0, EXPUNGE_FORCE);
     } else {
-	r = mailbox_delete(oldmailbox, 0);
+	r = mailbox_delete(oldmailbox, 0, NULL);
     }
 
     if(r) {
--- cyrus-imapd-2.3.8/imap/mailbox.h	2006-11-30 12:11:19.000000000 -0500
+++ cyrus-imapd-2.3.8p1/imap/mailbox.h	2007-03-15 17:46:11.000000000 -0400
@@ -362,7 +362,7 @@
 extern int mailbox_create(const char *name, char *partition,
 			  const char *acl, const char *uniqueid, int format,
 			  struct mailbox *mailboxp);
-extern int mailbox_delete(struct mailbox *mailbox, int delete_quota_root);
+extern int mailbox_delete(struct mailbox *mailbox, int delete_quota_root, char *deleteddir);
 
 extern int mailbox_rename_copy(struct mailbox *oldmailbox, 
 			       const char *newname, char *newpartition,
--- cyrus-imapd-2.3.8/imap/mboxlist.c	2007-02-05 13:41:47.000000000 -0500
+++ cyrus-imapd-2.3.8p1/imap/mboxlist.c	2007-03-15 17:46:11.000000000 -0400
@@ -889,7 +889,7 @@
  */
 int mboxlist_deletemailbox(const char *name, int isadmin, char *userid, 
 			   struct auth_state *auth_state, int checkacl,
-			   int local_only, int force)
+			   int local_only, int force, int now)
 {
     int r;
     char *acl;
@@ -902,6 +902,11 @@
     int mbtype;
     const char *p;
     mupdate_handle *mupdate_h = NULL;
+    char *partition, *deleteddir = NULL;
+    const char *root;
+    char deleteddot[ MAX_MAILBOX_NAME+1 ], deletedhash[ MAX_MAILBOX_NAME+1 ];
+    char deletedbuf[ MAX_MAILBOX_NAME+1 ];
+    struct timeval tv;
 
     if(!isadmin && force) return IMAP_PERMISSION_DENIED;
 
@@ -922,7 +927,7 @@
 	if (!isadmin) { r = IMAP_PERMISSION_DENIED; goto done; }
     }
 
-    r = mboxlist_mylookup(name, &mbtype, &path, &mpath, NULL, &acl, &tid, 1);
+    r = mboxlist_mylookup(name, &mbtype, &path, &mpath, &partition, &acl, &tid, 1);
     switch (r) {
     case 0:
 	break;
@@ -1014,7 +1019,19 @@
 
     if ((r && !force) || isremote) goto done;
 
-    if (!r || force) r = mailbox_delete(&mailbox, deletequotaroot);
+    if ( !now & config_getenum(IMAPOPT_DELETE_MODE) !=
+	    IMAP_ENUM_DELETE_MODE_IMMEDIATE ) {
+	root = config_partitiondir( partition );
+	if ( !root ) return IMAP_PARTITION_UNKNOWN;
+
+	snprintf( deleteddot, sizeof( deleteddot ), "%s/deleted./", root );
+	mailbox_hash_mbox( deletedhash, sizeof( deletedhash ), deleteddot, name );
+	gettimeofday( &tv, NULL );
+	snprintf( deletedbuf, sizeof( deletedbuf ), "%s.%X", deletedhash, tv.tv_sec );
+	deleteddir = deletedbuf;
+    }
+
+    if (!r || force) r = mailbox_delete(&mailbox, deletequotaroot, deleteddir);
 
     /*
      * See if we have to remove mailbox's quota root
--- cyrus-imapd-2.3.8/imap/mboxlist.h	2006-11-30 12:11:19.000000000 -0500
+++ cyrus-imapd-2.3.8p1/imap/mboxlist.h	2007-03-15 17:47:55.000000000 -0400
@@ -126,7 +126,7 @@
  * the planet */
 int mboxlist_deletemailbox(const char *name, int isadmin, char *userid, 
 			   struct auth_state *auth_state, int checkacl,
-			   int local_only, int force);
+			   int local_only, int force, int now);
 
 /* Rename/move a mailbox (hierarchical) */
 int mboxlist_renamemailbox(char *oldname, char *newname, char *partition, 
--- cyrus-imapd-2.3.8/imap/mupdate.c	2007-01-31 09:10:05.000000000 -0500
+++ cyrus-imapd-2.3.8p1/imap/mupdate.c	2007-03-15 17:46:11.000000000 -0400
@@ -2313,7 +2313,7 @@
 	    remote_boxes.head = r->next;
 	} else if (ret < 0) {
 	    /* Local without corresponding remote, delete it */
-	    mboxlist_deletemailbox(l->mailbox, 1, "", NULL, 0, 0, 0);
+	    mboxlist_deletemailbox(l->mailbox, 1, "", NULL, 0, 0, 0, 1);
 	    local_boxes.head = l->next;
 	} else /* (ret > 0) */ {
 	    /* Remote without corresponding local, insert it */
@@ -2328,7 +2328,7 @@
     if(l && !r) {
 	/* we have more deletes to do */
 	while(l) {
-	    mboxlist_deletemailbox(l->mailbox, 1, "", NULL, 0, 0, 0);
+	    mboxlist_deletemailbox(l->mailbox, 1, "", NULL, 0, 0, 0, 1);
 	    local_boxes.head = l->next;
 	    l = local_boxes.head;
 	}
--- cyrus-imapd-2.3.8/imap/sync_reset.c	2006-11-30 12:11:20.000000000 -0500
+++ cyrus-imapd-2.3.8p1/imap/sync_reset.c	2007-03-15 17:46:11.000000000 -0400
@@ -258,7 +258,7 @@
     if (r) goto fail;
 
     for (item = list->head ; item ; item = item->next) {
-        r=mboxlist_deletemailbox(item->name, 1, NULL, sync_authstate, 1, 0, 0);
+        r=mboxlist_deletemailbox(item->name, 1, NULL, sync_authstate, 1, 0, 0, 1);
 
         if (r) goto fail;
     }
@@ -274,7 +274,7 @@
     if (r) goto fail;
 
     for (item = list->head ; item ; item = item->next) {
-        r=mboxlist_deletemailbox(item->name, 1, NULL, sync_authstate, 1, 0, 0);
+        r=mboxlist_deletemailbox(item->name, 1, NULL, sync_authstate, 1, 0, 0, 1);
 
         if (r) goto fail;
     }
@@ -282,7 +282,7 @@
 
     /* Nuke inbox (recursive nuke possible?) */
     snprintf(buf, sizeof(buf)-1, "user.%s", user);
-    r = mboxlist_deletemailbox(buf, 1, "cyrus", sync_authstate, 1, 0, 0);
+    r = mboxlist_deletemailbox(buf, 1, "cyrus", sync_authstate, 1, 0, 0, 1);
     if (r && (r != IMAP_MAILBOX_NONEXISTENT)) goto fail;
 
     if ((r=user_deletedata(user, sync_userid, sync_authstate, 1)))
--- cyrus-imapd-2.3.8/imap/sync_server.c	2006-11-30 12:11:20.000000000 -0500
+++ cyrus-imapd-2.3.8p1/imap/sync_server.c	2007-03-15 17:46:11.000000000 -0400
@@ -1581,7 +1581,7 @@
 
     for (item = list->head ; item ; item = item->next) {
         r=mboxlist_deletemailbox(item->name, sync_userisadmin, sync_userid,
-				 sync_authstate, 0, 0, 1);
+				 sync_authstate, 0, 0, 1, 0);
 
         if (r) goto fail;
     }
@@ -1591,7 +1591,7 @@
     (sync_namespacep->mboxname_tointernal)(sync_namespacep, "INBOX",
 					   user, buf);
     r = mboxlist_deletemailbox(buf, sync_userisadmin, sync_userid,
-			       sync_authstate, 0, 0, 1);
+			       sync_authstate, 0, 0, 1, 0);
     if (r && (r != IMAP_MAILBOX_NONEXISTENT)) goto fail;
 
     if ((r=user_deletedata(user, sync_userid, sync_authstate, 1)))
@@ -2526,7 +2526,7 @@
 
     /* Delete with admin priveleges */
     r = mboxlist_deletemailbox(name, sync_userisadmin, sync_userid,
-			       sync_authstate, 0, 0, 1);
+			       sync_authstate, 0, 0, 1, 0);
 
     if (r)
         prot_printf(sync_out, "NO Failed to delete %s: %s\r\n",
--- cyrus-imapd-2.3.8/lib/imapoptions	2007-02-07 13:58:07.000000000 -0500
+++ cyrus-imapd-2.3.8p1/lib/imapoptions	2007-03-15 17:46:11.000000000 -0400
@@ -197,6 +197,14 @@
 { "defaultpartition", "default", STRING }
 /* The partition name used by default for new mailboxes. */
 
+{ "delete_mode", "immediate", ENUM("immediate", "delayed") }
+/* The mode in which mailboxes (and their contents) are deleted.
+   "Immediate" mode is the default behavior in which the mailboxes and
+   their contents are removed at the time of the DELETE.  In "delayed" mode,
+   the contents of mailboxes are copied to a holding area, to be purged
+   at a later time by "cyr_expire".  This allows accidentally removed
+   mailboxes to be recovered. */
+
 { "deleteright", "c", STRING }
 /* Deprecated - only used for backwards compatibility with existing
    installations.  Lists the old RFC 2086 right which was used to
