Index: imap/notify.c
===================================================================
RCS file: /cvs/src/cyrus/imap/notify.c,v
retrieving revision 1.17
diff -u -r1.17 notify.c
--- imap/notify.c	6 Jan 2010 17:01:38 -0000	1.17
+++ imap/notify.c	21 Jul 2010 13:45:49 -0000
@@ -145,6 +145,7 @@
 	       (struct sockaddr *)&sun_data, sizeof(sun_data));
     if (r < buflen) {
 	syslog(LOG_ERR, "unable to sendto() notify socket: %m");
+	close(soc);
 	return;
     }
 
