--- cyrus-imapd-2.3.14/imap/mboxname.c~	2009-04-24 16:35:20.000000000 -0400
+++ cyrus-imapd-2.3.14/imap/mboxname.c	2009-04-24 16:33:44.000000000 -0400
@@ -731,6 +731,15 @@
 
     unixsep = config_getswitch(IMAPOPT_UNIXHIERARCHYSEP);
 
+    /* Skip policy check on mailbox created in delayed delete namespace
+     * assuming the mailbox existed before and was OK then.
+     * This should allow mailboxes that are extremely long to be
+     * deleted when delayed_delete is enabled. 
+     * A thorough fix might remove the prefix and timestamp
+     * then continue with the check
+     */
+    if (mboxname_isdeletedmailbox(name)) return 0;
+
     if (strlen(name) > MAX_MAILBOX_NAME) return IMAP_MAILBOX_BADNAME;
     for (i = 0; i < NUM_BADMBOXPATTERNS; i++) {
 	g = glob_init(badmboxpatterns[i], 0);
