diff --git a/imap/imapd.c b/imap/imapd.c
index 7dda772..35ba4bd 100644
--- a/imap/imapd.c
+++ b/imap/imapd.c
@@ -8746,8 +8746,9 @@ static int xfer_user_cb(char *name,
     /* Skip remote mailbox */
     if (mbentry.mbtype & MBTYPE_REMOTE) return 0;
 
-    xfer_addmbox(xfer, name);
-
+    r = xfer_addmbox(xfer, name);
+    if (r) return r;
+    
     return 0;
 }
 
@@ -8874,7 +8875,8 @@ void cmd_xfer(char *tag, char *name, char *toserver, char *topart)
     if (r) goto done;
 
     /* we're always moving this mailbox */
-    xfer_addmbox(xfer, mailboxname);
+    r = xfer_addmbox(xfer, mailboxname);
+    if (r) goto done;
 
     /* if we are not moving a user, just move the one mailbox */
     if (!moving_user) {
