*** lib/tok.c.orig	Wed May  8 11:17:23 2013
--- lib/tok.c	Wed May  8 11:20:01 2013
***************
*** 47,52 ****
--- 47,76 ----
  #include "tok.h"
  #include "xmalloc.h"
  
+ #if defined(__SunOS)
+ static char* strsep(char** str, const char* delims) {
+     char* token;
+ 
+     if (*str==NULL) {
+         /* No more tokens */
+         return NULL;
+     }
+ 
+     token=*str;
+     while (**str!='\0') {
+         if (strchr(delims,**str)!=NULL) {
+             **str='\0';
+             (*str)++;
+             return token;
+         }
+         (*str)++;
+     }
+     /* There is no other token */
+     *str=NULL;
+     return token;
+ }
+ #endif /* defined(__SunOS) */
+ 
  void tok_init(tok_t *t, const char *str, const char *sep, int flags)
  {
      tok_initm(t, str ? xstrdup(str) : NULL, sep, flags|TOK_FREEBUFFER);
