--- cyrus-imapd-2.4.6/imap/proc.c.orig	2010-12-20 14:15:49.000000000 +0100
+++ cyrus-imapd-2.4.6/imap/proc.c	2011-01-19 14:40:57.000000000 +0100
@@ -76,8 +76,16 @@
 
 	procfile = fopen(procfname, "w+");
 	if (!procfile) {
-	    syslog(LOG_ERR, "IOERROR: creating %s: %m", procfname);
-	    fatal("can't write proc file", EC_IOERR);
+	    if (cyrus_mkdir(procfname, 0755) == -1) {
+		fatal("couldn't create proc directory", EC_IOERR);
+	    } else {
+		syslog(LOG_NOTICE, "created proc directory");
+		procfile = fopen(procfname, "w+");
+		if (!procfile) {
+		    syslog(LOG_ERR, "IOERROR: creating %s: %m", procfname);
+		    fatal("can't write proc file", EC_IOERR);
+		}
+	    }
 	}
     }
 
