Index: master.c
===================================================================
RCS file: /cvs/src/cyrus/master/master.c,v
retrieving revision 1.98
diff -u -r1.98 master.c
--- master.c	22 Jun 2004 19:00:05 -0000	1.98
+++ master.c	2 Dec 2004 18:54:50 -0000
@@ -1366,7 +1366,7 @@
 
 	Services[i].maxforkrate = maxforkrate;
 
- 	Services[nservices].maxfds = maxfds;
+ 	Services[i].maxfds = maxfds;
 
 	if (!strcmp(Services[i].proto, "tcp") ||
 	    !strcmp(Services[i].proto, "tcp4") ||
