Index: index.c
===================================================================
RCS file: /afs/andrew/system/cvs/src/cyrus/imap/index.c,v
retrieving revision 1.242
diff -u -r1.242 index.c
--- index.c	18 Jan 2008 19:17:08 -0000	1.242
+++ index.c	26 Feb 2008 17:21:35 -0000
@@ -62,6 +62,7 @@
 #include "assert.h"
 #include "charset.h"
 #include "exitcodes.h"
+#include "gmtoff.h"
 #include "hash.h"
 #include "imap_err.h"
 #include "global.h"
@@ -3076,14 +3077,26 @@
     if (searchargs->smaller && SIZE(msgno) >= searchargs->smaller) return 0;
     if (searchargs->larger && SIZE(msgno) <= searchargs->larger) return 0;
 
-    if (searchargs->after && (time_t) INTERNALDATE(msgno) < searchargs->after)
-      return 0;
-    if (searchargs->before && (time_t) INTERNALDATE(msgno) > searchargs->before)
-      return 0;
-    if (searchargs->sentafter && (time_t) SENTDATE(msgno) < searchargs->sentafter)
-      return 0;
-    if (searchargs->sentbefore && (time_t) SENTDATE(msgno) > searchargs->sentbefore)
-      return 0;
+    if (searchargs->after || searchargs->before) {
+	time_t date = INTERNALDATE(msgno);
+	struct tm *tm = localtime(&date);
+
+	/* normalize to GMT */
+	date -= gmtoff_of(tm, date);
+
+	if (searchargs->after && date < searchargs->after) return 0;
+	if (searchargs->before && date > searchargs->before) return 0;
+    }
+    if (searchargs->sentafter || searchargs->sentbefore) {
+	time_t date = SENTDATE(msgno);
+	struct tm *tm = localtime(&date);
+
+	/* normalize to GMT */
+	date -= gmtoff_of(tm, date);
+
+	if (searchargs->sentafter && date < searchargs->sentafter) return 0;
+	if (searchargs->sentbefore && date > searchargs->sentbefore) return 0;
+    }
 
     if (searchargs->modseq && MODSEQ(msgno) < searchargs->modseq) return 0;
 
