--- cyrus-imapd-2.2.12/lib/cyrusdb_berkeley.c.orig	Wed Nov 17 17:41:31 2004
+++ cyrus-imapd-2.2.12/lib/cyrusdb_berkeley.c	Wed Aug 10 12:15:07 2005
@@ -149,7 +149,12 @@
 #endif
     }
 
+#if ((DB_VERSION_MAJOR == 4 && DB_VERSION_MINOR <= 2) || (DB_VERSION_MAJOR < 4))
     dbenv->set_errcall(dbenv, db_err);
+#else
+    dbenv->set_errcall(dbenv, db_err);
+    dbenv->set_msgcall(dbenv, db_err);
+#endif
     snprintf(errpfx, sizeof(errpfx), "db%d", DB_VERSION_MAJOR);
     dbenv->set_errpfx(dbenv, errpfx);
 
