*** annotate.c- Thu Aug  4 15:06:51 2005
--- annotate.c  Thu Aug  4 12:46:21 2005
***************
*** 983,989 ****
                                          (void*) entries_ptr->entrypat));
      }

!     if (proxy_fetch_func && fdata->orig_entry
        && !hash_lookup(mbrock.server, &(fdata->server_table))) {
        /* xxx ignoring result */
        proxy_fetch_func(mbrock.server, fdata->orig_mailbox,
--- 983,989 ----
                                          (void*) entries_ptr->entrypat));
      }

!     if (mbrock.server && proxy_fetch_func && fdata->orig_entry
        && !hash_lookup(mbrock.server, &(fdata->server_table))) {
        /* xxx ignoring result */
        proxy_fetch_func(mbrock.server, fdata->orig_mailbox,

