Index: script.c
===================================================================
RCS file: /cvs/src/sieve/script.c,v
retrieving revision 1.63
diff -u -r1.63 script.c
--- script.c	30 Nov 2006 17:11:24 -0000	1.63
+++ script.c	9 Aug 2007 12:54:45 -0000
@@ -536,6 +536,11 @@
 	    syslog(LOG_ERR, "IOERROR: can not open sieve script %s: %m", fname);
 	    return SIEVE_FAIL;
 	}
+	if (fstat(fd, &sbuf) == -1) {
+	    syslog(LOG_ERR, "IOERROR: fstating sieve script %s: %m", fname);
+	    close(fd);
+	    return SIEVE_FAIL;
+	}
 
 	bc = (sieve_bytecode_t *) xzmalloc(sizeof(sieve_bytecode_t));
 
