--- cyrus-imapd-2.3-2.3.7-CLEAN/imap/squatter.c	2006-03-31 21:22:28.000000000 +0200
+++ cyrus-imapd-2.3-2.3.7/imap/squatter.c	2006-09-26 17:46:32.000000000 +0200
@@ -301,6 +301,13 @@
     char uid_validity_buf[30];
     char extname[MAX_MAILBOX_NAME+1];
     int use_annot = *((int *) rock);
+    int mbtype;
+
+    /* Skip remote mailboxes */
+    r = mboxlist_detail(name, &mbtype, NULL, NULL, NULL, NULL, NULL);
+    if (r) return r;
+    if (mbtype & MBTYPE_REMOTE)
+	    return 0;
 
     /* make sure the mailbox (or an ancestor) has
        /vendor/cmu/cyrus-imapd/squat set to "true" */
