--- backend.c.orig	2010-06-18 10:38:43.000000000 -0400
+++ backend.c	2010-06-18 10:20:51.000000000 -0400
@@ -444,6 +444,12 @@
 				  &ret->capability, AUTO_NO);
     }
 
+    if (!mechlist) {
+        syslog(LOG_ERR, "couldn't retrieve auth mechlist from backend server");
+        close(sock);
+        return NULL;
+    } 
+        
     /* now need to authenticate to backend server,
        unless we're doing LMTP/CSYNC on a UNIX socket (deliver/sync_client) */
     if ((server[0] != '/') ||
