diff -ru cyrus-imapd-2.4.12.orig/lib/libcyr_cfg.h cyrus-imapd-2.4.12/lib/libcyr_cfg.h
--- cyrus-imapd-2.4.12.orig/lib/libcyr_cfg.h	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/lib/libcyr_cfg.h	2011-10-08 18:06:50.780215073 +0000
@@ -45,8 +45,6 @@
 #ifndef INCLUDED_LIBCYR_CFG_H
 #define INCLUDED_LIBCYR_CFG_H
 
-#include <config.h>
-
 #include <stdio.h>
 #include <ctype.h>
 #include <string.h>
diff -ru cyrus-imapd-2.4.12.orig/lib/Makefile.in cyrus-imapd-2.4.12/lib/Makefile.in
--- cyrus-imapd-2.4.12.orig/lib/Makefile.in	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/lib/Makefile.in	2011-10-08 18:07:36.692215311 +0000
@@ -43,12 +43,6 @@
 #
 # $Id: Makefile.in,v 1.77 2010/06/28 12:06:31 brong Exp $
 
-# Authorization namespace.
-AUTH=auth.o auth_krb.o auth_unix.o auth_krb5.o auth_pts.o
-
-# ACL interpretation.  Only one choice for now:
-ACL=acl_afs.o
-
 srcdir = @srcdir@
 top_srcdir = @top_srcdir@
 VPATH = @srcdir@
@@ -78,33 +72,28 @@
 BUILTSOURCES = imapopts.h imapopts.c
 
 LIBCYR_HDRS = $(srcdir)/acl.h $(srcdir)/assert.h $(srcdir)/auth.h \
-	$(srcdir)/bsearch.h $(srcdir)/charset.h $(srcdir)/glob.h \
-	$(srcdir)/gmtoff.h $(srcdir)/imclient.h $(srcdir)/imparse.h \
-	$(srcdir)/cyr_lock.h $(srcdir)/map.h $(srcdir)/mkgmtime.h \
-	$(srcdir)/nonblock.h $(srcdir)/parseaddr.h $(srcdir)/prot.h \
-	$(srcdir)/retry.h $(srcdir)/sysexits.h $(srcdir)/strhash.h \
-	$(srcdir)/lsort.h $(srcdir)/stristr.h $(srcdir)/signals.h \
-	$(srcdir)/util.h $(srcdir)/xstrlcpy.h $(srcdir)/xstrlcat.h \
-	$(srcdir)/xmalloc.h $(srcdir)/imapurl.h \
-	$(srcdir)/cyrusdb.h $(srcdir)/iptostring.h $(srcdir)/rfc822date.h \
-	$(srcdir)/libcyr_cfg.h $(srcdir)/byteorder64.h \
-	$(srcdir)/md5.h $(srcdir)/crc32.h
-
-LIBCYR_OBJS = acl.o bsearch.o charset.o glob.o retry.o util.o \
-	libcyr_cfg.o mkgmtime.o prot.o parseaddr.o imclient.o imparse.o \
-	lsort.o stristr.o rfc822date.o signals.o cyrusdb.o strhash.o \
-	chartable.o imapurl.o nonblock_@WITH_NONBLOCK@.o lock_@WITH_LOCK@.o \
-	gmtoff_@WITH_GMTOFF@.o map_@WITH_MAP@.o $(ACL) $(AUTH) \
-	@LIBOBJS@ @CYRUSDB_OBJS@  \
-	iptostring.o xmalloc.o wildmat.o byteorder64.o \
-	xstrlcat.o xstrlcpy.o crc32.o
-
-LIBCYRM_HDRS = $(srcdir)/hash.h $(srcdir)/mpool.h $(srcdir)/xmalloc.h \
-	$(srcdir)/xstrlcat.h $(srcdir)/xstrlcpy.h $(srcdir)/util.h \
-	$(srcdir)/strhash.h $(srcdir)/libconfig.h $(srcdir)/assert.h \
-	imapopts.h
-LIBCYRM_OBJS = libconfig.o imapopts.o hash.o mpool.o xmalloc.o strhash.o \
-	xstrlcat.o xstrlcpy.o assert.o util.o signals.o @IPV6_OBJS@
+	$(srcdir)/bsearch.h $(srcdir)/charset.h $(srcdir)/cyr_lock.h \
+	$(srcdir)/cyrusdb.h $(srcdir)/glob.h $(srcdir)/gmtoff.h \
+	$(srcdir)/imapurl.h $(srcdir)/imclient.h $(srcdir)/imparse.h \
+	$(srcdir)/iptostring.h $(srcdir)/libcyr_cfg.h $(srcdir)/lsort.h \
+	$(srcdir)/map.h $(srcdir)/mkgmtime.h $(srcdir)/nonblock.h \
+	$(srcdir)/parseaddr.h $(srcdir)/retry.h $(srcdir)/rfc822date.h \
+	$(srcdir)/signals.h $(srcdir)/stristr.h $(srcdir)/sysexits.h
+
+LIBCYR_OBJS = acl.o acl_afs.o auth.o auth_krb.o auth_krb5.o auth_pts.o \
+	auth_unix.o bsearch.o byteorder64.o charset.o chartable.o crc32.o \
+	cyrusdb.o glob.o gmtoff_@WITH_GMTOFF@.o imapurl.o imclient.o \
+	imparse.o iptostring.o libcyr_cfg.o lock_@WITH_LOCK@.o lsort.o \
+	map_@WITH_MAP@.o mkgmtime.o nonblock_@WITH_NONBLOCK@.o parseaddr.o \
+	prot.o retry.o rfc822date.o signals.o stristr.o wildmat.o \
+	@CYRUSDB_OBJS@
+
+LIBCYRM_HDRS = $(srcdir)/assert.h $(scrdir)/exitcodes.h $(srcdir)/hash.h \
+	imapopts.h $(srcdir)/libconfig.h $(srcdir)/mpool.h \
+	$(srcdir)/strhash.h $(srcdir)/xmalloc.h
+
+LIBCYRM_OBJS = assert.o hash.o imapopts.o libconfig.o mpool.o strhash.o \
+	util.o xmalloc.o xstrlcat.o xstrlcpy.o @IPV6_OBJS@ @LIBOBJS@ 
 
 all: $(BUILTSOURCES) libcyrus_min.a libcyrus.a
 
diff -ru cyrus-imapd-2.4.12.orig/lib/prot.h cyrus-imapd-2.4.12/lib/prot.h
--- cyrus-imapd-2.4.12.orig/lib/prot.h	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/lib/prot.h	2011-10-08 18:06:50.892215072 +0000
@@ -52,6 +52,7 @@
 
 #include <sasl/sasl.h>
 
+#include "config.h"
 #ifdef HAVE_SSL
 #include <openssl/ssl.h>
 #endif /* HAVE_SSL */
