--- master.c.original	2005-09-19 21:48:33.000000000 +0200
+++ master.c	2005-09-19 22:11:10.000000000 +0200
@@ -1337,14 +1337,21 @@
 	/* skip non-primary instances */
 	if (Services[i].associate > 0)
 	    continue;
-	/* must have empty/same service name, listen and proto */
-	if ((!Services[i].name || !strcmp(Services[i].name, name)) &&
-	    (!Services[i].listen || !strcmp(Services[i].listen, listen)) &&
-	    (!Services[i].proto || !strcmp(Services[i].proto, proto)))
+	/* must have empty/same service name, listen and proto
+	 * and must not have the same service name with different
+	 * listen or proto
+	 */
+	if (Services[i].name && !strcmp(Services[i].name, name) &&
+	    ((Services[i].listen && strcmp(Services[i].listen, listen)) ||
+	    (Services[i].proto && strcmp(Services[i].proto, proto))))
 	    break;
+	else if ((!Services[i].name || !strcmp(Services[i].name, name)) &&
+	         (!Services[i].listen || !strcmp(Services[i].listen, listen)) &&
+	         (!Services[i].proto || !strcmp(Services[i].proto, proto)))
+	         break;
     }
 
-    /* we have duplicate service names in the config file */
+    /* we have duplicate service names or definitions in the config file */
     if ((i < nservices) && Services[i].exec) {
 	char buf[256];
 	snprintf(buf, sizeof(buf), "multiple entries for service '%s'", name);
