diff -ruN cyrus-sasl-2.1.24.prepatched/plugins/Makefile.am cyrus-sasl-2.1.24/plugins/Makefile.am
--- cyrus-sasl-2.1.24.prepatched/plugins/Makefile.am	2010-04-29 05:35:29.000000000 +0000
+++ cyrus-sasl-2.1.24/plugins/Makefile.am	2010-04-29 05:38:35.000000000 +0000
@@ -122,7 +122,8 @@
 libldapdb_la_LIBADD = $(LIB_LDAP) $(COMPAT_OBJS)
 
 libsql_la_SOURCES = sql.c sql_init.c $(common_sources)
-libsql_la_LDFLAGS = $(LIB_MYSQL) $(LIB_PGSQL) $(LIB_SQLITE) $(AM_LDFLAGS)
+libsql_la_LDFLAGS = $(LIB_MYSQL) $(LIB_PGSQL) $(LIB_SQLITE) $(LIB_SQLITE3) \
+	$(AM_LDFLAGS)
 libsql_la_DEPENDENCIES = $(COMPAT_OBJS)
 libsql_la_LIBADD = $(COMPAT_OBJS)
 
