--- cyrus-imapd-2.4.11/imap/mbpath.c.orig	2011-09-21 11:25:35.565792296 +0200
+++ cyrus-imapd-2.4.11/imap/mbpath.c	2011-09-21 11:26:47.538723687 +0200
@@ -135,7 +135,7 @@
 
     if ((rc = mboxlist_lookup(buf, &mbentry, NULL)) == 0) {
       if (mbentry.mbtype & MBTYPE_REMOTE) {
-	fprintf(stderr, "Remote mailbox: %s\n", argv[i]);
+	printf("%s\n", mbentry.partition);
       } else {
 	char *path = mboxname_metapath(mbentry.partition, mbentry.name, 0, 0);
 	printf("%s\n", path);
