diff -ur cyrus-imapd-2.4.11.orig/imap/tls.c cyrus-imapd-2.4.11/imap/tls.c
--- cyrus-imapd-2.4.11.orig/imap/tls.c	2011-09-08 18:57:42.000000000 +0000
+++ cyrus-imapd-2.4.11/imap/tls.c	2011-10-04 11:22:53.201359702 +0000
@@ -814,7 +814,7 @@
     int     sts;
     int     j;
     unsigned int n;
-    SSL_CIPHER *cipher;
+    const SSL_CIPHER *cipher;
     X509   *peer;
     const char *tls_protocol = NULL;
     const char *tls_cipher_name = NULL;
@@ -1256,7 +1256,7 @@
 			SSL_SESSION **sess)
 {
     int     sts;
-    SSL_CIPHER *cipher;
+    const SSL_CIPHER *cipher;
     X509   *peer;
     const char *tls_protocol = NULL;
     const char *tls_cipher_name = NULL;
diff -ur cyrus-imapd-2.4.11.orig/imtest/imtest.c cyrus-imapd-2.4.11/imtest/imtest.c
--- cyrus-imapd-2.4.11.orig/imtest/imtest.c	2011-09-08 18:57:42.000000000 +0000
+++ cyrus-imapd-2.4.11/imtest/imtest.c	2011-10-04 11:24:03.334359735 +0000
@@ -652,7 +652,7 @@
 int tls_start_clienttls(unsigned *layer, char **authid)
 {
     int     sts;
-    SSL_CIPHER *cipher;
+    const SSL_CIPHER *cipher;
     X509   *peer;
     
     if (verbose==1)
diff -ur cyrus-imapd-2.4.11.orig/lib/imclient.c cyrus-imapd-2.4.11/lib/imclient.c
--- cyrus-imapd-2.4.11.orig/lib/imclient.c	2011-09-08 18:57:42.000000000 +0000
+++ cyrus-imapd-2.4.11/lib/imclient.c	2011-10-04 11:21:56.006359672 +0000
@@ -1883,7 +1883,7 @@
 {
     int     sts;
     SSL_SESSION *session;
-    SSL_CIPHER *cipher;
+    const SSL_CIPHER *cipher;
     X509   *peer;
     int tls_cipher_usebits = 0;
     int tls_cipher_algbits = 0;
diff -ur cyrus-imapd-2.4.11.orig/sieve/addr.y cyrus-imapd-2.4.11/sieve/addr.y
--- cyrus-imapd-2.4.11.orig/sieve/addr.y	2011-09-08 18:57:42.000000000 +0000
+++ cyrus-imapd-2.4.11/sieve/addr.y	2011-10-04 11:31:18.101360018 +0000
@@ -59,7 +59,7 @@
 
 #define ADDRERR_SIZE 500
     
-int yyerror(char *msg);
+int yyerror(const char *msg);
 extern int yylex(void);
 
 #define YYERROR_VERBOSE /* i want better error messages! */
@@ -111,7 +111,7 @@
 %%
 
 /* copy address error message into buffer provided by sieve parser */
-int yyerror(char *s)
+int yyerror(const char *s)
 {
     extern char addrerr[ADDRERR_SIZE];
     
diff -ur cyrus-imapd-2.4.11.orig/sieve/sieve.c cyrus-imapd-2.4.11/sieve/sieve.c
diff -ur cyrus-imapd-2.4.11.orig/sieve/sieve.y cyrus-imapd-2.4.11/sieve/sieve.y
--- cyrus-imapd-2.4.11.orig/sieve/sieve.y	2011-09-08 18:57:42.000000000 +0000
+++ cyrus-imapd-2.4.11/sieve/sieve.y	2011-10-04 11:30:41.049359937 +0000
@@ -162,7 +162,7 @@
 #endif
 static int verify_utf8(char *s);
 
-int yyerror(char *msg);
+int yyerror(const char *msg);
 extern int yylex(void);
 extern void yyrestart(FILE *f);
 
@@ -713,7 +713,7 @@
     return t;
 }
 
-int yyerror(char *msg)
+int yyerror(const char *msg)
 {
     extern int yylineno;
     int ret;
