Index: sasl2.m4
===================================================================
RCS file: /afs/andrew.cmu.edu/system/cvs/src/cmulocal/sasl2.m4,v
retrieving revision 1.17.4.4
diff -u -r1.17.4.4 sasl2.m4
--- sasl2.m4	2002/08/21 15:32:21	1.17.4.4
+++ sasl2.m4	2002/09/30 20:33:43
@@ -202,6 +202,10 @@
 
 AC_DEFUN(CMU_SASL2_REQUIRE_VER, [
 	AC_REQUIRE([CMU_SASL2_REQUIRED])
+
+	cmu_saved_CPPFLAGS=$CPPFLAGS
+	CPPFLAGS="$CPPFLAGS $SASLFLAGS"
+
 	AC_TRY_CPP([
 #include <sasl/sasl.h>
 
@@ -219,6 +223,8 @@
 #error SASL version is less than $1.$2.$3
 #endif
 	],,AC_ERROR([Incorrect SASL headers found.  This package requires SASL $1.$2.$3 or newer.]))
+
+	CPPFLAGS=$cmu_saved_CPPFLAGS
 ])
 
 AC_DEFUN(CMU_SASL2_CHECKAPOP_REQUIRED, [
