diff -u aclocal.m4.orig aclocal.m4
--- aclocal.m4.orig	Mon Apr 14 20:07:47 2003
+++ aclocal.m4	Tue Apr 15 01:39:33 2003
@@ -1061,7 +1061,7 @@
 	    BDB_LIBADD=""
 	fi
 
-        for dbname in db-4.1 db4.1 db-4.0 db4.0 db-4 db4 db-3.3 db3.3 db-3.2 db3.2 db-3.1 db3.1 db-3 db3 db
+        for dbname in db-4.1 db4.1 db41 db-4.0 db4.0 db-4 db4 db-3.3 db3.3 db-3.2 db3.2 db-3.1 db3.1 db-3 db3 db
           do
             AC_CHECK_LIB($dbname, db_create, BDB_LIBADD="$BDB_LIBADD -l$dbname";
               dblib="berkeley"; break, dblib="no")
diff -Nur cyrus-sasl2/files/patch-cmulocal::berkdb.m4.orig cyrus-sasl2/files/patch-cmulocal::berkdb.m4
--- cmulocal/berkdb.m4.orig	Mon Mar 24 18:49:59 2003
+++ cmulocal/berkdb.m4	Tue Apr 15 01:39:33 2003
@@ -211,7 +211,7 @@
 	    BDB_LIBADD=""
 	fi
 
-        for dbname in db-4.1 db4.1 db-4.0 db4.0 db-4 db4 db-3.3 db3.3 db-3.2 db3.2 db-3.1 db3.1 db-3 db3 db
+        for dbname in db-4.1 db4.1 db41 db-4.0 db4.0 db-4 db4 db-3.3 db3.3 db-3.2 db3.2 db-3.1 db3.1 db-3 db3 db
           do
             AC_CHECK_LIB($dbname, db_create, BDB_LIBADD="$BDB_LIBADD -l$dbname";
               dblib="berkeley"; break, dblib="no")
diff -u config/ltmain.sh.orig config/ltmain.sh
 --- config/ltmain.sh.orig	Tue Sep  3 23:59:40 2002
 +++ config/ltmain.sh	Sat Oct 12 08:03:59 2002
@@ -1894,7 +1894,7 @@
 	  if test $? -eq 0 ; then
 	    ldd_output=`ldd conftest`
 	    for i in $deplibs; do
-	      name="`expr $i : '-l\(.*\)'`"
+	      name="`expr X$i : 'X-l\(.*\)'`"
 	      # If $name is empty we are operating on a -L argument.
 	      if test "$name" != "" ; then
 		libname=`eval \\$echo \"$libname_spec\"`
@@ -1919,7 +1919,7 @@
 	    # Error occured in the first compile.  Let's try to salvage the situation:
 	    # Compile a seperate program for each library.
 	    for i in $deplibs; do
-	      name="`expr $i : '-l\(.*\)'`"
+	      name="`expr X$i : 'X-l\(.*\)'`"
 	     # If $name is empty we are operating on a -L argument.
 	      if test "$name" != "" ; then
 		$rm conftest
@@ -1959,7 +1959,7 @@
 	  set dummy $deplibs_check_method
 	  file_magic_regex="`expr \"$deplibs_check_method\" : \"$2 \(.*\)\"`"
 	  for a_deplib in $deplibs; do
-	    name="`expr $a_deplib : '-l\(.*\)'`"
+	    name="`expr X$a_deplib : 'X-l\(.*\)'`"
 	    # If $name is empty we are operating on a -L argument.
 	    if test "$name" != "" ; then
 	      libname=`eval \\$echo \"$libname_spec\"`
diff -u configure.orig configure
--- configure.orig	Mon Apr 14 20:07:48 2003
+++ configure	Tue Apr 15 01:39:33 2003
@@ -2652,7 +2652,7 @@
 	    BDB_LIBADD=""
 	fi
 
-        for dbname in db-4.1 db4.1 db-4.0 db4.0 db-4 db4 db-3.3 db3.3 db-3.2 db3.2 db-3.1 db3.1 db-3 db3 db
+        for dbname in db-4.1 db4.1 db41 db-4.0 db4.0 db-4 db4 db-3.3 db3.3 db-3.2 db3.2 db-3.1 db3.1 db-3 db3 db
           do
             echo $ac_n "checking for db_create in -l$dbname""... $ac_c" 1>&6
 echo "configure:2659: checking for db_create in -l$dbname" >&5
@@ -3043,7 +3043,7 @@
 	    BDB_LIBADD=""
 	fi
 
-        for dbname in db-4.1 db4.1 db-4.0 db4.0 db-4 db4 db-3.3 db3.3 db-3.2 db3.2 db-3.1 db3.1 db-3 db3 db
+        for dbname in db-4.1 db4.1 db41 db-4.0 db4.0 db-4 db4 db-3.3 db3.3 db-3.2 db3.2 db-3.1 db3.1 db-3 db3 db
           do
             echo $ac_n "checking for db_create in -l$dbname""... $ac_c" 1>&6
 echo "configure:3050: checking for db_create in -l$dbname" >&5
diff -u saslauthd/aclocal.m4.orig saslauthd/aclocal.m4
--- saslauthd/aclocal.m4.orig	Mon Apr 14 20:07:50 2003
+++ saslauthd/aclocal.m4	Tue Apr 15 01:39:33 2003
@@ -1034,7 +1034,7 @@
 	    BDB_LIBADD=""
 	fi
 
-        for dbname in db-4.1 db4.1 db-4.0 db4.0 db-4 db4 db-3.3 db3.3 db-3.2 db3.2 db-3.1 db3.1 db-3 db3 db
+        for dbname in db-4.1 db4.1 db41 db-4.0 db4.0 db-4 db4 db-3.3 db3.3 db-3.2 db3.2 db-3.1 db3.1 db-3 db3 db
           do
             AC_CHECK_LIB($dbname, db_create, BDB_LIBADD="$BDB_LIBADD -l$dbname";
               dblib="berkeley"; break, dblib="no")
diff -u saslauthd/config/ltmain.sh.orig saslauthd/config/ltmain.sh
--- saslauthd/config/ltmain.sh.orig	Tue Dec  4 03:06:57 2001
+++ saslauthd/config/ltmain.sh	Tue Apr 15 02:33:18 2003
@@ -1861,7 +1861,7 @@
 	  if test $? -eq 0 ; then
 	    ldd_output=`ldd conftest`
 	    for i in $deplibs; do
-	      name="`expr $i : '-l\(.*\)'`"
+	      name="`expr X$i : 'X-l\(.*\)'`"
 	      # If $name is empty we are operating on a -L argument.
 	      if test "$name" != "" ; then
 		libname=`eval \\$echo \"$libname_spec\"`
@@ -1886,7 +1886,7 @@
 	    # Error occured in the first compile.  Let's try to salvage the situation:
 	    # Compile a seperate program for each library.
 	    for i in $deplibs; do
-	      name="`expr $i : '-l\(.*\)'`"
+	      name="`expr X$i : 'X-l\(.*\)'`"
 	     # If $name is empty we are operating on a -L argument.
 	      if test "$name" != "" ; then
 		$rm conftest
@@ -1926,7 +1926,7 @@
 	  set dummy $deplibs_check_method
 	  file_magic_regex="`expr \"$deplibs_check_method\" : \"$2 \(.*\)\"`"
 	  for a_deplib in $deplibs; do
-	    name="`expr $a_deplib : '-l\(.*\)'`"
+	    name="`expr X$a_deplib : 'X-l\(.*\)'`"
 	    # If $name is empty we are operating on a -L argument.
 	    if test "$name" != "" ; then
 	      libname=`eval \\$echo \"$libname_spec\"`
diff -u saslauthd/configure.orig saslauthd/configure
--- saslauthd/configure.orig	Mon Apr 14 20:07:50 2003
+++ saslauthd/configure	Tue Apr 15 01:39:33 2003
@@ -3441,7 +3441,7 @@
 	    BDB_LIBADD=""
 	fi
 
-        for dbname in db-4.1 db4.1 db-4.0 db4.0 db-4 db4 db-3.3 db3.3 db-3.2 db3.2 db-3.1 db3.1 db-3 db3 db
+        for dbname in db-4.1 db4.1 db41 db-4.0 db4.0 db-4 db4 db-3.3 db3.3 db-3.2 db3.2 db-3.1 db3.1 db-3 db3 db
           do
             echo $ac_n "checking for db_create in -l$dbname""... $ac_c" 1>&6
 echo "configure:3448: checking for db_create in -l$dbname" >&5
@@ -3832,7 +3832,7 @@
 	    BDB_LIBADD=""
 	fi
 
-        for dbname in db-4.1 db4.1 db-4.0 db4.0 db-4 db4 db-3.3 db3.3 db-3.2 db3.2 db-3.1 db3.1 db-3 db3 db
+        for dbname in db-4.1 db4.1 db41 db-4.0 db4.0 db-4 db4 db-3.3 db3.3 db-3.2 db3.2 db-3.1 db3.1 db-3 db3 db
           do
             echo $ac_n "checking for db_create in -l$dbname""... $ac_c" 1>&6
 echo "configure:3839: checking for db_create in -l$dbname" >&5
