diff -Naur cyrus-imapd-2.3.7.orig/imap/append.c cyrus-imapd-2.3.7/imap/append.c
--- cyrus-imapd-2.3.7.orig/imap/append.c	2006-10-03 17:46:03.000000000 +0200
+++ cyrus-imapd-2.3.7/imap/append.c	2006-10-04 10:53:13.000000000 +0200
@@ -652,6 +652,9 @@
     /* ok, we've successfully added a message */
     as->quota_used += message_index.size;
 
+    syslog(LOG_DEBUG, "append_fromstage: message %d added to %s",
+	   message_index.uid, mailbox->name );
+
     return 0;
 }
 
diff -Naur cyrus-imapd-2.3.7.orig/imap/imapd.c cyrus-imapd-2.3.7/imap/imapd.c
--- cyrus-imapd-2.3.7.orig/imap/imapd.c	2006-10-03 17:46:04.000000000 +0200
+++ cyrus-imapd-2.3.7/imap/imapd.c	2006-10-04 10:53:13.000000000 +0200
@@ -3494,6 +3494,8 @@
     else {
 	prot_printf(imapd_out, "%s OK %s\r\n", tag,
 		    error_message(IMAP_OK_COMPLETED));
+	syslog(LOG_DEBUG, "cmd_expunge: user %s, mailbox %s, sequence %s",
+	       imapd_userid, imapd_mailbox->name, "''");
     }
 }    
 
diff -Naur cyrus-imapd-2.3.7.orig/imap/mailbox.c cyrus-imapd-2.3.7/imap/mailbox.c
--- cyrus-imapd-2.3.7.orig/imap/mailbox.c	2006-10-03 17:46:04.000000000 +0200
+++ cyrus-imapd-2.3.7/imap/mailbox.c	2006-10-04 10:53:27.000000000 +0200
@@ -2455,6 +2455,8 @@
 	*(fname->tail)++ = '/';
 	fname->len++;
 	for (msgno = 0; msgno < numdeleted; msgno++) {
+	    syslog(LOG_DEBUG, "mailbox_expunge: removing mail %s:%d",
+		   mailbox->name, deleted[msgno]);
 	    mailbox_message_get_fname(mailbox, deleted[msgno],
 				      fname->tail,
 				      sizeof(fname->buf) - fname->len);
