*** ctl_mboxlist.Oc	Mon Feb  5 12:41:46 2007
--- ctl_mboxlist.c	Sun Mar  4 15:44:04 2007
***************
*** 267,279 ****
  		/* since it lives on another server, schedule it for a wipe */
  		struct mb_node *next;
  		
! 		if(warn_only) {
  		    printf("Remove Local Mailbox: %s\n", name);
! 		} else {
  		    next = xzmalloc(sizeof(struct mb_node));
  		    strlcpy(next->mailbox, name, sizeof(next->mailbox));
  		    next->next = wipe_head;
  		    wipe_head = next;
  		}
  		
  		skip_flag = 1;		
--- 267,283 ----
  		/* since it lives on another server, schedule it for a wipe */
  		struct mb_node *next;
  		
! 		if (config_mupdate_config !=
! 		    IMAP_ENUM_MUPDATE_CONFIG_UNIFIED) {
! 		  /* But not for a unified configuration */
! 		  if(warn_only) {
  		    printf("Remove Local Mailbox: %s\n", name);
! 		  } else {
  		    next = xzmalloc(sizeof(struct mb_node));
  		    strlcpy(next->mailbox, name, sizeof(next->mailbox));
  		    next->next = wipe_head;
  		    wipe_head = next;
+ 		  }
  		}
  		
  		skip_flag = 1;		
