diff -ur imapd-2.2.8-old/sieve/tree.h imapd-2.2.8-new/sieve/tree.h
--- imapd-2.2.8-old/sieve/tree.h	2004-08-29 23:44:18.000000000 -0400
+++ imapd-2.2.8-new/sieve/tree.h	2004-08-29 23:37:04.000000000 -0400
@@ -69,8 +69,8 @@
 	} h;
 	struct { /* it's an address or envelope test */
 	    int comptag;
-	    int relation; 
 	    char * comparator;
+	    int relation; 
 	    void *comprock;
 	    stringlist_t *sl;
 	    stringlist_t *pl;
