diff -u -r cyrus-imapd-2.3.11-old/imap/tls.c cyrus-imapd-2.3.11-new/imap/tls.c
--- cyrus-imapd-2.3.11-old/imap/tls.c	2007-11-28 12:51:27.000000000 +0100
+++ cyrus-imapd-2.3.11-new/imap/tls.c	2008-04-14 21:06:37.000000000 +0200
@@ -403,6 +403,10 @@
 			  const char *cert_file, const char *key_file)
 {
     if (cert_file != NULL) {
+	/* SSL_CTX_use_certificate_chain_file() requires an empty error stack.
+	 * To make sure there is no error from previous operation, we clear it here...
+	 */
+	ERR_clear_error();
 	if (SSL_CTX_use_certificate_chain_file(ctx, cert_file) <= 0) {
 	    syslog(LOG_ERR, "unable to get certificate from '%s'", cert_file);
 	    return (0);
