diff -urNad cyrus-imapd-2.3-2.3.7~/imap/cyr_expire.c cyrus-imapd-2.3-2.3.7/imap/cyr_expire.c
--- cyrus-imapd-2.3-2.3.7~/imap/cyr_expire.c	2006-10-03 16:23:39.000000000 +0200
+++ cyrus-imapd-2.3-2.3.7/imap/cyr_expire.c	2006-10-03 16:25:32.000000000 +0200
@@ -132,6 +132,18 @@
     char buf[MAX_MAILBOX_NAME+1] = "", *p;
     struct annotation_data attrib;
     int r, domainlen = 0;
+    int mbtype;
+
+    /* Skip remote mailboxes */
+    r = mboxlist_detail(name, &mbtype, NULL, NULL, NULL, NULL, NULL);
+    if (r) {
+        if (erock->verbose) {
+            printf("error looking up %s: %s\n",
+		    name, error_message(r));
+        }
+        return 1;
+    }
+    if (mbtype & MBTYPE_REMOTE) return 0;
 
     if (config_virtdomains && (p = strchr(name, '!')))
 	domainlen = p - name + 1;

