diff -Naur cyrus-imapd-2.3.1.orig/imap/backend.c cyrus-imapd-2.3.1/imap/backend.c
--- cyrus-imapd-2.3.1.orig/imap/backend.c	2005-02-25 07:46:14.000000000 +0100
+++ cyrus-imapd-2.3.1/imap/backend.c	2006-01-25 08:10:20.000000000 +0100
@@ -274,7 +274,7 @@
     /* need to (re)establish connection to server or create one */
     int sock = -1;
     int r;
-    int err;
+    int err = -1;
     struct addrinfo hints, *res0 = NULL, *res1 = NULL, *res;
     struct sockaddr_un sunsock;
     char buf[2048], *mechlist = NULL;
