diff -ruN cyrus-imapd-2.3.16.bak/doc/install-configure.html cyrus-imapd-2.3.16/doc/install-configure.html
--- cyrus-imapd-2.3.16.bak/doc/install-configure.html	2008-01-07 17:56:09.000000000 +0000
+++ cyrus-imapd-2.3.16/doc/install-configure.html	2010-05-24 02:31:36.000000000 +0000
@@ -193,8 +193,8 @@
    imaps     993/tcp
    pop3s     995/tcp
    kpop      1109/tcp
-   sieve     2000/tcp
    lmtp      2003/tcp
+   sieve     4190/tcp
    fud       4201/udp
 </pre>
 
diff -ruN cyrus-imapd-2.3.16.bak/doc/install-sieve.html cyrus-imapd-2.3.16/doc/install-sieve.html
--- cyrus-imapd-2.3.16.bak/doc/install-sieve.html	2001-10-23 18:14:48.000000000 +0000
+++ cyrus-imapd-2.3.16/doc/install-sieve.html	2010-05-24 02:26:06.000000000 +0000
@@ -13,10 +13,6 @@
 <tt>./configure</tt>, you did NOT compile the server with sieve
 support. 
 
-<p>A <a href="http://www.cyrusoft.com/sieve">brief introduction</a> to
-Sieve at is available from <a href="http://www.cyrusoft.com/">Cyrusoft
-International</a>.
-
 <h3>Configure outgoing mail</h3>
 
 <p>Some Sieve actions (redirect, vacation) can send outgoing mail.</p>
@@ -48,7 +44,7 @@
 <ol>
 
 <p>
-<li>The Sieve server, "<tt>timsieved</tt>", is used for transporting
+<li>"<tt>timsieved</tt>" is a ManageSieve server used for transporting
 user Sieve scripts to the sealed IMAP server.  It is incompatible with
 the "<tt>sieveusehomedir</tt>" option.  It is named after the
 principal author, Tim Martin, who desperately wanted something named
@@ -85,7 +81,7 @@
 <li>Next test authenticating to the sieve server. To do this run the
 "<tt>sieveshell</tt>" utility. You must specify the server.  If you
 run this utility from a different machine without the "sieve" entry in
-"/etc/services", port 2000 will be used.
+"/etc/services", port 4190 will be used.
 
 <pre>
   "<kbd>sieveshell foobar</kbd>"
diff -ruN cyrus-imapd-2.3.16.bak/doc/text/install-configure cyrus-imapd-2.3.16/doc/text/install-configure
--- cyrus-imapd-2.3.16.bak/doc/text/install-configure	2009-12-21 13:18:00.000000000 +0000
+++ cyrus-imapd-2.3.16/doc/text/install-configure	2010-05-24 02:31:07.000000000 +0000
@@ -152,8 +152,8 @@
            imaps     993/tcp
            pop3s     995/tcp
            kpop      1109/tcp
-           sieve     2000/tcp
            lmtp      2003/tcp
+           sieve     4190/tcp
            fud       4201/udp
 
     12. Remove "/etc/[x]inetd.conf" entries. Any imap, imaps, pop3, pop3s,
diff -ruN cyrus-imapd-2.3.16.bak/doc/text/install-sieve cyrus-imapd-2.3.16/doc/text/install-sieve
--- cyrus-imapd-2.3.16.bak/doc/text/install-sieve	2009-12-21 13:18:00.000000000 +0000
+++ cyrus-imapd-2.3.16/doc/text/install-sieve	2010-05-24 02:30:48.000000000 +0000
@@ -64,7 +64,7 @@
      3. Next test authenticating to the sieve server. To do this run the
         "sieveshell" utility. You must specify the server. If you run this
         utility from a different machine without the "sieve" entry in
-        "/etc/services", port 2000 will be used. 
+        "/etc/services", port 4190 will be used. 
 
           "sieveshell foobar"
             Please enter your password: ******
diff -ruN cyrus-imapd-2.3.16.bak/perl/sieve/lib/isieve.c cyrus-imapd-2.3.16/perl/sieve/lib/isieve.c
--- cyrus-imapd-2.3.16.bak/perl/sieve/lib/isieve.c	2009-02-16 18:18:10.000000000 +0000
+++ cyrus-imapd-2.3.16/perl/sieve/lib/isieve.c	2010-05-24 02:35:32.000000000 +0000
@@ -616,7 +616,7 @@
     } else {
 	serv = getservbyname("sieve", "tcp");
 	if (serv == NULL) {
-	    port = 2000;
+	    port = 4190;
 	} else {
 	    port = ntohs(serv->s_port);
 	}
diff -ruN cyrus-imapd-2.3.16.bak/perl/sieve/managesieve/managesieve.c cyrus-imapd-2.3.16/perl/sieve/managesieve/managesieve.c
--- cyrus-imapd-2.3.16.bak/perl/sieve/managesieve/managesieve.c	2010-01-17 21:08:52.000000000 +0000
+++ cyrus-imapd-2.3.16/perl/sieve/managesieve/managesieve.c	2010-05-24 02:33:36.000000000 +0000
@@ -261,7 +261,7 @@
       /* map port -> num */
       serv = getservbyname("sieve", "tcp");
       if (serv == NULL) {
-	  port = 2000;
+	  port = 4190;
       } else {
 	  port = ntohs(serv->s_port);
       }
diff -ruN cyrus-imapd-2.3.16.bak/perl/sieve/managesieve/managesieve.xs cyrus-imapd-2.3.16/perl/sieve/managesieve/managesieve.xs
--- cyrus-imapd-2.3.16.bak/perl/sieve/managesieve/managesieve.xs	2009-01-14 15:50:47.000000000 +0000
+++ cyrus-imapd-2.3.16/perl/sieve/managesieve/managesieve.xs	2010-05-24 02:32:27.000000000 +0000
@@ -241,7 +241,7 @@
       /* map port -> num */
       serv = getservbyname("sieve", "tcp");
       if (serv == NULL) {
-	  port = 2000;
+	  port = 4190;
       } else {
 	  port = ntohs(serv->s_port);
       }
