diff -urNad cyrus-imapd-2.2.13/configure.in /tmp/dpep.6LRag9/cyrus-imapd-2.2.13/configure.in
--- cyrus-imapd-2.2.13/configure.in	2006-09-25 20:13:04.000000000 +0200
+++ /tmp/dpep.6LRag9/cyrus-imapd-2.2.13/configure.in	2006-09-25 20:13:06.102258297 +0200
@@ -297,16 +297,15 @@
 fi
 if test "$spt_type" = ""; then
 	AC_MSG_CHECKING(for PS_STRINGS)
-	AC_CACHE_VAL(cyrus_cv_sys_psstrings, AC_TRY_CPP([
+AC_CACHE_VAL(cyrus_cv_sys_psstrings, AC_COMPILE_IFELSE([
 #include <machine/vmparam.h>
 #include <sys/exec.h>
-#ifndef PS_STRINGS
-#include </nonexistent>
-#endif],cyrus_cv_sys_psstrings=yes,cyrus_cv_sys_psstrings=no))
-	if test $cyrus_cv_sys_psstrings = yes; then
-		spt_type=SPT_PSSTRINGS
-	fi
-	AC_MSG_RESULT($cyrus_cv_sys_psstrings)
+void *p = PS_STRINGS;
+],cyrus_cv_sys_psstrings=yes,cyrus_cv_sys_psstrings=no))
+         if test $cyrus_cv_sys_psstrings = yes; then
+                 spt_type=SPT_PSSTRINGS
+         fi
+         AC_MSG_RESULT($cyrus_cv_sys_psstrings)
 fi
 if test "$spt_type" = ""; then
 	AC_MSG_CHECKING(for SCO)
