Index: imap/global.c
===================================================================
RCS file: /cvs/src/cyrus/imap/global.c,v
retrieving revision 1.23
diff -u -r1.23 global.c
--- imap/global.c	30 Mar 2007 18:51:00 -0000	1.23
+++ imap/global.c	26 Sep 2007 22:03:52 -0000
@@ -221,6 +221,9 @@
 			       config_getint(IMAPOPT_BERKELEY_TXNS_MAX));
 	libcyrus_config_setstring(CYRUSOPT_DELETERIGHT,
 				  config_getstring(IMAPOPT_DELETERIGHT));
+	libcyrus_config_setswitch(CYRUSOPT_NORMALIZEUID,
+ 				  config_getswitch(IMAPOPT_NORMALIZEUID));
+
 
 	/* Not until all configuration parameters are set! */
 	libcyrus_init();
