Index: saslauthd/auth_krb5.c
===================================================================
RCS file: /cvs/src/sasl/saslauthd/auth_krb5.c,v
retrieving revision 1.17
diff -u -r1.17 auth_krb5.c
--- saslauthd/auth_krb5.c	14 Feb 2005 05:50:49 -0000	1.17
+++ saslauthd/auth_krb5.c	26 Sep 2007 21:53:52 -0000
@@ -280,7 +280,7 @@
 	}
     }
     
-    if (krb5_kt_read_service_key(context, kt, server, 0,
+    if (krb5_kt_read_service_key(context, keytabname, server, 0,
 				 0, &keyblock)) {
 	goto fini;
     }
@@ -310,7 +310,7 @@
     }
     
     if (krb5_rd_req(context, &auth_context, &packet, 
-		    server, NULL, NULL, NULL)) {
+		    server, kt, NULL, NULL)) {
 	goto fini;
     }
 
@@ -322,6 +322,10 @@
     /* all is good now */
     result = 1;
  fini:
+    if (kt) {
+	krb5_kt_close(context, kt);
+	kt = NULL;
+    }
     krb5_free_data_contents(context, &packet);
     krb5_free_principal(context, server);
     
