From 1c059508ffcd44ae9363b34a9b1b66d3d9d4ca7b Mon Sep 17 00:00:00 2001
From: "Jeroen van Meeuwen (Kolab Systems)" <vanmeeuwen@kolabsys.com>
Date: Sun, 15 Apr 2012 11:39:11 +0100
Subject: [PATCH] Set COMPILE_ET to the full path, and use it (#3670)

---
 Makefile.am  |    8 ++++----
 configure.ac |    4 ++--
 2 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index 774a0e7..cb5a0ba 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -366,13 +366,13 @@ imap_unexpunge_SOURCES = imap/cli_fatal.c imap/mutex_fake.c imap/unexpunge.c
 imap_unexpunge_LDADD = $(ptclient_ptdump_LDADD)
 
 imap/imap_err.h imap/imap_err.c: imap/imap_err.et
-	cd imap && $(top_builddir)/../$(COMPILE_ET) ../$(top_srcdir)/imap/imap_err.et
+	cd imap && $(COMPILE_ET) ../$(top_srcdir)/imap/imap_err.et
 
 imap/mupdate_err.h imap/mupdate_err.c: imap/mupdate_err.et
-	cd imap && $(top_builddir)/../$(COMPILE_ET) ../$(top_srcdir)/imap/mupdate_err.et
+	cd imap && $(COMPILE_ET) ../$(top_srcdir)/imap/mupdate_err.et
 
 imap/nntp_err.h imap/nntp_err.c: imap/nntp_err.et
-	cd imap && $(top_builddir)/../$(COMPILE_ET) ../$(top_srcdir)/imap/nntp_err.et
+	cd imap && $(COMPILE_ET) ../$(top_srcdir)/imap/nntp_err.et
 
 imap/rfc822_header.c: imap/rfc822_header.st
 	${top_srcdir}/tools/compile_st.pl -c $< > $@
@@ -555,7 +555,7 @@ sieve_sievec_LDADD = sieve/libsieve.a lib/libcyrus.a lib/libcyrus_min.a ${COM_ER
 sieve_sieved_LDADD = ${sieve_sievec_LDADD}
 
 sieve/sieve_err.h sieve/sieve_err.c: sieve/sieve_err.et
-	cd sieve && $(top_builddir)/../$(COMPILE_ET) ../$(top_srcdir)/sieve/sieve_err.et
+	cd sieve && $(COMPILE_ET) ../$(top_srcdir)/sieve/sieve_err.et
 
 sieve_test_SOURCES = sieve/test.c imap/mutex_fake.c
 sieve_test_LDADD = imap/libimap.a ${sieve_sievec_LDADD}
diff --git a/configure.ac b/configure.ac
index eb4b1dd..972842d 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1035,8 +1035,8 @@ fi
 case "$with_com_err" in
   # built-in et
   yes) # use the com_err we're gonna build
-	  COM_ERR_LIBS="com_err/et/libcom_err.a"
-	  COMPILE_ET="com_err/et/compile_et"
+	  COM_ERR_LIBS="${top_srcdir}/com_err/et/libcom_err.a"
+	  COMPILE_ET="${top_srcdir}/com_err/et/compile_et"
 	  COM_ERR_CPPFLAGS="-I\${top_srcdir}/com_err/et"
 	  ;;
   "")   # no problem, we already have it in the paths
-- 
1.7.7.6

