diff -ru cyrus-imapd-2.2.8/imap/xversion.sh cyrus-imapd-2.2.8_patched/imap/xversion.sh
--- cyrus-imapd-2.2.8/imap/xversion.sh  2003-10-22 20:03:00.000000000 +0200
+++ cyrus-imapd-2.2.8_patched/imap/xversion.sh  2004-07-30 00:12:32.000000000 +0200
@@ -16,6 +16,10 @@
 
 printf "#define CYRUS_CVSDATE " >> xversion.h
 
+_CYRUS_POSIX2_VERSION=_POSIX2_VERSION
+_POSIX2_VERSION='199209'
+export _POSIX2_VERSION
+
 find .. -name '*.[chly]' -print | \
        xargs egrep '\$Id: ' | \
        $AWK ' # extract timestamp and surround with quotes
@@ -23,3 +27,6 @@
            printf "\"%s\"\n", substr($0, RSTART, RLENGTH)
        }' pattern="$DATEPAT $TIMEPAT" | \
        sort | tail -1 >> xversion.h
+
+_POSIX2_VERSION=_CYRUS_POSIX2_VERSION
+export _POSIX2_VERSION
