--- imap/proxy.c	24 Mar 2008 17:09:18 -0000	1.7
+++ imap/proxy.c	1 Sep 2008 02:42:56 -0000
@@ -268,7 +268,7 @@
 
 		    if (c == 0 || c < 0) break;
 		    prot_write(pout, buf, c);
-		} while (c == sizeof(buf));
+		} while (pin->cnt > 0);
 
 		if ((err = prot_error(pin)) != NULL) {
 		    if (serverout && !strcmp(err, PROT_EOF_STRING)) {
@@ -280,6 +280,8 @@
 			fatal("Lost connection to input stream",
 			      EC_UNAVAILABLE);
 		    }
+		} else {
+		    return 0;
 		}
 	    }
 	}
