--- imap/annotate.c.old	2006-05-24 14:39:10.000000000 -0700
+++ imap/annotate.c	2006-05-24 14:39:40.000000000 -0700
@@ -703,7 +703,7 @@
     struct mailbox mailbox;
     struct index_record record;
     int r = 0, msg;
-    unsigned long totsize = 0;
+    unsigned long long totsize = 0;
     char value[21];
     struct annotation_data attrib;
 
@@ -735,7 +735,7 @@
 
     mailbox_close(&mailbox);
 
-    if (r || snprintf(value, sizeof(value), "%lu", totsize) == -1)
+    if (r || snprintf(value, sizeof(value), "%llu", totsize) == -1)
 	return;
 
     memset(&attrib, 0, sizeof(attrib));
