Index: cyrus-imapd-2.2.1-BETA/acconfig.h
diff -u cyrus-imapd-2.2.1-BETA/acconfig.h.orig cyrus-imapd-2.2.1-BETA/acconfig.h
--- cyrus-imapd-2.2.1-BETA/acconfig.h.orig	Tue Jul 15 01:40:07 2003
+++ cyrus-imapd-2.2.1-BETA/acconfig.h	Fri Jul 25 14:30:45 2003
@@ -241,7 +241,7 @@
 #define	AF_INET6	AF_MAX
 #endif
 
-#ifndef HAVE_GETADDRINFO
+#if !defined(HAVE_GETADDRINFO) || !defined(HAVE_GETNAMEINFO)
 #include "gai.h"
 #endif
 
Index: cyrus-imapd-2.2.1-BETA/lib/gai.h
diff -u cyrus-imapd-2.2.1-BETA/lib/gai.h.orig cyrus-imapd-2.2.1-BETA/lib/gai.h
--- cyrus-imapd-2.2.1-BETA/lib/gai.h.orig	Fri Feb 14 05:33:12 2003
+++ cyrus-imapd-2.2.1-BETA/lib/gai.h	Fri Jul 25 14:32:51 2003
@@ -89,11 +89,16 @@
 };
 #endif
 
-int	getaddrinfo(const char *, const char *,
-		    const struct addrinfo *, struct addrinfo **);
+#ifndef HAVE_GETNAMEINFO
 int	getnameinfo(const struct sockaddr *, socklen_t, char *,
 		    size_t, char *, size_t, int);
+#endif
+
+#ifndef HAVE_GETADDRINFO
+int	getaddrinfo(const char *, const char *,
+		    const struct addrinfo *, struct addrinfo **);
 void	freeaddrinfo(struct addrinfo *);
 char	*gai_strerror(int);
+#endif
 
 #endif
