--- cyrus-imapd-2.4.11.orig/imap/mboxlist.c	2011-10-03 11:01:23.282702645 +0200
+++ cyrus-imapd-2.4.11/imap/mboxlist.c	2011-10-03 11:04:11.516730179 +0200
@@ -1165,7 +1165,8 @@
 		r = mupdate_delete(mupdate_h, oldname);
 	    if (!r) r = mupdate_reserve(mupdate_h, newname, buf);
 	}
-	if (!r) r = mupdate_activate(mupdate_h, newname, buf, newmailbox->acl);
+	if (!r) r = mupdate_activate(mupdate_h, newname, buf, 
+					partitionmove ? oldmailbox->acl : newmailbox->acl);
 	if (r) {
 	    syslog(LOG_ERR,
 		   "MUPDATE: can't commit mailbox entry for '%s'",
