diff -ruN cyrus-imapd-2.4.6.bak//imtest/imtest.c cyrus-imapd-2.4.6/imtest/imtest.c
--- cyrus-imapd-2.4.6.bak//imtest/imtest.c	2011-02-02 05:49:08.666775228 +0000
+++ cyrus-imapd-2.4.6/imtest/imtest.c	2011-02-02 05:59:07.699818504 +0000
@@ -1485,10 +1485,6 @@
     }
 
     do { /* look for the end of the capabilities */
-	if (ret) {
-	    free(ret);
-	    ret = NULL;
-	}
 	if (prot_fgets(str, sizeof(str), pin) == NULL) {
 	    if (!*str) imtest_fatal("prot layer failure");
 	    else break;
@@ -1510,6 +1506,8 @@
 	/* check for auth */
 	if (prot->capa_cmd.auth &&
 	    (tmp = strstr(str, prot->capa_cmd.auth)) != NULL) {
+	    if (ret)
+		free(ret);
 	    if (prot->capa_cmd.parse_mechlist)
 		ret = prot->capa_cmd.parse_mechlist(str, prot);
 	    else
@@ -2278,7 +2276,7 @@
       { "STARTTLS", "220", "454", 0 },
       { "AUTH", 512, 0, "235", "5", "334 ", "*", NULL, 0 },
       { NULL, NULL, NULL, },
-      &xmtp_do_auth, { "QUIT", "221" },
+      NULL, { "QUIT", "221" },
       &xmtp_init_conn, &generic_pipe, &xmtp_reset
     },
     { "smtp", "smtps", "smtp",
@@ -2287,7 +2285,7 @@
       { "STARTTLS", "220", "454", 0 },
       { "AUTH", 512, 0, "235", "5", "334 ", "*", NULL, 0 },
       { NULL, NULL, NULL, },
-      &xmtp_do_auth, { "QUIT", "221" },
+      NULL, { "QUIT", "221" },
       &xmtp_init_conn, &generic_pipe, &xmtp_reset
     },
     { "mupdate", NULL, "mupdate",
