--- cyrus-imapd-2.3.15.orig/imtest/imtest.c	2009-04-23 19:10:07.000000000 +0200
+++ cyrus-imapd-2.3.15/imtest/imtest.c	2009-09-24 14:14:00.000000000 +0200
@@ -2564,11 +2564,13 @@
 	    prot_free(pin);
 	    prot_free(pout);
 
+#ifdef HAVE_SSL
 	    /* Properly shutdown TLS so that session can be reused */
 	    if (tls_conn) {
 		SSL_shutdown(tls_conn);
 		SSL_set_shutdown(tls_conn, SSL_SENT_SHUTDOWN|SSL_RECEIVED_SHUTDOWN);
 	    }
+#endif /* HAVE_SSL */
 	    
 	    close(sock);
 	    
