Index: imapd.c
===================================================================
RCS file: /afs/andrew/system/cvs/src/cyrus/imap/imapd.c,v
retrieving revision 1.310
diff -c -r1.310 imapd.c
*** imapd.c	2001/08/03 21:18:06	1.310
--- imapd.c	2001/08/05 14:43:14
***************
*** 384,391 ****
--- 384,396 ----
      imapd_starttls_done = 0;
  #ifdef HAVE_SSL
      if (tls_conn) {
+ #ifdef TLS_REUSE
+ 	/* make sure we re-use sessions */
+ 	SSL_set_shutdown(tls_conn,SSL_SENT_SHUTDOWN|SSL_RECEIVED_SHUTDOWN);
+ #else
  	tls_free(&tls_conn);
  	tls_conn = NULL;
+ #endif
      }
  #endif
  
Index: pop3d.c
===================================================================
RCS file: /afs/andrew/system/cvs/src/cyrus/imap/pop3d.c,v
retrieving revision 1.101
diff -c -r1.101 pop3d.c
*** pop3d.c	2001/08/03 21:18:08	1.101
--- pop3d.c	2001/08/05 14:43:14
***************
*** 191,198 ****
--- 191,203 ----
      popd_starttls_done = 0;
  #ifdef HAVE_SSL
      if (tls_conn) {
+ #ifdef TLS_REUSE
+ 	/* make sure we re-use sessions */
+ 	SSL_set_shutdown(tls_conn,SSL_SENT_SHUTDOWN|SSL_RECEIVED_SHUTDOWN);
+ #else
  	tls_free(&tls_conn);
  	tls_conn = NULL;
+ #endif
      }
  #endif
  
Index: tls.c
===================================================================
RCS file: /afs/andrew/system/cvs/src/cyrus/imap/tls.c,v
retrieving revision 1.14
diff -c -r1.14 tls.c
*** tls.c	2001/05/21 18:28:18	1.14
--- tls.c	2001/08/05 14:43:14
***************
*** 368,373 ****
--- 368,374 ----
  {
      int     off = 0;
      int     verify_flags = SSL_VERIFY_NONE;
+     int     session_id_context = 1; /* anything will do */
      char   *CApath;
      char   *CAfile;
      char   *s_cert_file;
***************
*** 442,447 ****
--- 443,450 ----
  	verify_flags |= SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT
  	    | SSL_VERIFY_CLIENT_ONCE;
      SSL_CTX_set_verify(ctx, verify_flags, verify_callback);
+     SSL_CTX_set_session_id_context(ctx,(void*)&session_id_context,
+ 				   sizeof session_id_context);
  
      if (askcert || requirecert) {
        if (CAfile == NULL) {
***************
*** 623,636 ****
      }
  
      if (authid && *authid) {
! 	syslog(LOG_NOTICE, "starttls: %s with cipher %s (%d/%d bits)"
  	                   " authenticated as %s", 
  	       tls_protocol, tls_cipher_name,
  	       tls_cipher_usebits, tls_cipher_algbits, 
  	       *authid);
      } else {
! 	syslog(LOG_NOTICE, "starttls: %s with cipher %s (%d/%d bits)"
  	                   " no authentication", 
  	       tls_protocol, tls_cipher_name,
  	       tls_cipher_usebits, tls_cipher_algbits);
      }
--- 626,641 ----
      }
  
      if (authid && *authid) {
! 	syslog(LOG_NOTICE, "starttls: %s, %s with cipher %s (%d/%d bits)"
  	                   " authenticated as %s", 
+ 	       tls_conn->hit ? "Reused" : "New",
  	       tls_protocol, tls_cipher_name,
  	       tls_cipher_usebits, tls_cipher_algbits, 
  	       *authid);
      } else {
! 	syslog(LOG_NOTICE, "starttls: %s, %s with cipher %s (%d/%d bits)"
  	                   " no authentication", 
+ 	       tls_conn->hit ? "Reused" : "New",
  	       tls_protocol, tls_cipher_name,
  	       tls_cipher_usebits, tls_cipher_algbits);
      }
