diff -Naur cyrus-sasl-2.1.21.old/saslauthd/configure.in cyrus-sasl-2.1.21/saslauthd/configure.in
--- cyrus-sasl-2.1.21.old/saslauthd/configure.in	2005-05-07 05:54:18.000000000 +0100
+++ cyrus-sasl-2.1.21/saslauthd/configure.in	2005-09-14 17:13:47.523634679 +0100
@@ -197,7 +197,7 @@
 IPv6_CHECK_FUNC(getaddrinfo,
 		[AC_DEFINE(HAVE_GETADDRINFO,[],[Do we have a getaddrinfo() function?])], [sasl_cv_getaddrinfo=yes])
 if test $sasl_cv_getaddrinfo = yes; then
-	AC_LIBOBJ(getaddrinfo)
+	AC_LIBOBJ(../lib/getaddrinfo)
 fi
 
 dnl Check for getnameinfo
@@ -206,7 +206,7 @@
 IPv6_CHECK_FUNC(getnameinfo,
 		[AC_DEFINE(HAVE_GETNAMEINFO,[],[Do we have a getnameinfo() function?])], [sasl_cv_getnameinfo=yes])
 if test $sasl_cv_getnameinfo = yes; then
-	AC_LIBOBJ(getnameinfo)
+	AC_LIBOBJ(../lib/getnameinfo)
 fi
 
 IPv6_CHECK_SS_FAMILY()
