--- pop3d.c.orig        2007-10-02 16:55:20.000000000 -0500
+++ pop3d.c     2007-10-02 16:54:35.000000000 -0500
@@ -1199,6 +1199,8 @@
 void cmd_pass(char *pass)
 {
     int plaintextloginpause;
+    int sasl_result;
+    char *canon_user;

     if (!popd_userid) {
        prot_printf(popd_out, "-ERR [AUTH] Must give USER command\r\n");
@@ -1258,6 +1260,42 @@
        return;
     }
     else {
+
+        /* successful authentication */
+
+        /* get the userid from SASL --- already canonicalized from
+         * mysasl_proxy_policy()
+         */
+        sasl_result = sasl_getprop(popd_saslconn, SASL_USERNAME,
+                                   (const void **) &canon_user);
+        if (sasl_result != SASL_OK) {
+            prot_printf(popd_out,
+                        "-ERR [AUTH] weird SASL error %d getting SASL_USERNAME\r\n",
+                        sasl_result);
+            return;
+        }
+
+        /* If we're proxying, the authzid may contain a subfolder,
+           so re-canonify it */
+        if (config_getswitch(IMAPOPT_POPSUBFOLDERS) && strchr(canon_user, '+')) {
+            char userbuf[MAX_MAILBOX_NAME+1];
+            unsigned userlen;
+
+            sasl_result = popd_canon_user(popd_saslconn, NULL, canon_user, 0,
+                                          SASL_CU_AUTHID | SASL_CU_AUTHZID,
+                                          NULL, userbuf, sizeof(userbuf), &userlen);
+            if (sasl_result != SASL_OK) {
+                prot_printf(popd_out,
+                            "-ERR [AUTH] SASL canonification error %d\r\n",
+                            sasl_result);
+                return;
+            }
+
+            popd_userid = xstrdup(userbuf);
+        } else {
+            popd_userid = xstrdup(canon_user);
+        }
+
        syslog(LOG_NOTICE, "login: %s %s%s plaintext%s %s", popd_clienthost,
               popd_userid, popd_subfolder ? popd_subfolder : "",
               popd_starttls_done ? "+TLS" : "", "User logged in");
