diff --git a/imap/mboxname.c b/imap/mboxname.c
index ec9c385..ff68ede 100644
--- a/imap/mboxname.c
+++ b/imap/mboxname.c
@@ -1191,7 +1191,9 @@ int mboxname_policycheck(const char *name)
 	else {
 	    if (!strchr(GOODCHARS, *name) &&
 		/* If we're using unixhierarchysep, DOTCHAR is allowed */
-		!(unixsep && *name == DOTCHAR))
+		!(unixsep && *name == DOTCHAR) ||
+		/* If we're not using virtdomains, '@' is not permitted in the mboxname */
+		(!config_virtdomains && *name == '@'))
 		return IMAP_MAILBOX_BADNAME;
 	    name++;
 	    sawutf7 = 0;
