diff --git a/saslauthd/auth_httpform.c b/saslauthd/auth_httpform.c
index 45fb083..240869b 100644
--- a/saslauthd/auth_httpform.c
+++ b/saslauthd/auth_httpform.c
@@ -92,6 +92,7 @@ static const char *r_port = "80";       /* remote port (mech_option) */
 static const char *r_uri = NULL;        /* URI to call (mech_option) */
 static const char *formdata = NULL;     /* HTML form data (mech_option) */
 static struct addrinfo *ai = NULL;      /* remote host, as looked up    */
+const int esc_amp = 0;
 /* END PRIVATE DEPENDENCIES */
 
 #define NETWORK_IO_TIMEOUT 30		/* network I/O timeout (seconds) */
@@ -180,7 +181,7 @@ static char *url_escape(
         if (!(in >= 'a' && in <= 'z') &&
             !(in >= 'A' && in <= 'Z') &&
             !(in >= '0' && in <= '9') &&
-            in != '&' && in != '=' && in != '-' && in != '_') {
+            (in != '&' && !esc_amp ) && in != '=' && in != '-' && in != '_') {
 
             /* encode it */
             if (outidx+3 > alloc) {
@@ -242,7 +243,13 @@ static char *create_post_data(
     int biggest;
     size_t i;
     /* END VARIABLES */
-    
+
+    esc_amp = 1;
+    user = url_escape(user);
+    password = url_escape(password);
+    realm = url_escape(realm);
+    esc_amp = 0;
+
     /* calculate memory needed for creating the complete query string. */
     ulen = strlen(user);
     plen = strlen(password);
@@ -471,7 +478,6 @@ auth_httpform (
     int s=-1;                           /* socket to remote auth host   */
     struct addrinfo *r;                 /* remote socket address info   */
     char *req;                          /* request, with user and pw    */
-    char *escreq;                       /* URL-escaped request          */
     char *c;                            /* scratch pointer              */
     int rc;                             /* return code scratch area     */
     char postbuf[RESP_LEN];             /* request buffer               */
@@ -535,14 +541,6 @@ auth_httpform (
         syslog(LOG_WARNING, "auth_httpform: create_post_data == NULL");
         return strdup(RESP_IERROR);
     }
-    escreq = url_escape(req);
-    if (escreq == NULL) {
-        memset(req, 0, strlen(req));
-        free(req); 
-        close(s);
-        syslog(LOG_WARNING, "auth_httpform: url_escape == NULL");
-        return strdup(RESP_IERROR);
-    }
 
     postlen = snprintf(postbuf, RESP_LEN-1,
               "POST %s HTTP/1.1" CRLF
@@ -552,11 +550,11 @@ auth_httpform (
               "Content-Type: application/x-www-form-urlencoded" CRLF
               "Content-Length: %d" TWO_CRLF
               "%s",
-              r_uri, r_host, r_port, strlen(escreq), escreq);
+              r_uri, r_host, r_port, strlen(req), req);
 
     if (flags & VERBOSE) {
         syslog(LOG_DEBUG, "auth_httpform: sending %s %s %s",
-               r_host, r_uri, escreq);
+               r_host, r_uri, req);
     }
     
     /* send it */
@@ -568,8 +566,6 @@ auth_httpform (
         syslog(LOG_WARNING, "auth_httpform: failed to send request");
         memset(req, 0, strlen(req));
         free(req); 
-        memset(escreq, 0, strlen(escreq));
-        free(escreq);
         memset(postbuf, 0, postlen);
         close(s);
         return strdup(RESP_IERROR);
@@ -578,8 +574,6 @@ auth_httpform (
     /* don't need these any longer */
     memset(req, 0, strlen(req));
     free(req); 
-    memset(escreq, 0, strlen(escreq));
-    free(escreq);
     memset(postbuf, 0, postlen);
 
     /* read and parse the response */
