--- pop3proxyd.c.orig	2003-07-22 16:17:17.000000000 -0300
+++ pop3proxyd.c	2004-01-31 17:19:27.000000000 -0200
@@ -1167,7 +1167,7 @@
 
     /* we now do the actual SASL exchange */
     r = sasl_client_start(backend_saslconn, 
-			  config_getstring(buf, "KERBEROS_V4"),
+			  config_getstring(buf, NULL),
 			  NULL, &out, &outlen, &mechusing);
     if ((r != SASL_OK) && (r != SASL_CONTINUE)) {
 	return r;
