*** mupdate.Oc	Wed Jan 31 08:10:05 2007
--- mupdate.c	Sun Mar  4 15:11:08 2007
***************
*** 2313,2319 ****
--- 2313,2322 ----
  	    remote_boxes.head = r->next;
  	} else if (ret < 0) {
  	    /* Local without corresponding remote, delete it */
+ 	  if (config_mupdate_config != IMAP_ENUM_MUPDATE_CONFIG_UNIFIED) {
+ 	    /* But not for a unified configuration */
  	    mboxlist_deletemailbox(l->mailbox, 1, "", NULL, 0, 0, 0);
+ 	  }
  	    local_boxes.head = l->next;
  	} else /* (ret > 0) */ {
  	    /* Remote without corresponding local, insert it */
***************
*** 2328,2334 ****
      if(l && !r) {
  	/* we have more deletes to do */
  	while(l) {
! 	    mboxlist_deletemailbox(l->mailbox, 1, "", NULL, 0, 0, 0);
  	    local_boxes.head = l->next;
  	    l = local_boxes.head;
  	}
--- 2331,2340 ----
      if(l && !r) {
  	/* we have more deletes to do */
  	while(l) {
! 	    if (config_mupdate_config != IMAP_ENUM_MUPDATE_CONFIG_UNIFIED) {
! 	      /* But not for a unified configuration */
! 	      mboxlist_deletemailbox(l->mailbox, 1, "", NULL, 0, 0, 0);
! 	    }
  	    local_boxes.head = l->next;
  	    l = local_boxes.head;
  	}
