Index: imap/tls.c
===================================================================
RCS file: /cvs/src/cyrus/imap/tls.c,v
retrieving revision 1.47.2.3
diff -u -r1.47.2.3 tls.c
--- imap/tls.c	28 Oct 2005 14:51:18 -0000	1.47.2.3
+++ imap/tls.c	27 Sep 2006 07:21:26 -0000
@@ -226,7 +226,7 @@
     int     err;
     int     depth;
 
-    syslog(LOG_ERR,"Doing a peer verify");
+    syslog(LOG_DEBUG,"Doing a peer verify");
 
     err_cert = X509_STORE_CTX_get_current_cert(ctx);
     err = X509_STORE_CTX_get_error(ctx);
