Index: perl/sieve/scripts/sieveshell.pl
===================================================================
RCS file: /cvs/src/cyrus/perl/sieve/scripts/sieveshell.pl,v
retrieving revision 1.24
diff -u -0 -r1.24 sieveshell.pl
--- perl/sieve/scripts/sieveshell.pl	30 Nov 2006 17:11:24 -0000	1.24
+++ perl/sieve/scripts/sieveshell.pl	14 Mar 2008 16:43:28 -0000
@@ -88 +88 @@
-if (! $ex eq "") {
+if ( $ex && $ex !~ m{/|\w+\.\w+} ) {
@@ -95,0 +96,2 @@
+} elsif ($ex){
+	open($filehandle, "$ex"); 
@@ -174 +176,3 @@
-my $term = Term::ReadLine->new("sieveshell");
+my $term;
+if ($interactive){
+	$term = Term::ReadLine->new("sieveshell");
@@ -176 +180,2 @@
-$term->ornaments(0);
+	$term->ornaments(0);
+}
@@ -180 +185 @@
-  $term->addhistory($_);
+  $term->addhistory($_) if $interactive;
