Index: sieveshell.pl
===================================================================
RCS file: /cvs/src/cyrus/perl/sieve/scripts/sieveshell.pl,v
retrieving revision 1.24
diff -u -r1.24 sieveshell.pl
--- sieveshell.pl	30 Nov 2006 17:11:24 -0000	1.24
+++ sieveshell.pl	8 Mar 2008 05:16:04 -0000
@@ -171,15 +171,20 @@
     die "unable to connect to server";
 }
 
-my $term = Term::ReadLine->new("sieveshell");
+my $term;
+if ($interactive){
+   $term = Term::ReadLine->new("sieveshell");
 
-$term->ornaments(0);
+   $term->ornaments(0);
+}
 
-while(defined($_  = ($interactive ? $term->readline('> ') : <$filehandle>))){
+   while(defined($_  = ($interactive ? $term->readline('> ') : <$filehandle>))){
   
-  $term->addhistory($_);
+      $term->addhistory($_) if $interactive;
 
-  my @words = split ' ',$_;
+   open(SCRIPT,"$_");
+   foreach my $line (<SCRIPT>){
+     my @words = split ' ',$line;
   my $str;
     if ($#words < 0) {
 	next;
@@ -274,6 +279,7 @@
     } else {
 	print "Invalid command: $words[0]\n";
     } 
+   }
 }
 
 __END__
