diff --git a/imap/idled.c b/imap/idled.c
index 8850ce1..03ea4c3 100644
--- a/imap/idled.c
+++ b/imap/idled.c
@@ -273,6 +273,10 @@ int main(int argc, char **argv)
     pid_t pid;
     char *alt_config = NULL;
 
+    if ((geteuid()) == 0 && (become_cyrus(/*is_master*/0) != 0)) {
+	fatal("must run as the Cyrus user", EC_USAGE);
+    }
+
     p = getenv("CYRUS_VERBOSE");
     if (p) verbose = atoi(p) + 1;
 
