Index: sync_client.c
===================================================================
RCS file: /data/ucvs/hermes/src/cyrus-imapd-2.4/imap/sync_client.c,v
retrieving revision 1.15
diff -u -d -r1.15 sync_client.c
--- sync_client.c	18 Jan 2012 15:23:03 -0000	1.15
+++ sync_client.c	3 Oct 2012 13:38:12 -0000
@@ -1307,6 +1307,9 @@
     r = mailbox_open_irl(local->name, &mailbox);
     if (r == IMAP_MAILBOX_NONEXISTENT) {
 	/* been deleted in the meanwhile... */
+        if (!remote)
+            goto done;  /* Promote and retry */
+
 	r = folder_delete(remote->name);
 	goto done;
     }
