--- /usr/lib/perl5/Cyrus/IMAP.pm.orig   2011-10-04 21:53:03.000000000 +0200
+++ /usr/lib/perl5/Cyrus/IMAP.pm        2014-01-27 15:53:44.726444230 +0100
@@ -202,8 +202,8 @@
   if (defined $first &&
       $first =~ /^-\w+|Mechanism|Service|Authz|User|Minssf|Maxssf|Password|Tlskey|Notls|CAfile|CApath$/) {
     (undef, %opts) = @_;
-    foreach (qw(mechanism service authz user minssf maxssf password tlskey notls)) {
-      $opts{'-' . $_} = $opts{ucfirst($_)} if !defined($opts{'-' . $_});
+    foreach (qw(Mechanism Service Authz User Minssf Maxssf Password Tlskey Notls CAfile CApath)) {
+      $opts{'-' . lc $_} = $opts{$_} if !defined($opts{'-' . lc $_});
     }
   } else {
     (undef, $opts{-mechanism}, $opts{-service}, $opts{-authz}, $opts{-user},
@@ -283,7 +283,7 @@
   }
 
   if (!$rc && $logindisabled) {
-    $self->_starttls('', '', '', '') || return undef;
+    $self->_starttls('', '', $opts{-cafile}, $opts{-capath}) || return undef;
 
     # Refetch all relevent capabilities
     ($starttls, $logindisabled, $availmechs) = (0, 0, "");

