diff --git a/imap/annotate.c b/imap/annotate.c
index ff53735..1c7473c 100644
--- a/imap/annotate.c
+++ b/imap/annotate.c
@@ -3067,9 +3067,23 @@ int annotate_msg_copy(const char *oldmboxname, uint32_t olduid,
 		      const char *newmboxname, uint32_t newuid,
 		      const char *userid)
 {
-    return _annotate_rewrite(oldmboxname, olduid, userid,
-			     newmboxname, newuid, userid,
-			     /*copy*/1);
+    int r;
+
+    r = annotatemore_begin();
+    if (r)
+	goto out;
+
+    r = _annotate_rewrite(oldmboxname, olduid, userid,
+			  newmboxname, newuid, userid,
+			  /*copy*/1);
+    if (r)
+	goto out;
+    r = annotatemore_commit();
+
+out:
+    if (r)
+	annotatemore_abort();
+    return r;
 }
 
 /*************************  Annotation Initialization  ************************/
