diff -Naur cyrus-imapd-2.2.12/imap/global.c cyrus-imapd-2.2.12.patched/imap/global.c
--- cyrus-imapd-2.2.12/imap/global.c	2004-11-23 18:40:15.000000000 +0100
+++ cyrus-imapd-2.2.12.patched/imap/global.c	2006-10-24 18:38:59.000000000 +0200
@@ -120,7 +120,7 @@
     
     /* xxx we lose here since we can't have the prefix until we load the
      * config file */
-    openlog(config_ident, LOG_PID, SYSLOG_FACILITY);
+    openlog(config_ident, LOG_PID | LOG_PERROR, SYSLOG_FACILITY);
 
     /* Load configuration file.  This will set config_dir when it finds it */
     config_read(alt_config);
@@ -137,7 +137,7 @@
 	strlcat(ident_buf, ident, size);
 
 	closelog();
-	openlog(ident_buf, LOG_PID, SYSLOG_FACILITY);
+	openlog(ident_buf, LOG_PID | LOG_PERROR, SYSLOG_FACILITY);
 
 	/* don't free the openlog() string! */
     }
diff -Naur cyrus-imapd-2.2.12/lib/util.c cyrus-imapd-2.2.12.patched/lib/util.c
--- cyrus-imapd-2.2.12/lib/util.c	2004-09-09 18:22:01.000000000 +0200
+++ cyrus-imapd-2.2.12.patched/lib/util.c	2006-11-05 22:58:02.000000000 +0100
@@ -301,11 +301,7 @@
     shutdown(1, SHUT_RD);
     dup2(devnull, 1);
 
-    /* stderr */
-    shutdown(2, SHUT_RD);
-    dup2(devnull, 2);
-
-    if (devnull > 2) close(devnull);
+    if (devnull > 1) close(devnull);
 }
 
 
diff -Naur cyrus-imapd-2.2.12/master/masterconf.c cyrus-imapd-2.2.12.patched/master/masterconf.c
--- cyrus-imapd-2.2.12/master/masterconf.c	2003-11-25 18:37:19.000000000 +0100
+++ cyrus-imapd-2.2.12.patched/master/masterconf.c	2006-10-24 18:40:13.000000000 +0200
@@ -76,7 +76,7 @@
 
     /* Open the log file with the appropriate facility so we 
      * correctly log any config errors */
-    openlog(ident, LOG_PID, SYSLOG_FACILITY);
+    openlog(ident, LOG_PID | LOG_PERROR, SYSLOG_FACILITY);
 
     config_ident = ident;
     config_read(alt_config);
@@ -92,7 +92,7 @@
 
 	/* Reopen the log with the new prefix */
 	closelog();
-	openlog(buf, LOG_PID, SYSLOG_FACILITY);
+	openlog(buf, LOG_PID | LOG_PERROR, SYSLOG_FACILITY);
 
         /* don't free the openlog() string! */
     }
diff -Naur cyrus-imapd-2.2.12/master/service.c cyrus-imapd-2.2.12.patched/master/service.c
--- cyrus-imapd-2.2.12/master/service.c	2004-12-17 17:32:25.000000000 +0100
+++ cyrus-imapd-2.2.12.patched/master/service.c	2006-11-05 22:42:29.000000000 +0100
@@ -515,10 +515,6 @@
 	    syslog(LOG_ERR, "can't duplicate accepted socket: %m");
 	    service_abort(EX_OSERR);
 	}
-	if (fd != 2 && dup2(fd, 2) < 0) {
-	    syslog(LOG_ERR, "can't duplicate accepted socket: %m");
-	    service_abort(EX_OSERR);
-	}
 
 	/* tcp only */
 	if(soctype == SOCK_STREAM) {
diff -Naur cyrus-imapd-2.2.12/ptclient/ptexpire.c cyrus-imapd-2.2.12.patched/ptclient/ptexpire.c
--- cyrus-imapd-2.2.12/ptclient/ptexpire.c	2004-05-26 18:05:59.000000000 +0200
+++ cyrus-imapd-2.2.12.patched/ptclient/ptexpire.c	2006-10-24 18:42:06.000000000 +0200
@@ -114,7 +114,7 @@
 
     if (geteuid() == 0) fatal("must run as the Cyrus user", EC_USAGE);
     
-    openlog("ptexpire", LOG_PID, SYSLOG_FACILITY);
+    openlog("ptexpire", LOG_PID | LOG_PERROR, SYSLOG_FACILITY);
 
     while ((opt = getopt(argc, argv, "C:E:")) != EOF) {
 	switch (opt) {
diff -Naur cyrus-imapd-2.2.12/syslog/syslog.c cyrus-imapd-2.2.12.patched/syslog/syslog.c
--- cyrus-imapd-2.2.12/syslog/syslog.c	2003-06-04 05:27:56.000000000 +0200
+++ cyrus-imapd-2.2.12.patched/syslog/syslog.c	2006-10-24 18:41:11.000000000 +0200
@@ -158,7 +158,7 @@
 
         /* Get connected. */
         if (!connected)
-                openlog(LogTag, LogStat | LOG_NDELAY, 0);
+                openlog(LogTag, LogStat | LOG_NDELAY | LOG_PERROR, 0);
 
         /* Build the message. */
         (void)time(&now);
