--- mailbox.c	7 Oct 2008 19:15:38 -0000	1.182
+++ mailbox.c	14 Oct 2008 17:38:22 -0000
@@ -2874,9 +2874,11 @@
 
     strlcpy(nbuf, mailbox->name, sizeof(nbuf));
     ntail = nbuf + strlen(nbuf);
+
+    syslog(LOG_NOTICE, "Deleted mailbox %s", mailbox->name);
+    if (updatenotifier) updatenotifier(mailbox->name);
     mailbox_close(mailbox);
 
-    syslog(LOG_NOTICE, "Deleted mailbox %s", nbuf);
 
     do {
 	/* Check if the mailbox has children */
@@ -2918,10 +2920,6 @@
 	}
 	r = mboxlist_lookup(nbuf, NULL, NULL);
     } while(r == IMAP_MAILBOX_NONEXISTENT);
-
-    *ntail = '\0';
-
-    if (updatenotifier) updatenotifier(nbuf);
 
     return 0;
 }
