--- imap/mboxlist.c.orig	2011-10-04 14:53:03.000000000 -0500
+++ imap/mboxlist.c	2012-02-27 07:31:23.638307615 -0600
@@ -997,6 +997,7 @@
     char *mboxent = NULL;
     char *p;
     mupdate_handle *mupdate_h = NULL;
+    int len1,len2;
 
     /* 1. open mailbox */
     r = mailbox_open_iwl(oldname, &oldmailbox);
@@ -1035,8 +1036,9 @@
 	    goto done;
 	}
     } else if ((p = mboxname_isusermailbox(oldname, 1))) {
-	if (!strncmp(p, userid, config_virtdomains ? strcspn(userid, "@") :
-		     strlen(userid))) {
+        len1 = strlen(p);
+        len2 = config_virtdomains ? strcspn(userid, "@") : strlen(userid);
+        if (len1==len2 && !strncmp(p, userid, len1)) {
 	    /* Special case of renaming inbox */
 	    if (!(myrights & ACL_DELETEMBOX)) {
 		r = IMAP_PERMISSION_DENIED;
