--- imap/imapd.c.orig	2011-10-04 12:53:03.000000000 -0700
+++ imap/imapd.c	2011-12-14 10:32:54.000000000 -0800
@@ -2819,7 +2819,7 @@
 	if (capa_is_disabled(base_capabilities[i].str))
 	    continue;
 	/* Don't show "MAILBOX-REFERRALS" if disabled by config */
-	if (disable_referrals && 
+	if (config_getswitch(IMAPOPT_PROXYD_DISABLE_MAILBOX_REFERRALS) && 
 	    !strcmp(base_capabilities[i].str, "MAILBOX-REFERRALS"))
 	    continue;
 	/* Don't show if they're not shown at this level of login */
@@ -5979,7 +5979,11 @@
     clock_t start = clock();
     char mytime[100];
 
-    if (listargs->sel & LIST_SEL_REMOTE) supports_referrals = !disable_referrals;
+    if (listargs->sel & LIST_SEL_REMOTE) {
+	if (!config_getswitch(IMAPOPT_PROXYD_DISABLE_MAILBOX_REFERRALS)) {
+	    supports_referrals = !disable_referrals;
+	}
+    }
 
     list_callback_calls = 0;
 
