From b1f33fd10599103840c33d6577fe429f3199773b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=D8yvind=20Kolbu?= <oyvind.kolbu@usit.uio.no>
Date: Wed, 6 Apr 2011 13:17:20 +0200
Subject: [PATCH 8/8] Accept both quoted and non-quoted charset.

---
 imap/imapd.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/imap/imapd.c b/imap/imapd.c
index 6e85604..d8b8546 100644
--- a/imap/imapd.c
+++ b/imap/imapd.c
@@ -4738,7 +4738,7 @@ void cmd_thread(char *tag, int usinguid)
     }
 
     /* get charset */
-    c = getstring(imapd_in, imapd_out, &arg);
+    c = getastring(imapd_in, imapd_out, &arg);
     if (c != ' ') {
 	prot_printf(imapd_out, "%s BAD Missing charset in Thread\r\n",
 		    tag);
-- 
1.7.1

