Index: lmtpd.c
===================================================================
RCS file: /afs/andrew/system/cvs/src/cyrus/imap/lmtpd.c,v
retrieving revision 1.96
diff -u -c -r1.96 lmtpd.c
cvs server: conflicting specifications of output style
*** lmtpd.c	2002/04/15 14:42:17	1.96
--- lmtpd.c	2002/04/23 20:24:44
***************
*** 156,162 ****
  extern int optind;
  extern char *optarg;
  extern int errno;
! static int dupelim = 0;		/* eliminate duplicate messages with
  				   same message-id */
  static int singleinstance = 1;	/* attempt single instance store */
  const char *BB = "";
--- 156,163 ----
  extern int optind;
  extern char *optarg;
  extern int errno;
! static int have_dupdb = 1;	/* duplicate delivery db is initialized */
! static int dupelim = 1;		/* eliminate duplicate messages with
  				   same message-id */
  static int singleinstance = 1;	/* attempt single instance store */
  const char *BB = "";
***************
*** 271,283 ****
      }
  
      dupelim = config_getswitch("duplicatesuppression", 1);
!     if (dupelim) {
! 	/* initialize duplicate delivery database */
! 	if (duplicate_init(NULL, 0) != 0) {
! 	    syslog(LOG_ERR, 
! 		   "lmtpd: unable to init duplicate delivery database\n");
! 	    dupelim = 0;
! 	}
      }
  
      /* so we can do mboxlist operations */
--- 272,282 ----
      }
  
      dupelim = config_getswitch("duplicatesuppression", 1);
!     /* initialize duplicate delivery database */
!     if (duplicate_init(NULL, 0) != 0) {
! 	syslog(LOG_ERR, 
! 	       "lmtpd: unable to init duplicate delivery database\n");
! 	dupelim = have_dupdb = 0;
      }
  
      /* so we can do mboxlist operations */
***************
*** 1061,1068 ****
  	return NULL;
      }
      
!     if (!dupelim) {
! 	/* duplicate delivery suppression is needed for sieve */
  	return NULL;
      }
  
--- 1060,1067 ----
  	return NULL;
      }
      
!     if (!have_dupdb) {
! 	/* duplicate delivery database is needed for sieve */
  	return NULL;
      }
  
***************
*** 1236,1254 ****
  		snprintf(namebuf, sizeof(namebuf), "%s+%s", rcpt,
  			 plus ? plus : "");
  		
- 		/* is this the first time we've sieved the message? */
- 		if (msgdata->id) {
- 		    char *sdb = make_sieve_db(namebuf);
- 		    
- 		    if (duplicate_check(msgdata->id, strlen(msgdata->id),
- 					sdb, strlen(sdb))) {
- 			logdupelem(msgdata->id, sdb);
- 			/* done it before ! */
- 			r = 0;
- 			goto donercpt;
- 		    }
- 		}
- 		
  		r = sieve_script_parse(sieve_interp, f, (void *) sdata, &s);
  		fclose(f);
  		if (r == SIEVE_OK) {
--- 1235,1240 ----
***************
*** 1307,1313 ****
  	    }
  	}
  
-     donercpt:
  	free(rcpt);
  	msg_setrcpt_status(msgdata, n, r);
      }
--- 1293,1298 ----
