diff -C 3 -r ../cyrus-imapd-2.2.12.clean/imap/imapd.c ./imap/imapd.c
*** ../cyrus-imapd-2.2.12.clean/imap/imapd.c	2005-02-14 09:39:55 +0300
--- ./imap/imapd.c	2005-10-10 19:42:03 +0400
***************
*** 1692,1697 ****
--- 1692,1708 ----
      }
  }
  
+ static void userid_tointernal(struct namespace *namespace, char *userid)
+ {
+     char hiersep_orig = namespace->hier_sep;
+ 
+     namespace->hier_sep='/';
+     mboxname_hiersep_tointernal(namespace, userid,
+ 				config_virtdomains ?
+ 				strcspn(userid, "@") : 0);
+     namespace->hier_sep=hiersep_orig;
+ }
+ 
  /*
   * Perform a LOGIN command
   */
***************
*** 1843,1851 ****
      }
  
      /* Translate any separators in userid */
!     mboxname_hiersep_tointernal(&imapd_namespace, imapd_userid,
! 				config_virtdomains ?
! 				strcspn(imapd_userid, "@") : 0);
  
      freebuf(&passwdbuf);
      return;
--- 1854,1860 ----
      }
  
      /* Translate any separators in userid */
!     userid_tointernal(&imapd_namespace, imapd_userid);
  
      freebuf(&passwdbuf);
      return;
***************
*** 1989,1997 ****
      }
  
      /* Translate any separators in userid */
!     mboxname_hiersep_tointernal(&imapd_namespace, imapd_userid,
! 				config_virtdomains ?
! 				strcspn(imapd_userid, "@") : 0);
  
      return;
  }
--- 1998,2004 ----
      }
  
      /* Translate any separators in userid */
!     userid_tointernal(&imapd_namespace, imapd_userid);
  
      return;
  }
