diff --git a/imap/cyr_expire.c b/imap/cyr_expire.c
index 6091461..c3f4f7d 100644
--- a/imap/cyr_expire.c
+++ b/imap/cyr_expire.c
@@ -81,7 +81,7 @@ static int verbose = 0;
 void usage(void)
 {
     fprintf(stderr,
-	    "cyr_expire [-C <altconfig>] -E <days> [-X <expunge-days>] [-p prefix] [-a] [-v]\n");
+	    "cyr_expire [-C <altconfig>] [-E <expire-duration>] [-D <delete-days] [-X <expunge-days>] [-p prefix] [-a] [-v]\n");
     exit(-1);
 }
 
@@ -391,7 +391,7 @@ int main(int argc, char *argv[])
 	}
     }
 
-    if (!expire_seconds) usage();
+    if (!expire_seconds && !delete_seconds && !expunge_seconds) usage();
 
     sigemptyset(&action.sa_mask);
     action.sa_flags = 0;
diff --git a/man/cyr_expire.8 b/man/cyr_expire.8
index 7708136..8bd3934 100644
--- a/man/cyr_expire.8
+++ b/man/cyr_expire.8
@@ -52,7 +52,9 @@ cyr_expire \- expire messages and duplicate delivery database entries
 [
 .BI \-D " delete-days"
 ]
+[
 .BI \-E " expire-duration"
+]
 [
 .BI \-X " expunge-days"
 ]
@@ -101,6 +103,8 @@ expired using the value given to the \fB-E\fR option.
 .PP
 \fBCyr_expire\fR reads its configuration options out of the \fIimapd.conf\fR(5)
 file unless specified otherwise by \fB-C\fR.
+.PP
+\fBCyr_expire\fR requires at least one of \fB-D -E\fR or \fB-X\fR to be supplied.
 .SH OPTIONS
 .TP
 \fB\-C\fI config-file\fR
