--- compile_et.apple	2005-04-16 05:29:05.000000000 +1000
+++ compile_et.fixed	2012-06-25 12:18:05.370038003 +1000
@@ -23,6 +23,7 @@
 
 use strict;
 use integer;
+use Config;
 
 #
 # Tables and constants used to calculate the base error code from the name
@@ -152,6 +153,8 @@
         $etBase += $charToNumber{ substr($etName, $i, 1) };
     }
     $etBase <<= $errorCodeRange;
+    # Sign-extend etBase on 64 platforms
+    $etBase |= (~0 << 32) if $Config{longsize} == 8 && $etBase & 0x80000000;
     printf "Error Table \"%s\" has base $etBase\n", $etName, $etBase;
 }
 
