diff -ur cyrus-imapd-2.4.11.orig/imap/Makefile.in cyrus-imapd-2.4.11/imap/Makefile.in
--- cyrus-imapd-2.4.11.orig/imap/Makefile.in	2011-09-08 18:57:42.000000000 +0000
+++ cyrus-imapd-2.4.11/imap/Makefile.in	2011-10-04 11:48:35.294360462 +0000
@@ -347,12 +347,6 @@
 	$(CC) $(LDFLAGS) -o $@ unexpunge.o $(CLIOBJS) \
 	libimap.a $(DEPLIBS) $(LIBS)
 
-make_md5: make_md5.o libimap.a mutex_fake.o $(DEPLIBS)
-	$(CC) $(LDFLAGS) -o make_md5 make_md5.o libimap.a mutex_fake.o $(DEPLIBS) $(LIBS)
-
-make_sha1: make_sha1.o libimap.a mutex_fake.o $(DEPLIBS)
-	$(CC) $(LDFLAGS) -o make_sha1 make_sha1.o libimap.a mutex_fake.o $(DEPLIBS) $(LIBS)
-
 sync_client: sync_client.o sync_support.o \
 	backend.o tls.o imapparse.o libimap.a mutex_fake.o $(DEPLIBS)
 	$(CC) $(LDFLAGS) -o \
Only in cyrus-imapd-2.4.11.orig/imap: signals.c
Only in cyrus-imapd-2.4.11.orig/imap: signals.h
--- cyrus-imapd-2.4.11.orig/lib/Makefile.in	2011-09-08 18:57:42.000000000 +0000
+++ cyrus-imapd-2.4.11/lib/Makefile.in	2011-10-04 11:50:40.311360522 +0000
@@ -102,9 +102,9 @@
 LIBCYRM_HDRS = $(srcdir)/hash.h $(srcdir)/mpool.h $(srcdir)/xmalloc.h \
 	$(srcdir)/xstrlcat.h $(srcdir)/xstrlcpy.h $(srcdir)/util.h \
 	$(srcdir)/strhash.h $(srcdir)/libconfig.h $(srcdir)/assert.h \
-	imapopts.h $(srcdir)/crc32.h
+	imapopts.h
 LIBCYRM_OBJS = libconfig.o imapopts.o hash.o mpool.o xmalloc.o strhash.o \
-	xstrlcat.o xstrlcpy.o assert.o util.o signals.o crc32.o @IPV6_OBJS@
+	xstrlcat.o xstrlcpy.o assert.o util.o signals.o @IPV6_OBJS@
 
 all: $(BUILTSOURCES) libcyrus_min.a libcyrus.a
 
diff -ur cyrus-imapd-2.4.11.orig/sieve/addr.y cyrus-imapd-2.4.11/sieve/addr.y
--- cyrus-imapd-2.4.11.orig/sieve/addr.y	2011-09-08 18:57:42.000000000 +0000
+++ cyrus-imapd-2.4.11/sieve/addr.y	2011-10-04 11:52:58.112360579 +0000
@@ -70,7 +70,7 @@
 %}
 
 %token ATOM QTEXT DTEXT
-
+%name-prefix "addr"
 %%
 sieve_address: addrspec			/* simple address */
 	| phrase '<' addrspec '>'	/* name & addr-spec */
--- cyrus-imapd-2.4.11.orig/sieve/Makefile.in	2011-09-08 18:57:42.000000000 +0000
+++ cyrus-imapd-2.4.11/sieve/Makefile.in	2011-10-04 11:53:27.026360644 +0000
@@ -106,7 +106,7 @@
 #	$(LEX) -t -Paddr $(srcdir)/addr-lex.l > $@
 
 addr.c addr.h: addr.y
-	$(YACC) $(YFLAGS) -p addr $(srcdir)/addr.y
+	$(YACC) $(YFLAGS) $(srcdir)/addr.y
 	mv -f y.tab.c addr.c
 	mv -f y.tab.h addr.h
 
