diff -ur cyrus-sasl-2.1.7.orig/java/CyrusSasl/GenericClient.java cyrus-sasl-2.1.6/java/CyrusSasl/GenericClient.java
--- cyrus-sasl-2.1.7.orig/java/CyrusSasl/GenericClient.java	2001-02-12 14:53:10.000000000 -0600
+++ cyrus-sasl-2.1.7/java/CyrusSasl/GenericClient.java	2002-07-17 12:04:38.000000000 -0500
@@ -112,7 +112,7 @@
 	}
 
 	if (wantpass==1) {
-	    pc = new PasswordCallback("Please enter your password");
+	    pc = new PasswordCallback("Please enter your password", false);
 	    cbs[pos] = pc;
 	    pos++;
 	}
@@ -124,7 +124,7 @@
 	}
 	
 	try {
-	    cbh.invokeCallback(cbs);
+	    cbh.handle(cbs);
 	} catch (UnsupportedCallbackException e) {
 	    throw new SaslException("Unsupported callback",null);
 	} catch (IOException e) {
@@ -138,7 +138,7 @@
 	    this.authid = nc2.getName();
 	}
 	if (pc!=null) {
-	    this.password = pc.getPassword();
+	    this.password = new String(pc.getPassword());
 	}
 	if (rc!=null) {
 	    this.realm = rc.getRealm();
