diff -Nrub cyrus-imapd-2.3.13/imap/tls.c working_copy/imap/tls.c
--- cyrus-imapd-2.3.13/imap/tls.c	2008-04-15 18:58:08.000000000 +0100
+++ working_copy/imap/tls.c	2008-11-04 16:50:38.000000000 +0000
@@ -868,7 +868,7 @@
 	tv.tv_sec = timeout;
 	tv.tv_usec = 0;
 
-	sts = select(1, &rfds, NULL, NULL, &tv);
+	sts = select(readfd+1, &rfds, NULL, NULL, &tv);
 	if (sts <= 0) {
 	    if (sts == 0) {
 		syslog(LOG_DEBUG, "SSL_accept() timed out -> fail");
