--- cyrus-imapd-2.3.7/imap/lmtpd.c	2006-05-23 15:09:36.000000000 +0200
+++ cyrus-imapd-2.3.7.ok/imap/lmtpd.c	2007-03-15 16:28:50.390913880 +0100
@@ -419,9 +419,11 @@
 	       that we may have to restart this transaction from scratch */
 	    fatal("error communicating with MUPDATE server", EC_TEMPFAIL);
 	}
-
-	type |= MBTYPE_REMOTE;
-	if (server) *server = (char *) mailboxdata->server;
+	else {
+	  type |= MBTYPE_REMOTE;
+	  if (server) *server = (char *) mailboxdata->server;
+	  if (aclp) *aclp = (char *) mailboxdata->acl;
+        }
     }
     else {
 	/* do a local lookup and kick the slave if necessary */
