--- cyrus-imapd-2.2.6-vanilla/lib/imclient.c	2003-10-22 20:50:12.000000000 +0200
+++ cyrus-imapd-2.2.6/lib/imclient.c	2004-08-07 16:14:48.282143000 +0200
@@ -235,7 +235,9 @@
     memset(&hints, 0, sizeof(hints));
     hints.ai_family = PF_UNSPEC;
     hints.ai_socktype = SOCK_STREAM;
+#ifndef GETADDR_NOFLAGS
     hints.ai_flags = AI_CANONNAME;
+#endif
     if (getaddrinfo(host, port, &hints, &res0))
 	return -1;
     for (res = res0; res; res = res->ai_next) {
