diff -Nur cyrus-imapd-2.3.8.dist/imap/imapd.c cyrus-imapd-2.3.8/imap/imapd.c
--- cyrus-imapd-2.3.8.dist/imap/imapd.c	2007-02-05 19:49:55.000000000 +0100
+++ cyrus-imapd-2.3.8/imap/imapd.c	2007-03-27 10:21:43.000000000 +0200
@@ -527,7 +527,7 @@
     backend_cached = NULL;
     backend_inbox = backend_current = NULL;
     proxy_cmdcnt = 0;
-    disable_referrals = 0;
+    disable_referrals = config_getswitch(IMAPOPT_PROXYD_DISABLE_MAILBOX_REFERRALS);
     supports_referrals = 0;
 
     if (imapd_mailbox) {
@@ -2592,6 +2592,10 @@
 
     prot_printf(imapd_out, CAPA_POSTAUTH_STRING);
 
+	if (! disable_referrals) {
+		(imapd_out, " MAILBOX-REFERRALS");
+	}
+
     if (idle_enabled()) {
 	prot_printf(imapd_out, " IDLE");
     }
diff -Nur cyrus-imapd-2.3.8.dist/imap/version.h cyrus-imapd-2.3.8/imap/version.h
--- cyrus-imapd-2.3.8.dist/imap/version.h	2007-02-06 19:07:42.000000000 +0100
+++ cyrus-imapd-2.3.8/imap/version.h	2007-03-27 09:24:19.000000000 +0200
@@ -63,7 +63,7 @@
 #define CAPA_PREAUTH_STRING "IMAP4 IMAP4rev1 LITERAL+ ID"
 
 #define CAPA_POSTAUTH_STRING " ACL RIGHTS=kxte QUOTA " \
-	"MAILBOX-REFERRALS NAMESPACE UIDPLUS " \
+	"NAMESPACE UIDPLUS " \
 	"NO_ATOMIC_RENAME UNSELECT " \
 	"CHILDREN MULTIAPPEND BINARY " \
 	"SORT SORT=MODSEQ THREAD=ORDEREDSUBJECT THREAD=REFERENCES " \
diff -Nur cyrus-imapd-2.3.8.dist/lib/imapoptions cyrus-imapd-2.3.8/lib/imapoptions
--- cyrus-imapd-2.3.8.dist/lib/imapoptions	2007-02-07 19:58:07.000000000 +0100
+++ cyrus-imapd-2.3.8/lib/imapoptions	2007-03-27 10:24:13.000000000 +0200
@@ -748,6 +748,10 @@
    connections that these referrals would cause, thus resulting in a higher
    authentication load on the respective backend server. */
 
+{ "proxyd_disable_mailbox_referrals", 0, SWITCH }
+/* Set to true to disable the use of mailbox-referrals on the
+   proxy servers. */
+
 { "proxyservers", NULL, STRING }
 /* A list of users and groups that are allowed to proxy for other
    users, separated by spaces.  Any user listed in this will be
