--- config.h.in	2006-05-18 15:38:02.000000000 -0400
+++ config.h.in	2008-10-09 02:18:14.523335000 -0400
@@ -101,6 +101,9 @@
 /* Do we have a getnameinfo() function? */
 #undef HAVE_GETNAMEINFO
 
+/* Define to 1 if you have the `getpassphrase' function. */
+#undef HAVE_GETPASSPHRASE
+
 /* Define to 1 if you have the `getpwnam' function. */
 #undef HAVE_GETPWNAM
 
--- configure.in	2006-05-18 15:06:13.000000000 -0400
+++ configure.in	2008-10-09 01:52:11.306245000 -0400
@@ -1115,7 +1115,7 @@
 
 #AC_FUNC_MEMCMP
 #AC_FUNC_VPRINTF
-AC_CHECK_FUNCS(gethostname getdomainname getpwnam getspnam gettimeofday inet_aton memcpy mkdir select socket strchr strdup strerror strspn strstr strtol jrand48)
+AC_CHECK_FUNCS(gethostname getdomainname getpwnam getspnam gettimeofday inet_aton memcpy mkdir select socket strchr strdup strerror strspn strstr strtol jrand48 getpassphrase)
 
 if test $enable_cmulocal = yes; then
     AC_WARN([enabling CMU local kludges])
