--- cyrus-imapd/imap/sync_reset.c-DIST	2011-08-10 09:43:41.000000000 +0100
+++ cyrus-imapd/imap/sync_reset.c	2011-08-10 09:46:50.000000000 +0100
@@ -144,9 +144,7 @@
 
     /* Nuke subscriptions */
     list = sync_name_list_create();
-    r = (sync_namespacep->mboxlist_findsub)(sync_namespacep, "*", 1,
-					    (char *)userid, sync_authstate,
-					    addmbox_sub, (void *)list, 1);
+    r = mboxlist_allsubs(userid, addmbox_sub, list);
     if (r) goto fail;
 
     /* ignore failures here - the subs file gets deleted soon anyway */
