diff -Nur cyrus-imapd-2.3.8/imap/lmtpd.c cyrus-imapd-2.3.8.tuc/imap/lmtpd.c
--- cyrus-imapd-2.3.8/imap/lmtpd.c	2007-02-05 19:41:47.000000000 +0100
+++ cyrus-imapd-2.3.8.tuc/imap/lmtpd.c	2007-07-18 13:53:22.000000000 +0200
@@ -485,8 +485,8 @@
 
     r = append_setup(&as, mailboxname, MAILBOX_FORMAT_NORMAL,
 		     authuser, authstate, acloverride ? 0 : ACL_POST, 
-		     quotaoverride ? -1 :
-		     config_getswitch(IMAPOPT_LMTP_STRICT_QUOTA) ? size : 0);
+		     quotaoverride ? (long)-1 :
+		     config_getswitch(IMAPOPT_LMTP_STRICT_QUOTA) ? (long)size : 0);
 
     if (!r && !content->body) {
 	/* parse the message body if we haven't already,
