--- ../cyrus-imapd-2.2.6.kolab-orig/lib/auth_unix.c	2004-10-09 00:07:45.000000000 +0200
+++ lib/auth_unix.c	2004-10-09 00:29:58.000000000 +0200
@@ -274,7 +274,7 @@ static struct group* fgetgrnam(const cha
     if (!groupfile) groupfile = fopen("/etc/group", "r");
     if (groupfile) {
        while ((grp = fgetgrent(groupfile))) {
-         if (strcmp(grp->gr_name, name) == 0) {
+         if (strcasecmp(grp->gr_name, name) == 0) {
            fclose(groupfile);
            return grp;
          }
@@ -386,7 +386,7 @@ struct auth_state *auth_newstate(const c
     if (groupfile) {
        while ((grp = fgetgrent(groupfile))) {
          for (mem = grp->gr_mem; *mem; mem++) {
-            if (!strcmp(*mem, identifier)) break;
+            if (!strcasecmp(*mem, identifier)) break;
          }
 
          if (*mem || (pwd && pwd->pw_gid == grp->gr_gid)) {
