Index: imap/unexpunge.c
===================================================================
RCS file: /afs/andrew/system/cvs/src/cyrus/imap/unexpunge.c,v
retrieving revision 1.12
diff -u -r1.12 unexpunge.c
--- imap/unexpunge.c	24 Mar 2008 17:09:20 -0000	1.12
+++ imap/unexpunge.c	24 Nov 2008 04:46:57 -0000
@@ -87,7 +87,7 @@
 void usage(void)
 {
     fprintf(stderr,
-	    "unexpunge [-C <altconfig>] -l <mailbox>\n"
+	    "unexpunge [-C <altconfig>] -l [-H] <mailbox>\n"
             "unexpunge [-C <altconfig>] -t time-interval [ -d ] [ -v ] mailbox\n"
 	    "unexpunge [-C <altconfig>] -a [-d] [-v] <mailbox>\n"
 	    "unexpunge [-C <altconfig>] -u [-d] [-v] <mailbox> <uid>...\n");
Index: man/unexpunge.8
===================================================================
RCS file: /afs/andrew/system/cvs/src/cyrus/man/unexpunge.8,v
retrieving revision 1.4
diff -u -r1.4 unexpunge.8
--- man/unexpunge.8	4 Apr 2008 12:47:06 -0000	1.4
+++ man/unexpunge.8	24 Nov 2008 04:46:57 -0000
@@ -50,6 +50,9 @@
 .I config-file
 ]
 .B \-l
+[
+.B \-H
+]
 .I mailbox
 .br
 .B unexpunge
@@ -111,6 +114,9 @@
 .B \-l
 List the expunged messages in the specified mailbox which are available
 for restoration.
+.TP
+.B \-H
+List the messages above in human readable format.
 .TP
 .BI \-t " time-interval"
 Unexpunge messages which were expunged in the last time-interval seconds.
