--- cyrus-imapd-2.5.7/imap/mbdump.c.orig	2016-01-18 21:10:50.678324750 +0100
+++ cyrus-imapd-2.5.7/imap/mbdump.c	2016-04-09 01:46:09.731883027 +0200
@@ -825,6 +825,7 @@
     const char *sieve_path = NULL;
     int sieve_usehomedir = config_getswitch(IMAPOPT_SIEVEUSEHOMEDIR);
     const char *userid = NULL;
+    int first_annotation = 1;
     char *annotation = NULL;
     struct buf content = BUF_INITIALIZER;
     char *seen_file = NULL;
@@ -897,8 +898,6 @@
     mailbox_get_usage(mailbox, old_quota_usage);
 
     astate = annotate_state_new();
-    r = annotate_state_set_mailbox(astate, mailbox);
-    if (r) goto done;
 
     while(1) {
 	char fnamebuf[MAX_MAILBOX_PATH + 1024];
@@ -922,6 +921,16 @@
 	    int i;
 	    char *tmpuserid;
 
+	    /* is this the fist annotation? re-read cyrus.header/index */
+	    if (first_annotation) {
+		first_annotation = 0;
+		mailbox_close(&mailbox);
+                r = mailbox_open_exclusive(mbname, &mailbox);
+		if (r) goto done;
+                r = annotate_state_set_mailbox(astate, mailbox);
+                if (r) goto done;
+	    }
+
 	    for(i=2; file.s[i]; i++) {
 		if(file.s[i] == '/') break;
 	    }
