diff -urN cyrus-imapd-2.4.11.orig/imap/imapd.c cyrus-imapd-2.4.11/imap/imapd.c
--- cyrus-imapd-2.4.11.orig/imap/imapd.c	2011-09-08 20:57:42.000000000 +0200
+++ cyrus-imapd-2.4.11/imap/imapd.c	2011-09-24 23:52:55.795700411 +0200
@@ -6634,11 +6634,14 @@
 	    c = getword(imapd_in, &arg);
 	    if (c != ' ' && c != ')') goto badlist;
 	    if (arg.s[0] == '\0') goto badlist;
-	    newquota = 0;
-	    for (p = arg.s; *p; p++) {
-		if (!Uisdigit(*p)) goto badlist;
-		newquota = newquota * 10 + *p - '0';
-                if (newquota < 0) goto badlist; /* overflow */
+	    /* accept "(storage -1)" as "()" to make move from unpatched cyrus possible */
+	    if (strcmp(arg.s, "-1") != 0) {
+		newquota = 0;
+		for (p = arg.s; *p; p++) {
+		    if (!Uisdigit(*p)) goto badlist;
+		    newquota = newquota * 10 + *p - '0';
+                    if (newquota < 0) goto badlist; /* overflow */
+		}
 	    }
 	    if (c == ')') break;
 	}
@@ -8854,9 +8857,16 @@
     
     /* note use of + to force the setting of a nonexistant
      * quotaroot */
-    prot_printf(xfer->be->out, "Q01 SETQUOTA {" SIZE_T_FMT "+}\r\n" \
-		"+%s (STORAGE %d)\r\n",
-		strlen(extname)+1, extname, quota.limit);
+    if (quota.limit == -1) {
+	prot_printf(xfer->be->out, "Q01 SETQUOTA {" SIZE_T_FMT "+}\r\n" \
+		    "+%s ()\r\n",
+		    strlen(extname)+1, extname);
+    }
+    else {
+	prot_printf(xfer->be->out, "Q01 SETQUOTA {" SIZE_T_FMT "+}\r\n" \
+		    "+%s (STORAGE %d)\r\n",
+		    strlen(extname)+1, extname, quota.limit);
+    }
     r = getresult(xfer->be->in, "Q01");
     if (r) syslog(LOG_ERR,
 		  "Could not move mailbox: %s, " \
