--- imap/backend.c	22 Apr 2008 13:11:17 -0000	1.55
+++ imap/backend.c	19 Sep 2008 02:37:30 -0000
@@ -305,6 +371,7 @@
     int sock = -1;
     int r;
     int err = -1;
+    int ask = 0;
     struct addrinfo hints, *res0 = NULL, *res;
     struct sockaddr_un sunsock;
     char buf[2048], *mechlist = NULL;
@@ -402,7 +469,7 @@
 				  &ret->capability, AUTO_BANNER);
 	if (!mechlist && !ret->capability) {
 	    /* found no capabilities in banner -> get them explicitly */
-	    prot->banner.is_capa = 0;
+	    ask = 1;
 	}
     }
     else {
@@ -417,9 +484,10 @@
 	    }
 	} while (strncasecmp(buf, prot->banner.resp,
 			     strlen(prot->banner.resp)));
+	ask = 1;
     }
 
-    if (!prot->banner.is_capa) {
+    if (ask) {
 	/* get the capabilities */
 	mechlist = ask_capability(ret->out, ret->in, prot,
 				  &ret->capability, AUTO_NO);
