Index: cyrus-imapd/perl/sieve/lib/request.c
===================================================================
RCS file: /home/cvs/repository/distribution/Network/cyrus-imapd/perl/sieve/lib/request.c,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 request.c
--- cyrus-imapd/perl/sieve/lib/request.c	13 Feb 2003 20:15:53 -0000	1.1.1.1
+++ cyrus-imapd/perl/sieve/lib/request.c	15 Feb 2010 16:08:10 -0000
@@ -356,6 +356,8 @@
     cnt+=amount;
   }
 
+  fclose(stream);
+
   prot_printf(pout,"\r\n");
   prot_flush(pout);
 
