From a72a8fbc24aca7b3cb5e981f68f7ec37febf3b27 Mon Sep 17 00:00:00 2001
From: =?utf-8?q?G=C3=A1bor=20Gomb=C3=A1s?= <gombasg@sztaki.hu>
Date: Sat, 29 Dec 2007 12:58:12 +0100
Subject: [PATCH] Fix multiple definition of TOUPPER/TOLOWER when SNMP is enabled

---
 master/master.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/master/master.c b/master/master.c
index 9ac3ab1..93c0393 100644
--- a/master/master.c
+++ b/master/master.c
@@ -91,6 +91,10 @@
 
   #include "cyrusMasterMIB.h"
 
+  /* Use our own definitions for these */
+  #undef TOUPPER
+  #undef TOLOWER
+
 #elif defined(HAVE_UCDSNMP)
   #include <ucd-snmp/ucd-snmp-config.h>
   #include <ucd-snmp/ucd-snmp-includes.h>
