--- cyrus-imapd-2.4.17/imap/mboxlist.c.orig	2013-01-07 13:48:24.000000000 +0100
+++ cyrus-imapd-2.4.17/imap/mboxlist.c	2013-01-07 13:52:32.000000000 +0100
@@ -3090,6 +3090,12 @@
 	break;
     }
 
+    if (config_auditlog)
+	syslog(LOG_NOTICE, "auditlog: subscribe sessionid=<%s> "
+	       "uid=<%s> mailbox=<%s> add=<%d> force=<%d> result=<%d>",
+	       session_id(),
+	       userid, name, add, force, r);
+
     sync_log_subscribe(userid, name);
     mboxlist_closesubs(subs);
     return r;
