--- imapd.conf.5	2007-08-15 13:23:15.000000000 -0400
+++ imapd.conf.5.pepper	2008-01-05 14:50:05.000000000 -0500
@@ -93,7 +93,7 @@
 typically used on backend servers in a Murder so that users can
 subscribe to mailboxes that don't reside on their "home" server.
 This option can also be used as a workaround for IMAP clients which
-don't play well with nonexistent or unselectable mailboxes (eg.
+don't play well with nonexistent or unselectable mailboxes (e.g.,
 Microsoft Outlook). 
 .IP "\fBallowanonymouslogin:\fR 0" 5
 Permit logins by the user "anonymous" using any password.  Also
@@ -103,8 +103,8 @@
 .PP
 Note that this command requires that SASL is compiled with APOP
 support, that the plaintext passwords are available in a SASL auxprop
-backend (eg. sasldb), and that the system can provide enough entropy
-(eg. from /dev/urandom) to create a challenge in the banner. 
+backend (e.g., sasldb), and that the system can provide enough entropy
+(e.g., from /dev/urandom) to create a challenge in the banner. 
 .IP "\fBallownewnews:\fR 0" 5
 Allow use of the NNTP NEWNEWS command.
 .PP
@@ -117,7 +117,7 @@
 or XFER.
 .PP
 Note that measures should be taken to make sure that the user being
-moved is not logged in, and can not login during the move.  Failure
+moved is not logged in, and cannot login during the move.  Failure
 to do so may result in the user's meta-data (seen state,
 subscriptions, etc) being corrupted or out of date. 
 .IP "\fBaltnamespace:\fR 0" 5
@@ -158,7 +158,7 @@
 environment. 
 .IP "\fBclient_timeout:\fR 10" 5
 Number of seconds to wait before returning a timeout failure when
-performing a client connection (e.g. in a murder environment) 
+performing a client connection (e.g., in a Murder environment) 
 .IP "\fBconfigdirectory:\fR <none>" 5
 The pathname of the IMAP configuration directory.  This field is
 required. 
@@ -209,20 +209,20 @@
 previously read messages being marked as unread in Microsoft
 Outlook, at the expense of a loss of performance/scalability. 
 .IP "\fBfoolstupidclients:\fR 0" 5
-If enabled, only list the personal namespace when a LIST "*" is performed.
-(it changes the request to a LIST "INBOX*" 
+If enabled, only list the personal namespace when a LIST "*" is performed
+(it changes the request to LIST "INBOX*").
 .IP "\fBforce_sasl_client_mech:\fR <none>" 5
 Force preference of a given SASL mechanism for client side operations
-(e.g. murder environments).  This is separate from (and overridden by)
+(e.g., Murder environments).  This is separate from (and overridden by)
 the ability to use the <host shortname>_mechs option to set preferred
 mechanisms for a specific host 
 .IP "\fBfulldirhash:\fR 0" 5
 If enabled, uses an improved directory hashing scheme which hashes
-the entire username instead of using just the first letter.  This
+on the entire username instead of just the first letter.  This
 changes hash algorithm used for quota and user directories and if
 \fIhashimapspool\fR is enabled, the entire mail spool.
 .PP
-Note that this option can NOT be changed on a live system.  The
+Note that this option CANNOT be changed on a live system.  The
 server must be quiesced and then the directories moved with the
 \fBrehash\fR utility. 
 .IP "\fBhashimapspool:\fR 0" 5
@@ -245,7 +245,7 @@
 .IP "\fBimapidlepoll:\fR 60" 5
 The interval (in seconds) for polling for mailbox changes and
 ALERTs while running the IDLE command.  This option is used when
-idled is not enabled or can not be contacted.  The minimum value is
+idled is not enabled or cannot be contacted.  The minimum value is
 1.  A value of 0 will disable IDLE. 
 .IP "\fBimapidresponse:\fR 1" 5
 If enabled, the server responds to an ID command with a parameter 
@@ -384,7 +384,7 @@
 ldap_use_sasl are enabled, ldap_version will be automatically
 set to 3. 
 .IP "\fBlmtp_downcase_rcpt:\fR 0" 5
-If enabled, lmtpd will convert the recipient address to lowercase
+If enabled, lmtpd will convert the recipient addresses to lowercase
 (up to a '+' character, if present). 
 .IP "\fBlmtp_fuzzy_mailbox_match:\fR 0" 5
 If enabled, and the mailbox specified in the detail part of the
@@ -459,7 +459,7 @@
 .IP "\fBmd5_dir:\fR <none>" 5
 Top level directory for MD5 store manipulated by make_md5. File
 structure within this directory is one file for each user on the system,
-hashed on the first letter of the userid (e.g: /var/imap/md5/d/dpc22). 
+hashed on the first letter of the userid (e.g., /var/imap/md5/d/dpc22). 
 .IP "\fBmd5_user_map:\fR <none>" 5
 Map file (cdb) to allow partial make_md5 runs. Maps username to UID 
 .IP "\fBmunge8bit:\fR 1" 5
@@ -580,7 +580,7 @@
 connections before the minpoll restriction is triggered.  Additionally,
 one padding entry is recovered every \fIpopminpoll\fR minutes.
 This allows for the occasional polling rate faster than popminpoll, 
-(i.e. for clients that require a send/receive to send mail) but still 
+(i.e., for clients that require a send/receive to send mail) but still 
 enforces the rate long-term.  Default is 1 (disabled).
 .br
 .sp
@@ -657,7 +657,7 @@
 The percent of quota utilization over which the server generates
 warnings. 
 .IP "\fBquotawarnkb:\fR 0" 5
-The maximum amount of free space (in kB) in which to give a quota
+The maximum amount of free space (in kB) at which to give a quota
 warning (if this value is 0, or if the quota is smaller than this
 amount, than warnings are always given). 
 .IP "\fBreject8bit:\fR 0" 5
@@ -717,7 +717,7 @@
 previously installed script will be unaffected by this option and
 will continue to execute regardless of the extensions used.  This
 option has no effect on options that are disabled at compile time
-(e.g. "regex"). 
+(e.g., "regex"). 
 
 Allowed values: \fIfileinto\fR, \fIreject\fR, \fIvacation\fR, \fIimapflags\fR, \fInotify\fR, \fIinclude\fR, \fIenvelope\fR, \fIbody\fR, \fIrelational\fR, \fIregex\fR, \fIsubaddress\fR, \fIcopy\fR
 .IP "\fBsieve_maxscriptsize:\fR 32" 5
@@ -776,7 +776,7 @@
 sync_client(8) for "rolling" replication. 
 .IP "\fBsync_machineid:\fR -1" 5
 Machine ID of this server which must be unique within a cluster.
-Any negative number, the default, will disable the use of UUIDs for
+Any negative number, including the default, will disable the use of UUIDs for
 replication. 
 .IP "\fBsync_password:\fR <none>" 5
 The default password to use when authenticating to a sync server. 
@@ -792,7 +792,7 @@
 .IP "\fBsyslog_prefix:\fR <none>" 5
 String to be prepended to the process name in syslog entries. 
 .IP "\fBtemp_path:\fR /tmp" 5
-The pathname to store temporary files in 
+The pathname to store temporary files in.
 .IP "\fBtimeout:\fR 30" 5
 The length of the IMAP server's inactivity autologout timer,       
 in minutes.  The minimum value is 30, the default. 
@@ -800,7 +800,7 @@
 File containing one or more Certificate Authority (CA) certificates. 
 .IP "\fBtls_ca_path:\fR <none>" 5
 Path to directory with certificates of CAs.  This directory must
-have filenames with the hashed value of the certificate (see
+have filenames with the hashed value of the certificates (see
 openssl(XXX)). 
 .IP "\fBtlscache_db:\fR berkeley-nosync" 5
 The cyrusdb backend to use for the TLS cache. 
@@ -824,7 +824,7 @@
 .IP "\fBumask:\fR 077" 5
 The umask value used by various Cyrus IMAP programs. 
 .IP "\fBusername_tolower:\fR 1" 5
-Convert usernames to all lowercase before login/authenticate.  This
+Convert usernames to all lowercase before login/authentication.  This
 is useful with authentication backends which ignore case during
 username lookups (such as LDAP).  
 .IP "\fBuserprefix:\fR Other Users" 5
