--- configure.in.old	2005-12-08 21:11:32.117913937 +0000
+++ configure.in	2005-12-08 21:11:55.510243542 +0000
@@ -857,12 +857,12 @@
     if test $ac_cv_header_ldap_h = yes -a $ac_cv_header_lber_h = yes; then
         CMU_OPENLDAP_API
 
-        if test $cmu_cv_openldap_api = yes; then
+        if test "$cmu_cv_openldap_api" = yes; then
             AC_CHECK_LIB(ldap, ldap_initialize, [ cmu_link_openldap="-lldap -llber" ], [ cmu_link_openldap=no ],-llber)
         fi
     fi
 
-    if test $cmu_cv_openldap_api = no -o "$cmu_link_openldap" = no; then
+    if test "$cmu_cv_openldap_api" = no -o "$cmu_link_openldap" = no; then
         AC_MSG_ERROR([Cannot enable LDAPDB plugin: Could not locate OpenLDAP])
     else
         CMU_OPENLDAP_COMPAT
