--- cmulocal/sasl2.m4.orig	Thu Aug 22 10:32:02 2002
+++ cmulocal/sasl2.m4	Thu Aug 22 10:52:07 2002
@@ -117,6 +117,11 @@
 
 	if test ${with_staticsasl} != "no"; then
 	  if test -d ${with_staticsasl}; then
+	    if test -d ${with_staticsasl}/lib64 ; then
+	      ac_cv_sasl_where_lib=${with_staticsasl}/lib64
+	    else
+	      ac_cv_sasl_where_lib=${with_staticsasl}/lib
+	    fi
 	    ac_cv_sasl_where_lib=${with_staticsasl}/lib
 	    ac_cv_sasl_where_inc=${with_staticsasl}/include
 
@@ -130,14 +135,17 @@
 
 	  AC_CHECK_HEADER(sasl/sasl.h, [
 	    AC_CHECK_HEADER(sasl/saslutil.h, [
-	     if test -r ${with_staticsasl}/lib/libsasl2.a; then
-		ac_cv_found_sasl=yes
-		AC_MSG_CHECKING(for static libsasl)
-		LIB_SASL="$LIB_SASL ${with_staticsasl}/lib/libsasl2.a"
-	     else
-	        AC_MSG_CHECKING(for static libsasl)
-		AC_ERROR([Could not find ${with_staticsasl}/lib/libsasl2.a])
-	     fi
+	     for i42 in lib64 lib; do
+	        if test -r ${with_staticsasl}/$i42/libsasl2.a; then
+		  ac_cv_found_sasl=yes
+		  AC_MSG_CHECKING(for static libsasl)
+		  LIB_SASL="$LIB_SASL ${with_staticsasl}/$i42/libsasl2.a"
+		fi
+	     done
+             if test ! "$ac_cv_found_sasl" = "yes"; then
+	          AC_MSG_CHECKING(for static libsasl)
+		  AC_ERROR([Could not find ${with_staticsasl}/lib*/libsasl2.a])
+             fi
 	    ])])
 
 	  AC_MSG_RESULT(found)
