commit 8fd43cc3558f416845158fdaced26814729b1cc1
Author: Matt Selsky <selsky@columbia.edu>
Date:   Mon Jun 14 22:30:34 2010 -0400

    Sync "usage" with man page.

diff --git a/imap/ctl_mboxlist.c b/imap/ctl_mboxlist.c
index 11f95a5..261d386 100644
--- a/imap/ctl_mboxlist.c
+++ b/imap/ctl_mboxlist.c
@@ -920,13 +920,13 @@ void do_verify(void)
 void usage(void)
 {
     fprintf(stderr, "DUMP:\n");
-    fprintf(stderr, "  ctl_mboxlist [-C <alt_config>] -d [-x] [-f filename] [-p partition]\n");
+    fprintf(stderr, "  ctl_mboxlist [-C <alt_config>] -d [-x] [-p partition] [-f filename]\n");
     fprintf(stderr, "UNDUMP:\n");
     fprintf(stderr,
 	    "  ctl_mboxlist [-C <alt_config>] -u [-f filename]"
 	    "    [< mboxlist.dump]\n");
     fprintf(stderr, "MUPDATE populate:\n");
-    fprintf(stderr, "  ctl_mboxlist [-C <alt_config>] -m [-a] [-w] [-f filename]\n");
+    fprintf(stderr, "  ctl_mboxlist [-C <alt_config>] -m [-a] [-w] [-i] [-f filename]\n");
     fprintf(stderr, "VERIFY:\n");
     fprintf(stderr, "  ctl_mboxlist [-C <alt_config>] -v [-f filename]\n");
     exit(1);
