--- imap/imapd.c	8 Oct 2008 15:47:06 -0000	1.553
+++ imap/imapd.c	12 Jan 2009 21:16:30 -0000
@@ -8676,8 +8732,8 @@
 	    }
 	} else if(!r) {
 	    /* this was a successful user delete, and we need to delete
-	       certain user meta-data (but not seen state!) */
-	    user_deletedata(mailboxname+5, imapd_userid, imapd_authstate, 0);
+	       certain user meta-data */
+	    user_deletedata(mailboxname+5, imapd_userid, imapd_authstate, 1);
 	    sync_log_user(mboxname_inbox_touserid(mailboxname));
 	}
 	
