--- cyrus-imapd-2.1.16/imap/mboxlist.c.orig	2004-02-25 17:00:24.684572832 -0300
+++ cyrus-imapd-2.1.16/imap/mboxlist.c	2004-02-25 17:00:28.380630279 -0300
@@ -855,7 +855,7 @@
 	if (!isadmin) { r = IMAP_PERMISSION_DENIED; goto done; }
     }
 
-    r = mboxlist_mylookup(name, &mbtype, &path, NULL, &acl, NULL, 1);
+    r = mboxlist_mylookup(name, &mbtype, &path, NULL, &acl, &tid, 1);
     switch (r) {
     case 0:
 	break;
