*** reconstruct.8~	Mon Jun 19 08:59:37 2006
--- reconstruct.8	Fri Aug 18 11:04:34 2006
***************
*** 61,66 ****
--- 61,73 ----
  [
  .B \-f
  ]
+ [
+ .B \-k  
+ ]
+ [
+ .B \-s
+ ]
+ 
  .IR mailbox ...
  .br
  .B reconstruct
***************
*** 130,135 ****
--- 137,150 ----
  a cyrus.header found there as new mailboxes.  Useful for
  restoring mailboxes from backups.
  .TP
+ .B \-k
+ Preserve the cyrus.expunge file and the corresponding message files 
+ instead of deleting them. 
+ .TP
+ .B \-s
+ Adds synchronization records to the log, so the synchronization 
+ service can replicate the changes.
+ .TP
  .B \-m
  .B NOTE: CURRENTLY UNAVAILABLE
  .br
