--- cyrus-imapd-2.2.7/perl/imap/Makefile.PL.perlfix	2004-07-28 14:46:56.000000000 +0200
+++ cyrus-imapd-2.2.7/perl/imap/Makefile.PL	2004-07-28 14:47:15.000000000 +0200
@@ -43,7 +43,7 @@
 use Config;
 
 my $SASL_INC = $ENV{SASL_INC};
-my $SASL_LIB = $ENV{SASL_LIB} || "-lsasl2";
+my $SASL_LIB = $ENV{SASL_LIB} || "-lsasl2 -ldb";
 
 my $OPENSSL_INC = $ENV{OPENSSL_INC};
 my $OPENSSL_LIB = $ENV{OPENSSL_LIB};
