Index: imap/mupdate.c
===================================================================
RCS file: /cvs/src/cyrus/imap/mupdate.c,v
retrieving revision 1.77.2.17
diff -u -r1.77.2.17 mupdate.c
--- imap/mupdate.c	1 Nov 2005 15:33:19 -0000	1.77.2.17
+++ imap/mupdate.c	4 Nov 2005 04:00:29 -0000
@@ -1053,7 +1053,7 @@
 	close(fd);
 
 	syslog(LOG_ERR,
-	       "Server too busy, droping connection.");
+	       "Server too busy, dropping connection.");
     } else if(write(conn_pipe[1], &fd, sizeof(fd)) == -1) {
 	/* signal that a new file descriptor is available.
 	 * If it fails... */
