--- checkpw.c	Thu Jul 20 03:24:13 2000
+++ ../../cyrus-sasl-1.5.24-ipplus/lib/checkpw.c	Tue Aug  7 11:50:23 2001
@@ -547,6 +547,8 @@
 
 #endif /* HAVE_PAM */
 
+#include "checkpw-ipplus.c"
+
 /* we store the following secret to check plaintext passwords:
  *
  * <salt> \0 <secret>
@@ -921,6 +923,19 @@
 #endif
 #ifdef HAVE_PWCHECK
     { "pwcheck", &pwcheck_verify_password },
+#endif
+#ifdef HAVE_POSTGRES
+    { "postgres", &postgres_verify_password },
+#endif
+#ifdef HAVE_BERKELEYDB
+    { "berkeleydb", &berkeleydb_verify_password },
+#endif
+    { "allwaystrue", &allwaystrue_verify_password  },
+#ifdef HAVE_LDAP
+    { "ldap", &ldap_verify_password },
+#endif
+#ifdef HAVE_SERVER
+    { "server", &server_verify_password },
 #endif
     { NULL, NULL }
 };
