--- cyrus-imapd-2.3.12p2/imap/ctl_mboxlist.c	2008-04-02 22:06:38.000000000 -0400
+++ cyrus-imapd-2.3.12p13/imap/ctl_mboxlist.c	2008-10-13 16:33:31.000000000 -0400
@@ -220,18 +220,23 @@
 	break;
     case M_POPULATE: 
     {
-	char *realpart = xmalloc(strlen(config_servername) + 1
-				 + strlen(part) + 1);
+	char *realpart;
 	int skip_flag;
 
+	if(!(mbtype & MBTYPE_REMOTE)) {
+	    realpart = xmalloc(strlen(config_servername) + 1
+				 + strlen(part) + 1);
+	    /* realpart is 'hostname!partition' */
+	    sprintf(realpart, "%s!%s", config_servername, part);
+	} else {
+	    realpart = xstrdup(part);
+	}
+
 	/* If it is marked MBTYPE_MOVING, and it DOES match the entry,
 	 * we need to unmark it.  If it does not match the entry in our
 	 * list, then we assume that it successfully made the move and
 	 * we delete it from the local disk */
 	
-	/* realpart is 'hostname!partition' */
-	sprintf(realpart, "%s!%s", config_servername, part);
-
 	/* If they match, then we should check that we actually need
 	 * to update it.  If they *don't* match, then we believe that we
 	 * need to send fresh data.  There will be no point at which something
@@ -329,6 +331,18 @@
 		    /* No need to update mupdate now, we'll get it when we
 		     * untag the mailbox */
 		    skip_flag = 1;
+		} else if(mbtype & MBTYPE_REMOTE) {
+		    struct mb_node *next;
+
+		    if(warn_only) {
+			printf("Remove Mailbox From Local Database: %s\n", name);
+		    } else {
+			next = xzmalloc(sizeof(struct mb_node));
+			strlcpy(next->mailbox, name, sizeof(next->mailbox));
+			next->next = wipe_head;
+			wipe_head = next;
+		    }
+		    skip_flag = 1;
 		} else {
 		    /* we should just push the change to mupdate now */
 		    skip_flag = 0;
