--- cyrus-imapd-2.2.12.orig/perl/imap/IMAP/Shell.pm	2006-10-24 19:13:31.000000000 +0200
+++ cyrus-imapd-2.2.12/perl/imap/IMAP/Shell.pm	2006-10-24 19:21:58.000000000 +0200
@@ -451,6 +451,12 @@
     cyradm_usage() if @ARGV;
   }
   my $cyradm;
+  if ($pw && -f $pw) {
+    my $pwh = IO::File->new($pw, O_RDONLY);
+    $pw = <$pwh>;
+    chomp($pw);
+    $pwh->close();
+  }
   if ($server ne '') {
     $cyradm = Cyrus::IMAP::Admin->new($server, $port)
       or die "cyradm: cannot connect to server\n";
