? annotations.diff
Index: imap/annotate.c
===================================================================
RCS file: /cvs/src/cyrus/imap/annotate.c,v
retrieving revision 1.16.2.16
diff -u -p -u -r1.16.2.16 annotate.c
--- imap/annotate.c	1 Jun 2006 13:23:23 -0000	1.16.2.16
+++ imap/annotate.c	17 Sep 2006 10:29:17 -0000
@@ -1790,6 +1790,15 @@ const struct annotate_st_entry mailbox_r
     { "/vendor/cmu/cyrus-imapd/condstore", ATTRIB_TYPE_BOOLEAN, BACKEND_ONLY,
       ATTRIB_VALUE_SHARED | ATTRIB_CONTENTTYPE_SHARED,
       ACL_ADMIN, annotation_set_condstore, NULL },
+    { "/vendor/kolab/folder-test", ATTRIB_TYPE_STRING, BACKEND_ONLY,
+      ATTRIB_VALUE_SHARED | ATTRIB_VALUE_PRIV,
+      ACL_ADMIN, annotation_set_todb, NULL },
+    { "/vendor/kolab/folder-type", ATTRIB_TYPE_STRING, BACKEND_ONLY,
+      ATTRIB_VALUE_SHARED | ATTRIB_VALUE_PRIV,
+      ACL_ADMIN, annotation_set_todb, NULL },
+    { "/vendor/kolab/incidences-for", ATTRIB_TYPE_STRING, BACKEND_ONLY,
+      ATTRIB_VALUE_SHARED | ATTRIB_VALUE_PRIV,
+      ACL_ADMIN, annotation_set_todb, NULL },
     { NULL, 0, ANNOTATION_PROXY_T_INVALID, 0, 0, NULL, NULL }
 };
 
