diff -u -r -p cyrus-imapd-2.3.13/lib/acl_afs.c cyrus-imapd.patched/lib/acl_afs.c
--- cyrus-imapd-2.3.13/lib/acl_afs.c	Mon Mar 24 18:43:08 2008
+++ cyrus-imapd.patched/lib/acl_afs.c	Tue Feb  3 15:12:16 2009
@@ -137,6 +137,12 @@ void *canonrock;
 	/* trying to delete invalid/non-existent identifier */
     }
 
+    /* Prevent ACLs for empty identifiers - most clients can't handle them */
+    if (strlen(identifier) == 0) {
+        if (newidentifier) free(newidentifier);
+        return -1;
+    }
+
     /* Find any existing entry for 'identifier' in 'acl' */
     for (thisid = nextid = *acl; *thisid; thisid = nextid) {
 	rights = strchr(thisid, '\t');
