--- cyrus-imapd-2.3.11.orig/imap/cyr_expire.c	2008-02-05 16:53:32.000000000 +0100
+++ cyrus-imapd-2.3.11/imap/cyr_expire.c	2008-02-05 16:44:33.000000000 +0100
@@ -55,6 +55,7 @@
 #include <syslog.h>
 #include <errno.h>
 #include <signal.h>
+#include <sasl/sasl.h>
 
 #include "annotate.h"
 #include "cyrusdb.h"
@@ -408,6 +409,7 @@
     if (!expire_days) usage();
 
     cyrus_init(alt_config, "cyr_expire", 0);
+    global_sasl_init(1,0,NULL);
 
     annotatemore_init(0, NULL, NULL);
     annotatemore_open(NULL);
@@ -507,6 +509,7 @@
     annotatemore_close();
     annotatemore_done();
     duplicate_done();
+    sasl_done();
     cyrus_done();
 
     exit(r);
