diff -ruN cyrus-imapd/imap/imapd.c cyrus-imapd.new/imap/imapd.c
--- cyrus-imapd/imap/imapd.c    2006-01-13 23:06:28.000000000 +0100
+++ cyrus-imapd.new/imap/imapd.c        2006-03-20 01:29:39.330385000 +0100
@@ -822,7 +822,12 @@
     const char *err;

     prot_printf(imapd_out,
-               "* OK %s Cyrus IMAP4 %s server ready\r\n", config_servername,
+               "* OK %s Cyrus IMAP4 ", config_servername);
+
+    if (config_getswitch(IMAPOPT_SHOWVERSION))
+           prot_printf(imapd_out, "%s ", CYRUS_VERSION);
+
+    prot_printf(imapd_out, "server ready\r\n", config_servername,
                CYRUS_VERSION);

     ret = snprintf(motdfilename, sizeof(motdfilename), "%s/msg/motd",
diff -ruN cyrus-imapd/lib/imapoptions cyrus-imapd.new/lib/imapoptions
--- cyrus-imapd/lib/imapoptions 2006-01-17 00:40:14.000000000 +0100
+++ cyrus-imapd.new/lib/imapoptions     2006-03-20 01:16:06.091561000 +0100
@@ -756,6 +756,9 @@
 /* If using the alternate IMAP namespace, the prefix for the shared
    namespace.  The hierarchy delimiter will be automatically appended. */

+{ "showversion", 1, SWITCH }
+/* If disabled, the server doesn't show its version on login. */
+
 { "sieve_maxscriptsize", 32, INT }
 /* Maximum size (in kilobytes) any sieve script can be, enforced at
    submission by timsieved(8). */

