*** mbdump.Oc	Mon Feb  5 12:41:47 2007
--- mbdump.c	Mon Mar 12 18:26:16 2007
***************
*** 418,424 ****
  
  	    if (!mbdir) {
  		syslog(LOG_ERR,
! 		       "could not dump sieve scripts in %s: %m)", mbpath);
  	    } else {
  		char tag_fname[2048];
  	    
--- 418,424 ----
  
  	    if (!mbdir) {
  		syslog(LOG_ERR,
! 		       "could not dump sieve scripts in %s: %m)", sieve_path);
  	    } else {
  		char tag_fname[2048];
  	    
***************
*** 439,444 ****
--- 439,448 ----
  				     "SIEVE-%s", next->d_name);
  			}
  
+ 			/* construct path/filename */
+ 			snprintf(filename, sizeof(filename), "%s/%s",
+ 				 sieve_path, next->d_name);
+ 
  			/* dump file */
  			r = dump_file(0, !tag, pin, pout, filename, tag_fname);
  			if (r) goto done;
