Index: cyrus-imapd-2.3.14/lib/auth_krb5.c
===================================================================
--- cyrus-imapd-2.3.14.orig/lib/auth_krb5.c
+++ cyrus-imapd-2.3.14/lib/auth_krb5.c
@@ -137,7 +137,7 @@ static char *mycanonifyid(const char *id
 
     /* build dummy princ to compare realms */
     if (krb5_build_principal(context,&princ_dummy,
-			     strlen(realm),realm,"dummy",0))
+			     strlen(realm),realm,"dummy",NULL))
     {
 	krb5_free_principal(context,princ);
 	krb5_free_context(context);
