From b03c49e3cd24df576e3a541a92009f05664e1832 Mon Sep 17 00:00:00 2001
From: =?utf-8?q?G=C3=A1bor=20Gomb=C3=A1s?= <gombasg@sztaki.hu>
Date: Sat, 29 Dec 2007 17:13:29 +0100
Subject: [PATCH] Fix the name of imclient_havetls() in imclient.h

---
 lib/imclient.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/lib/imclient.h b/lib/imclient.h
index d0efe5b..3df760e 100644
--- a/lib/imclient.h
+++ b/lib/imclient.h
@@ -88,7 +88,7 @@ extern int imclient_authenticate(struct imclient *imclient,
 				 int minssf, 
 				 int maxssf);
 
-extern int imclient_have_tls (void);
+extern int imclient_havetls (void);
 
 extern int imclient_starttls(struct imclient *imclient,
 			     char *cert_file, 

