--- imap/imapd.c.orig	Wed Jul  9 23:51:18 2003
+++ imap/imapd.c	Thu Jul 10 00:21:15 2003
@@ -308,8 +308,9 @@
 	realm++;
 	val = config_getstring("loginrealms", "");
 	while (*val) {
-	    if (!strncasecmp(val, realm, strlen(realm)) &&
-		(!val[strlen(realm)] || isspace((int) val[strlen(realm)]))) {
+	    if ((!strncasecmp(val, realm, strlen(realm)) &&
+		(!val[strlen(realm)] || isspace((int) val[strlen(realm)])))
+                || (val[0]=='*')) {
 		break;
 	    }
 	    /* not this realm, try next one */
--- imap/config.c.orig	Mon Apr 14 15:31:36 2003
+++ imap/config.c	Sun Aug 31 03:49:24 2003
@@ -481,8 +481,9 @@
 	realm++;
 	val = config_getstring("loginrealms", "");
 	while (*val) {
-	    if (!strncasecmp(val, realm, strlen(realm)) &&
-		(!val[strlen(realm)] || isspace((int) val[strlen(realm)]))) {
+	    if ((!strncasecmp(val, realm, strlen(realm)) &&
+		(!val[strlen(realm)] || isspace((int) val[strlen(realm)])))
+                || (val[0]=='*')) {
 		break;
 	    }
 	    /* not this realm, try next one */
