Index: cyrus-sasl-2.1.15/acconfig.h
diff -u cyrus-sasl-2.1.15/acconfig.h.orig cyrus-sasl-2.1.15/acconfig.h
--- cyrus-sasl-2.1.15/acconfig.h.orig	Sat Jul 26 01:53:30 2003
+++ cyrus-sasl-2.1.15/acconfig.h	Sat Jul 26 01:54:30 2003
@@ -296,8 +296,12 @@
 #ifndef HAVE_GETADDRINFO
 #define	getaddrinfo	sasl_getaddrinfo
 #define	freeaddrinfo	sasl_freeaddrinfo
-#define	getnameinfo	sasl_getnameinfo
 #define	gai_strerror	sasl_gai_strerror
+#endif
+#ifndef HAVE_GETNAMEINFO
+#define	getnameinfo	sasl_getnameinfo
+#endif
+#if !defined(HAVE_GETADDRINFO) || !defined(HAVE_GETNAMEINFO)
 #include "gai.h"
 #endif
 
Index: cyrus-sasl-2.1.15/include/gai.h
diff -u cyrus-sasl-2.1.15/include/gai.h.orig cyrus-sasl-2.1.15/include/gai.h
--- cyrus-sasl-2.1.15/include/gai.h.orig	Wed May 14 06:34:28 2003
+++ cyrus-sasl-2.1.15/include/gai.h	Sat Jul 26 01:55:54 2003
@@ -96,12 +96,17 @@
 extern "C" {
 #endif
 
-int	getaddrinfo(const char *, const char *,
-		    const struct addrinfo *, struct addrinfo **);
+#ifndef HAVE_GETNAMEINFO
 int	getnameinfo(const struct sockaddr *, socklen_t, char *,
 		    size_t, char *, size_t, int);
+#endif
+
+#ifndef HAVE_GETADDRINFO
+int	getaddrinfo(const char *, const char *,
+		    const struct addrinfo *, struct addrinfo **);
 void	freeaddrinfo(struct addrinfo *);
 char	*gai_strerror(int);
+#endif
 
 #ifdef __cplusplus
 }
