diff --git a/imap/lmtpengine.c b/imap/lmtpengine.c
index 7159c34..5027eb6 100644
--- a/imap/lmtpengine.c
+++ b/imap/lmtpengine.c
@@ -1817,7 +1817,10 @@ static void pushmsg(struct protstream *in, struct protstream *out,
 
     if (!isdotstuffed) {
 	/* signify end of message */
-	prot_printf(out, "\r\n.\r\n");
+	if (!lastline_hadendline) {
+	    prot_printf(out, "\r\n");
+	}
+	prot_printf(out, ".\r\n");
     }
 }
 
