--- imap/unexpunge.c.orig	Fri Mar 31 22:22:28 2006
+++ imap/unexpunge.c	Tue Nov  7 12:42:59 2006
@@ -504,7 +504,7 @@
 
     if (r || expunge_fd == -1) {
 	/* mailbox corrupt/nonexistent -- skip it */
-	syslog(LOG_WARNING, "unable to open/lock mailbox %s", argv[optind]);
+	syslog(LOG_WARNING, "unable to open/lock mailbox %s", argv[optind-1]);
 	if (doclose) mailbox_close(&mailbox);
 	return 0;
     }
