diff -Naur cyrus-sasl-2.1.21.old/saslauthd/auth_pam.c cyrus-sasl-2.1.21/saslauthd/auth_pam.c
--- cyrus-sasl-2.1.21.old/saslauthd/auth_pam.c	2005-05-15 07:34:43.000000000 +0100
+++ cyrus-sasl-2.1.21/saslauthd/auth_pam.c	2005-09-14 16:48:41.114079000 +0100
@@ -46,10 +46,6 @@
 #include "mechanisms.h"
 #include <stdio.h>
 
-#ifdef HAVE_CONFIG_H
-#include <config.h>
-#endif
-
 #ifdef AUTH_PAM
 
 # include <string.h>
diff -Naur cyrus-sasl-2.1.21.old/saslauthd/configure.in cyrus-sasl-2.1.21/saslauthd/configure.in
--- cyrus-sasl-2.1.21.old/saslauthd/configure.in	2005-05-07 05:54:18.000000000 +0100
+++ cyrus-sasl-2.1.21/saslauthd/configure.in	2005-09-14 17:24:48.297581540 +0100
@@ -88,6 +88,7 @@
     CPPFLAGS="$CPPFLAGS -I${with_pam}/include"
     LDFLAGS="$LDFLAGS -L${with_pam}/lib"
   fi
+  AC_CHECK_HEADERS(security/pam_appl.h pam/pam_appl.h)
   cmu_save_LIBS="$LIBS"
   AC_CHECK_LIB(pam, pam_start, [
 	  AC_CHECK_HEADER(security/pam_appl.h,,
