--- cyrus-imapd-2.3.12p2/imap/mailbox.c	2008-03-24 13:09:17.000000000 -0400
+++ cyrus-imapd-2.3.12p11/imap/mailbox.c	2008-07-11 14:18:35.000000000 -0400
@@ -2845,6 +2845,8 @@
     strlcpy(nbuf, mailbox->name, sizeof(nbuf));
     ntail = nbuf + strlen(nbuf);
     mailbox_close(mailbox);
+    /* log the delete after close, but before mucking with nbuf */
+    syslog(LOG_NOTICE, "Deleted mailbox %s", nbuf);
 
     do {
 	/* Check if the mailbox has children */
@@ -2888,8 +2894,6 @@
     } while(r == IMAP_MAILBOX_NONEXISTENT);
 
     *ntail = '\0';
-    syslog(LOG_NOTICE, "Deleted mailbox %s", nbuf);
-
     if (updatenotifier) updatenotifier(nbuf);
 
     return 0;
