*** checkpw.Oc	Tue Jul 22 16:44:32 2003
--- checkpw.c	Wed Aug 20 09:13:16 2003
***************
*** 542,548 ****
      arg.rbuf = response;
      arg.rsize = sizeof(response);
  
!     door_call(s, &arg);
  
      if (arg.data_ptr != response || arg.data_size >= sizeof(response)) {
  	/* oh damn, we got back a really long response */
--- 542,552 ----
      arg.rbuf = response;
      arg.rsize = sizeof(response);
  
!     if (door_call(s, &arg) < 0) {
!       /* Parameters are undefined */
! 	sasl_seterror(conn, 0, "door call to saslauthd server failed: %m", errno);
! 	goto fail;
!     }
  
      if (arg.data_ptr != response || arg.data_size >= sizeof(response)) {
  	/* oh damn, we got back a really long response */
