diff -urN cyrus-imapd-2.2.12.orig/imap/lmtpproxyd.c cyrus-imapd-2.2.12/imap/lmtpproxyd.c
--- cyrus-imapd-2.2.12.orig/imap/lmtpproxyd.c	2004-12-17 17:32:16.000000000 +0100
+++ cyrus-imapd-2.2.12/imap/lmtpproxyd.c	2005-08-24 14:37:26.000000000 +0200
@@ -231,6 +231,10 @@
 
     while ((opt = getopt(argc, argv, "q")) != EOF) {
 	switch(opt) {
+	case 'a':
+	    mylmtp.preauth = 1;
+	    break;
+
 	case 'q':
 	    quotaoverride = 1;
 	    break;
