diff -Naur cyrus-imapd-2.4.2/contrib/cyrus-graphtools.1.0/cgi-bin/cyrus_master.pl cyrus-imapd-2.4.2.new/contrib/cyrus-graphtools.1.0/cgi-bin/cyrus_master.pl
--- cyrus-imapd-2.4.2/contrib/cyrus-graphtools.1.0/cgi-bin/cyrus_master.pl	2010-10-19 05:08:28.000000000 +0200
+++ cyrus-imapd-2.4.2.new/contrib/cyrus-graphtools.1.0/cgi-bin/cyrus_master.pl	2010-10-26 16:45:16.498789198 +0200
@@ -1,4 +1,4 @@
-#!/usr/local/bin/perl5
+#!/usr/bin/perl
 
 #
 # Created by Alison Greenwald <alison@andrew.cmu.edu> 21 Sep 2000
diff -Naur cyrus-imapd-2.4.2/contrib/cyrus-graphtools.1.0/cgi-bin/graph_cyrus_db.pl cyrus-imapd-2.4.2.new/contrib/cyrus-graphtools.1.0/cgi-bin/graph_cyrus_db.pl
--- cyrus-imapd-2.4.2/contrib/cyrus-graphtools.1.0/cgi-bin/graph_cyrus_db.pl	2010-10-19 05:08:28.000000000 +0200
+++ cyrus-imapd-2.4.2.new/contrib/cyrus-graphtools.1.0/cgi-bin/graph_cyrus_db.pl	2010-10-26 16:45:28.102738590 +0200
@@ -1,4 +1,4 @@
-#!/usr/local/bin/perl5 
+#!/usr/bin/perl
 
 #
 # Created by Alison Greenwald <alison@andrew.cmu.edu> 21 Sep 2000
diff -Naur cyrus-imapd-2.4.2/contrib/cyrus-graphtools.1.0/cgi-bin/graph_cyrus_db-sum.pl cyrus-imapd-2.4.2.new/contrib/cyrus-graphtools.1.0/cgi-bin/graph_cyrus_db-sum.pl
--- cyrus-imapd-2.4.2/contrib/cyrus-graphtools.1.0/cgi-bin/graph_cyrus_db-sum.pl	2010-10-19 05:08:28.000000000 +0200
+++ cyrus-imapd-2.4.2.new/contrib/cyrus-graphtools.1.0/cgi-bin/graph_cyrus_db-sum.pl	2010-10-26 16:45:22.536931298 +0200
@@ -1,4 +1,4 @@
-#!/usr/local/bin/perl5
+#!/usr/bin/perl
 
 #
 # Created by Alison Greenwald 21 Sep 2000
diff -Naur cyrus-imapd-2.4.2/contrib/cyrus-graphtools.1.0/script/cyrus.pl cyrus-imapd-2.4.2.new/contrib/cyrus-graphtools.1.0/script/cyrus.pl
--- cyrus-imapd-2.4.2/contrib/cyrus-graphtools.1.0/script/cyrus.pl	2010-10-19 05:08:28.000000000 +0200
+++ cyrus-imapd-2.4.2.new/contrib/cyrus-graphtools.1.0/script/cyrus.pl	2010-10-26 16:45:33.544169475 +0200
@@ -1,4 +1,4 @@
-#!/usr/local/bin/perl5 
+#!/usr/bin/perl
 
 # This will read information from the cyrus MIB for all devices specified
 # in cyrusrc 
diff -Naur cyrus-imapd-2.4.2/contrib/mupdate-test.pl cyrus-imapd-2.4.2.new/contrib/mupdate-test.pl
--- cyrus-imapd-2.4.2/contrib/mupdate-test.pl	2010-10-19 05:08:28.000000000 +0200
+++ cyrus-imapd-2.4.2.new/contrib/mupdate-test.pl	2010-10-26 16:45:44.238829842 +0200
@@ -1,4 +1,4 @@
-#!/usr/local/bin/perl
+#!/usr/bin/perl
 
 # Create a workload against a murder frontend
 # that will give the MUPDATE server a workout.
diff -Naur cyrus-imapd-2.4.2/perl/imap/examples/auditmbox.pl cyrus-imapd-2.4.2.new/perl/imap/examples/auditmbox.pl
--- cyrus-imapd-2.4.2/perl/imap/examples/auditmbox.pl	2010-10-19 05:08:28.000000000 +0200
+++ cyrus-imapd-2.4.2.new/perl/imap/examples/auditmbox.pl	2010-10-26 16:47:48.402852914 +0200
@@ -1,4 +1,4 @@
-#! /usr/bin/perl -w
+#!/usr/bin/perl -w
 #
 # Copyright (c) 1994-2008 Carnegie Mellon University.  All rights reserved.
 #
diff -Naur cyrus-imapd-2.4.2/perl/imap/examples/imapcollate.pl cyrus-imapd-2.4.2.new/perl/imap/examples/imapcollate.pl
--- cyrus-imapd-2.4.2/perl/imap/examples/imapcollate.pl	2010-10-19 05:08:28.000000000 +0200
+++ cyrus-imapd-2.4.2.new/perl/imap/examples/imapcollate.pl	2010-10-26 16:47:45.144486424 +0200
@@ -1,4 +1,4 @@
-#! /usr/bin/perl -w
+#!/usr/bin/perl -w
 #
 # Copyright (c) 1994-2008 Carnegie Mellon University.  All rights reserved.
 #
diff -Naur cyrus-imapd-2.4.2/perl/imap/examples/imapdu.pl cyrus-imapd-2.4.2.new/perl/imap/examples/imapdu.pl
--- cyrus-imapd-2.4.2/perl/imap/examples/imapdu.pl	2010-10-19 05:08:28.000000000 +0200
+++ cyrus-imapd-2.4.2.new/perl/imap/examples/imapdu.pl	2010-10-26 16:50:40.416830168 +0200
@@ -1,4 +1,4 @@
-#! /usr/local/bin/perl -w
+#!/usr/bin/perl -w
 #
 # Copyright (c) 1994-2008 Carnegie Mellon University.  All rights reserved.
 #
diff -Naur cyrus-imapd-2.4.2/perl/imap/examples/test-imsp.pl cyrus-imapd-2.4.2.new/perl/imap/examples/test-imsp.pl
--- cyrus-imapd-2.4.2/perl/imap/examples/test-imsp.pl	2010-10-19 05:08:28.000000000 +0200
+++ cyrus-imapd-2.4.2.new/perl/imap/examples/test-imsp.pl	2010-10-26 16:45:56.902012999 +0200
@@ -1,4 +1,4 @@
-#!/usr/local/bin/perl -w
+#!/usr/bin/perl -w
 
 # test-imsp.pl
 #
diff -Naur cyrus-imapd-2.4.2/tools/arbitronsort.pl cyrus-imapd-2.4.2.new/tools/arbitronsort.pl
--- cyrus-imapd-2.4.2/tools/arbitronsort.pl	2010-10-19 05:08:28.000000000 +0200
+++ cyrus-imapd-2.4.2.new/tools/arbitronsort.pl	2010-10-26 16:45:03.533772317 +0200
@@ -1,4 +1,4 @@
-#!/usr/local/bin/perl
+#!/usr/bin/perl
 
 #
 # This script takes the output of arbitron (run without the -o option)
