--- cyrus-imapd-2.2.8.dist/perl/imap/IMAP/Admin.pm	2004-02-19 14:50:12.000000000 -0800
+++ cyrus-imapd-2.2.8/perl/imap/IMAP/Admin.pm	2005-01-11 08:08:31.410441790 -0800
@@ -337,6 +337,7 @@
 			next unless $d{-text} =~ s/^\(([^\)]*)\) //;
 			my $attrs = $1;
 			my $sep = '';
+			my $mbox;
 			# NIL or (attrs) "sep" "str"
 			if ($d{-text} =~ /^N/) {
 			  return if $d{-text} !~ s/^NIL//;
@@ -345,8 +346,10 @@
 			  $sep = $1;
 			}
 			return unless $d{-text} =~ s/^ //;
-			my $mbox;
-			if ($d{-text} =~ /\"(([^\\\"]*\\)*[^\\\"]*)\"/) {
+                        if ($d{-text} =~ /{\d+}(.*)/) {
+                            # cope with literals (?)
+			    (undef, $mbox) = split(/\n/, $d{-text});
+                        } elsif ($d{-text} =~ /\"(([^\\\"]*\\)*[^\\\"]*)\"/) {
 			  ($mbox = $1) =~ s/\\(.)/$1/g;
 			} else {
 			  $d{-text} =~ /^([]!\#-[^-~]+)/;
