Index: configure.in
===================================================================
RCS file: /home/cvs/debian/cyrus21-imapd/configure.in,v
retrieving revision 1.12
retrieving revision 1.13
diff -u -r1.12 -r1.13
--- configure.in	16 Apr 2002 20:12:42 -0000	1.12
+++ configure.in	10 May 2002 13:27:28 -0000	1.13
@@ -50,6 +50,10 @@
 AC_CONFIG_HEADER(config.h)
 AC_CANONICAL_HOST
 
+dnl Useful hook for distributions
+AC_ARG_WITH(extraident,[  --with-extraindent=STRING  use STRING as extra version information],
+	[AC_DEFINE_UNQUOTED(EXTRA_IDENT,"$withval", [Extra version information for imap/version.h])])
+
 AC_CHECK_PROG(MAKEDEPEND,makedepend,makedepend,[`cd ${srcdir};pwd`/tools/not-mkdep])
 if test "$MAKEDEPEND" != "makedepend"; then
         AC_MSG_WARN([Makedepend is not installed on this system.  You should compile and install the version from the makedepend subdirectory.])
