Index: sieve.y
===================================================================
RCS file: /afs/andrew/system/cvs/src/sieve/sieve.y,v
retrieving revision 1.19.4.4
diff -u -c -r1.19.4.4 sieve.y
cvs server: conflicting specifications of output style
*** sieve.y	27 Feb 2003 18:39:44 -0000	1.19.4.4
--- sieve.y	13 Mar 2003 15:16:56 -0000
***************
*** 121,126 ****
--- 121,127 ----
  
  int yyerror(char *msg);
  extern int yylex(void);
+ extern void yyrestart(FILE *f);
  
  #define YYERROR_VERBOSE /* i want better error messages! */
  %}
***************
*** 534,543 ****
  commandlist_t *sieve_parse(sieve_script_t *script, FILE *f)
  {
      commandlist_t *t;
-     extern FILE *yyin;
  
-     yyin = f;
      parse_script = script;
      if (yyparse()) {
  	t = NULL;
      } else {
--- 535,543 ----
  commandlist_t *sieve_parse(sieve_script_t *script, FILE *f)
  {
      commandlist_t *t;
  
      parse_script = script;
+     yyrestart(f);
      if (yyparse()) {
  	t = NULL;
      } else {
