*** tools/config2header.orig	Mon Dec  7 23:32:17 2015
--- tools/config2header	Mon Dec  7 23:36:23 2015
***************
*** 124,129 ****
--- 124,141 ----
  #include <string.h>
  #include "imapopts.h"
  
+ /*
+  * Sun C Compilers are more strict than GNU and won't allow type
+  * casting to a union
+  */
+ 
+ #if defined(__SUNPRO_C) || defined(__SUNPRO_CC)
+ #define U_CFG_V
+ #else
+ #define U_CFG_V (union config_value)
+ #endif
+ 
+ 
  EXPORTED struct imapopt_s imapopts[] =
  {
    { IMAPOPT_ZERO, "", 0, OPT_NOTOPT, { NULL }, { NULL }, { { NULL, IMAP_ENUM_ZERO } } },
***************
*** 150,156 ****
  
  	if ($3 eq "STRING") {
  	    $def = $use_gcc_extension
! 			? "(union config_value)((const char *) $2)"
  			: "{(void *)($2)}";
  	} elsif ($3 eq "ENUM") {
  	    my @values = eval $4;
--- 162,168 ----
  
  	if ($3 eq "STRING") {
  	    $def = $use_gcc_extension
! 			? "U_CFG_V((const char *) $2)"
  			: "{(void *)($2)}";
  	} elsif ($3 eq "ENUM") {
  	    my @values = eval $4;
***************
*** 163,169 ****
  	    $e = "IMAP_ENUM_" . uc($opt) . "_" . uc($def);
  	    $e =~ s/[^0-9A-Z_a-z]/_/g;
  	    $def = $use_gcc_extension
! 			? "(union config_value)((enum enum_value) $e)"
  			: "{(void *)($e)}";
  
  	    # output the enum_options
--- 175,181 ----
  	    $e = "IMAP_ENUM_" . uc($opt) . "_" . uc($def);
  	    $e =~ s/[^0-9A-Z_a-z]/_/g;
  	    $def = $use_gcc_extension
! 			? "U_CFG_V((enum enum_value) $e)"
  			: "{(void *)($e)}";
  
  	    # output the enum_options
***************
*** 189,195 ****
  	    my $count = 0;
  
  	    $def = $use_gcc_extension
! 			? "(union config_value)((const char *) $2)"
  			: "{(void*)($2)}";
  
  	    # output the enum_options
--- 201,207 ----
  	    my $count = 0;
  
  	    $def = $use_gcc_extension
! 			? "U_CFG_V((const char *) $2)"
  			: "{(void*)($2)}";
  
  	    # output the enum_options
***************
*** 219,225 ****
  	    }
  	    $e .= "0";
  	    $def = $use_gcc_extension
! 			? "(union config_value)((unsigned long) $e)"
  			: "{(void *)($e)}";
  
  	    # output the enum_options
--- 231,237 ----
  	    }
  	    $e .= "0";
  	    $def = $use_gcc_extension
! 			? "U_CFG_V((unsigned long) $e)"
  			: "{(void *)($e)}";
  
  	    # output the enum_options
***************
*** 242,248 ****
  	    if ($count > $enum_size) { $enum_size = $count; }
  	} else {
  	    $def = $use_gcc_extension
! 			? "(union config_value)((long) $2)"
  			: "{(void*)$2}";
  	}
  
--- 254,260 ----
  	    if ($count > $enum_size) { $enum_size = $count; }
  	} else {
  	    $def = $use_gcc_extension
! 			? "U_CFG_V((long) $2)"
  			: "{(void*)$2}";
  	}
  
