Index: sieve.y
===================================================================
RCS file: /cvs/src/sieve/sieve.y,v
retrieving revision 1.33
diff -u -r1.33 sieve.y
--- sieve.y	30 Nov 2006 17:11:25 -0000	1.33
+++ sieve.y	9 Aug 2007 12:55:33 -0000
@@ -201,7 +201,7 @@
 	;
 
 require: REQUIRE stringlist ';'	{ if (!check_reqs($2)) {
-                                    yyerror("unsupported feature");
+                                    yyerror("Unsupported features in require line");
 				    YYERROR; 
                                   } }
 	;
@@ -221,7 +221,7 @@
 	;
 
 action: REJCT STRING             { if (!parse_script->support.reject) {
-				     yyerror("reject not required");
+				     yyerror("reject require missing");
 				     YYERROR;
 				   }
 				   if (!verify_utf8($2)) {
@@ -230,7 +230,7 @@
 				   $$ = new_command(REJCT);
 				   $$->u.str = $2; }
 	| FILEINTO copy STRING	 { if (!parse_script->support.fileinto) {
-				     yyerror("fileinto not required");
+				     yyerror("fileinto require missing");
 	                             YYERROR;
                                    }
 				   if (!verify_mailbox($3)) {
@@ -245,7 +245,7 @@
 	| STOP			 { $$ = new_command(STOP); }
 	| DISCARD		 { $$ = new_command(DISCARD); }
 	| VACATION vtags STRING  { if (!parse_script->support.vacation) {
-				     yyerror("vacation not required");
+				     yyerror("vacation require missing");
 				     YYERROR;
 				   }
 				   if (($2->mime == -1) && !verify_utf8($3)) {
@@ -254,7 +254,7 @@
   				   $$ = build_vacation(VACATION,
 					    canon_vtags($2), $3); }
         | SETFLAG stringlist     { if (!parse_script->support.imapflags) {
-                                    yyerror("imapflags not required");
+                                    yyerror("imapflags require missing");
                                     YYERROR;
                                    }
                                   if (!verify_stringlist($2, verify_flag)) {
@@ -263,7 +263,7 @@
                                   $$ = new_command(SETFLAG);
                                   $$->u.sl = $2; }
          | ADDFLAG stringlist     { if (!parse_script->support.imapflags) {
-                                    yyerror("imapflags not required");
+                                    yyerror("imapflags require missing");
                                     YYERROR;
                                     }
                                   if (!verify_stringlist($2, verify_flag)) {
@@ -272,7 +272,7 @@
                                   $$ = new_command(ADDFLAG);
                                   $$->u.sl = $2; }
          | REMOVEFLAG stringlist  { if (!parse_script->support.imapflags) {
-                                    yyerror("imapflags not required");
+                                    yyerror("imapflags require missing");
                                     YYERROR;
                                     }
                                   if (!verify_stringlist($2, verify_flag)) {
@@ -281,18 +281,18 @@
                                   $$ = new_command(REMOVEFLAG);
                                   $$->u.sl = $2; }
          | MARK                   { if (!parse_script->support.imapflags) {
-                                    yyerror("imapflags not required");
+                                    yyerror("imapflags require missing");
                                     YYERROR;
                                     }
                                   $$ = new_command(MARK); }
          | UNMARK                 { if (!parse_script->support.imapflags) {
-                                    yyerror("imapflags not required");
+                                    yyerror("imapflags require missing");
                                     YYERROR;
                                     }
                                   $$ = new_command(UNMARK); }
 
          | NOTIFY ntags           { if (!parse_script->support.notify) {
-				       yyerror("notify not required");
+				       yyerror("notify require missing");
 				       $$ = new_command(NOTIFY); 
 				       YYERROR;
 	 			    } else {
@@ -300,7 +300,7 @@
 				             canon_ntags($2));
 				    } }
          | DENOTIFY dtags         { if (!parse_script->support.notify) {
-                                       yyerror("notify not required");
+                                       yyerror("notify require missing");
 				       $$ = new_command(DENOTIFY);
 				       YYERROR;
 				    } else {
@@ -310,14 +310,14 @@
 			YYERROR; } } }
 
 	 | INCLUDE location STRING { if (!parse_script->support.include) {
-				     yyerror("include not required");
+				     yyerror("include require missing");
 	                             YYERROR;
                                    }
 	                           $$ = new_command(INCLUDE);
 				   $$->u.inc.location = $2;
 				   $$->u.inc.script = $3; }
          | RETURN		 { if (!parse_script->support.include) {
-                                    yyerror("include not required");
+                                    yyerror("include require missing");
                                     YYERROR;
                                   }
                                    $$ = new_command(RETURN); }
@@ -482,7 +482,7 @@
 	| BODY btags stringlist
 				 {
 				     if (!parse_script->support.body) {
-                                       yyerror("body not required");
+                                       yyerror("body require missing");
 				       YYERROR;
 				     }
 					
@@ -513,7 +513,7 @@
 
 addrorenv: ADDRESS		 { $$ = ADDRESS; }
 	| ENVELOPE		 {if (!parse_script->support.envelope)
-	                              {yyerror("envelope not required"); YYERROR;}
+	                              {yyerror("envelope require missing"); YYERROR;}
 	                          else{$$ = ENVELOPE; }
 	                         }
 
@@ -542,7 +542,7 @@
 			yyerror("duplicate comparator tag"); YYERROR; }
 				   else if (!strcmp($3, "i;ascii-numeric") &&
 					    !parse_script->support.i_ascii_numeric) {
-			yyerror("comparator-i;ascii-numeric not required");
+			yyerror("comparator-i;ascii-numeric require missing");
 			YYERROR; }
 				   else { $$->comparator = $3; } }
 	;
@@ -565,7 +565,7 @@
 			 yyerror("duplicate comparator tag"); YYERROR; }
 				   else if (!strcmp($3, "i;ascii-numeric") &&
 					    !parse_script->support.i_ascii_numeric) { 
-			 yyerror("comparator-i;ascii-numeric not required");  YYERROR; }
+			 yyerror("comparator-i;ascii-numeric require missing");  YYERROR; }
 				   else { 
 				     $$->comparator = $3; } }
         ;
@@ -606,7 +606,7 @@
 			 yyerror("duplicate comparator tag"); YYERROR; }
 				   else if (!strcmp($3, "i;ascii-numeric") &&
 					    !parse_script->support.i_ascii_numeric) { 
-			 yyerror("comparator-i;ascii-numeric not required");  YYERROR; }
+			 yyerror("comparator-i;ascii-numeric require missing");  YYERROR; }
 				   else { 
 				     $$->comparator = $3; } }
         ;
@@ -616,12 +616,12 @@
 	| LOCALPART		 { $$ = LOCALPART; }
 	| DOMAIN                 { $$ = DOMAIN; }
 	| USER                   { if (!parse_script->support.subaddress) {
-				     yyerror("subaddress not required");
+				     yyerror("subaddress require missing");
 				     YYERROR;
 				   }
 				   $$ = USER; }  
 	| DETAIL                { if (!parse_script->support.subaddress) {
-				     yyerror("subaddress not required");
+				     yyerror("subaddress require missing");
 				     YYERROR;
 				   }
 				   $$ = DETAIL; }
@@ -630,19 +630,19 @@
 	| CONTAINS		 { $$ = CONTAINS; }
 	| MATCHES		 { $$ = MATCHES; }
 	| REGEX			 { if (!parse_script->support.regex) {
-				     yyerror("regex not required");
+				     yyerror("regex require missing");
 				     YYERROR;
 				   }
 				   $$ = REGEX; }
 	;
 
 relcomp: COUNT			 { if (!parse_script->support.relational) {
-				     yyerror("relational not required");
+				     yyerror("relational require missing");
 				     YYERROR;
 				   }
 				   $$ = COUNT; }
 	| VALUE			 { if (!parse_script->support.relational) {
-				     yyerror("relational not required");
+				     yyerror("relational require missing");
 				     YYERROR;
 				   }
 				   $$ = VALUE; }
@@ -655,7 +655,7 @@
 
 copy: /* empty */		 { $$ = 0; }
 	| COPY			 { if (!parse_script->support.copy) {
-				     yyerror("copy not required");
+				     yyerror("copy require missing");
 	                             YYERROR;
                                    }
 				   $$ = COPY; }
