--- cyrus-imapd-2.3.12p2/imap/mailbox.c	2008-03-24 13:09:17.000000000 -0400
+++ cyrus-imapd-2.3.12p11/imap/mailbox.c	2008-07-11 14:18:35.000000000 -0400
@@ -2871,8 +2873,12 @@
 	    *mtail ='\0';
 	}
 
-	/* Check if parent mailbox exists */
 	*ntail = '\0';
+	if (mboxname_isdeletedmailbox(nbuf)) {
+	    /* don't check for a parent of delayed deleted mailboxes */
+	    break;
+	}
+	/* Check if parent mailbox exists */
 	ntail = strrchr(nbuf, '.');
 	if (!ntail || strchr(ntail, '!')) {
 	    /* Hit top of hierarchy or domain separator */
