*** ../cyrus-imapd-2.2.12.clean/lib/glob.c	2003-12-08 23:57:13 +0300
--- ./lib/glob.c	2005-10-10 19:42:04 +0400
***************
*** 284,312 ****
  		/* look for a match with first char following '%',
  		 * stop at a sep_char unless we're doing "*%"
  		 */
! 		ptr = phier;
! 		while (ptr != pend && *ghier != *ptr
! 		       && (*ptr != g->sep_char ||
  			   (!*ghier && gstar && *gstar == '%' && min
! 			    && ptr - start < *min))) {
! 		    ++ptr;
  		}
! 		if (ptr == pend) {
  		    gptr = ghier;
  		    break;
  		}
! 		if (*ptr == g->sep_char) {
  		    if (!*ghier && min
! 			&& *min < ptr - start && ptr != pend
! 			&& *ptr == g->sep_char
! 			) {
! 			*min = gstar ? ptr - start + 1 : -1;
! 			return (ptr - start);
  		    }
  		    ghier = NULL;
  		    sepfound = 1;
  		} else {
! 		    phier = ++ptr;
  		    gptr = ghier + 1;
  		}
  	    }
--- 285,311 ----
  		/* look for a match with first char following '%',
  		 * stop at a sep_char unless we're doing "*%"
  		 */
! 		while (phier != pend && *ghier != *phier
! 		       && (*phier != g->sep_char ||
  			   (!*ghier && gstar && *gstar == '%' && min
! 			    && phier - start < *min))) {
! 		    ++phier;
  		}
! 		if (phier == pend) {
  		    gptr = ghier;
+ 		    ptr = phier;
  		    break;
  		}
! 		if (*ghier != *phier && *phier == g->sep_char) {
  		    if (!*ghier && min
! 			&& *min < phier - start && phier != pend) {
! 			*min = gstar ? phier - start + 1 : -1;
! 			return (phier - start);
  		    }
  		    ghier = NULL;
  		    sepfound = 1;
  		} else {
! 		    ptr = ++phier;
  		    gptr = ghier + 1;
  		}
  	    }
***************
*** 376,404 ****
  		/* look for a match with first char following '%',
  		 * stop at a sep_char unless we're doing "*%"
  		 */
! 		ptr = phier;
! 		while (ptr != pend && (unsigned char) *ghier != TOLOWER(*ptr)
! 		       && (*ptr != g->sep_char ||
  			   (!*ghier && gstar && *gstar == '%' && min
! 			    && ptr - start < *min))) {
! 		    ++ptr;
  		}
! 		if (ptr == pend) {
  		    gptr = ghier;
  		    break;
  		}
! 		if (*ptr == g->sep_char) {
  		    if (!*ghier && min
! 			&& *min < ptr - start && ptr != pend
! 			&& *ptr == g->sep_char
! 			) {
! 			*min = gstar ? ptr - start + 1 : -1;
! 			return (ptr - start);
  		    }
  		    ghier = NULL;
  		    sepfound = 1;
  		} else {
! 		    phier = ++ptr;
  		    gptr = ghier + 1;
  		}
  	    }
--- 375,401 ----
  		/* look for a match with first char following '%',
  		 * stop at a sep_char unless we're doing "*%"
  		 */
! 		while (phier != pend && (unsigned char) *ghier != TOLOWER(*phier)
! 		       && (*phier != g->sep_char ||
  			   (!*ghier && gstar && *gstar == '%' && min
! 			    && phier - start < *min))) {
! 		    ++phier;
  		}
! 		if (phier == pend) {
  		    gptr = ghier;
+ 		    ptr = phier;
  		    break;
  		}
! 		if (*ghier != *phier && *phier == g->sep_char) {
  		    if (!*ghier && min
! 			&& *min < phier - start && phier != pend) {
! 			*min = gstar ? phier - start + 1 : -1;
! 			return (phier - start);
  		    }
  		    ghier = NULL;
  		    sepfound = 1;
  		} else {
! 		    ptr = ++phier;
  		    gptr = ghier + 1;
  		}
  	    }
