--- nntpd.c	2011-03-04 14:11:29.131749449 -0500
+++ nntpd.c.old	2010-12-20 08:15:49.000000000 -0500
@@ -1053,7 +1053,7 @@
 		be = backend_current;
 		if (arg1.len &&
 		    (!is_newsgroup(arg1.s) ||
-		     (r = open_group(arg1.s, 1, &be, NULL)))) goto nogroup;
+		     (r = open_group(arg1.s, 0, &be, NULL)))) goto nogroup;
 		else if (be) {
 		    prot_printf(be->out, "%s", cmd.s);
 		    if (arg1.len) {
@@ -2494,7 +2494,7 @@
 	}
     }
     else {
-	prot_printf(nntp_out, "%s %u %u %c\r\n", name,
+	prot_printf(nntp_out, "%s %u %u %c\r\n", name+strlen(newsprefix),
 		    group_state->exists ? index_getuid(group_state, group_state->exists) :
 		    group_state->mailbox->i.last_uid,
 		    group_state->exists ? index_getuid(group_state, 1) :
