--- cyrus-sasl.bak/plugins/gssapi.c	2011-06-03 01:59:43.000000000 -0500
+++ cyrus-sasl2-2.1.24~rc1.dfsg1+cvs2011-05-23/plugins/gssapi.c	2011-06-03 02:01:59.000000000 -0500
@@ -1477,10 +1477,10 @@
 	}
 
 	/* Setup req_flags properly */
-	req_flags = GSS_C_INTEG_FLAG;
+	req_flags = GSS_C_MUTUAL_FLAG | GSS_C_SEQUENCE_FLAG;
 	if (params->props.max_ssf > params->external_ssf) {
 	    /* We are requesting a security layer */
-	    req_flags |= GSS_C_MUTUAL_FLAG | GSS_C_SEQUENCE_FLAG;
+	    req_flags |= GSS_C_INTEG_FLAG;
 	    /* Any SSF bigger than 1 is confidentiality. */
 	    /* Let's check if the client of the API requires confidentiality,
 	       and it wasn't already provided by an external layer */
