*** sync_reset.c-	Tue Dec 13 10:31:10 2005
--- sync_reset.c	Thu Jul 27 14:11:30 2006
***************
*** 124,129 ****
--- 124,133 ----
      if (sync_authstate) auth_freestate(sync_authstate);
  
      seen_done();
+ 
+     quotadb_close();
+     quotadb_done();
+ 
      mboxlist_close();
      mboxlist_done();
      exit(code);
***************
*** 356,361 ****
--- 360,368 ----
      mboxlist_open(NULL);
      mailbox_initialize();
  
+     quotadb_init(0);
+     quotadb_open(NULL);
+ 
      signals_set_shutdown(&shut_down);
      signals_add_handlers(0);
  
