--- cyrus-imapd/imap/sync_client.c-DIST	2011-08-10 09:51:57.000000000 +0100
+++ cyrus-imapd/imap/sync_client.c	2011-08-10 09:52:48.000000000 +0100
@@ -2661,6 +2661,11 @@
     /* links to sockets */
     sync_in = sync_backend->in;
     sync_out = sync_backend->out;
+
+    /* Force use of LITERAL+ so we don't need two way communications */
+    prot_setisclient(sync_in, 1);
+    prot_setisclient(sync_out, 1);
+
 }
 
 static void replica_disconnect()
--- cyrus-imapd/imap/sync_server.c-DIST	2011-08-10 09:52:05.000000000 +0100
+++ cyrus-imapd/imap/sync_server.c	2011-08-10 09:53:19.000000000 +0100
@@ -349,6 +349,10 @@
     sync_in = prot_new(0, 0);
     sync_out = prot_new(1, 1);
 
+    /* Force use of LITERAL+ so we don't need two way communications */
+    prot_setisclient(sync_in, 1);
+    prot_setisclient(sync_out, 1);
+
     /* Find out name of client host */
     salen = sizeof(sync_remoteaddr);
     if (getpeername(0, (struct sockaddr *)&sync_remoteaddr, &salen) == 0 &&
