--- imap/proxy.c	6 Jan 2010 17:01:39 -0000	1.9
+++ imap/proxy.c	4 Aug 2010 19:11:23 -0000
@@ -158,6 +158,9 @@
 
     if (current && *current && !strcmp(server, (*current)->hostname)) {
 	/* this is our current backend */
+	if ((*current)->timeout) {
+	    (*current)->timeout->mark = time(NULL) + IDLE_TIMEOUT;
+	}
 	return *current;
     }
 
@@ -166,8 +169,10 @@
 	if (!strcmp(server, ((*cache)[i])->hostname)) {
 	    ret = (*cache)[i];
 	    /* ping/noop the server */
-	    if ((ret->sock != -1) && backend_ping(ret)) {
-		backend_disconnect(ret);
+	    if (backend_ping(ret)) {
+		proxy_downserver(ret);
+	    } else if (ret->timeout) {
+		ret->timeout->mark = time(NULL) + IDLE_TIMEOUT;
 	    }
 	    break;
 	}
@@ -185,8 +190,6 @@
 	    ret->timeout = prot_addwaitevent(clientin,
 					     time(NULL) + IDLE_TIMEOUT,
 					     backend_timeout, ret);
-
-	    ret->timeout->mark = time(NULL) + IDLE_TIMEOUT;
 	}
     }
 
