--- imap/lmtpd.c	2003/02/26 08:39:34	1.1
+++ imap/lmtpd.c	2003/02/26 10:21:08
@@ -914,6 +914,9 @@
 	fprintf(sm, "\r\nThis is a MIME-encapsulated message\r\n\r\n");
 	fprintf(sm, "--%d/%s\r\n", (int) p, config_servername);
     } else {
+	fprintf(sm, "MIME-Version: 1.0\r\n");
+	fprintf(sm, "Content-Type: text/plain; charset=utf-8\r\n");
+	fprintf(sm, "Content-Transfer-Encoding: 8bit\r\n");
 	fprintf(sm, "\r\n");
     }
 
