diff --git a/makedepend/def.h b/makedepend/def.h
index a3d4bec..efd8eb2 100644
--- a/makedepend/def.h
+++ b/makedepend/def.h
@@ -163,7 +163,7 @@ char			*realloc();
 
 char			*copy();
 char			*base_name();
-char			*getline();
+char			*get_line();
 struct symtab		*slookup();
 struct symtab		*isdefined();
 struct symtab		*fdefined();
diff --git a/makedepend/main.c b/makedepend/main.c
index 257f673..96e14e3 100644
--- a/makedepend/main.c
+++ b/makedepend/main.c
@@ -549,7 +549,7 @@ match(str, list)
  * Get the next line.  We only return lines beginning with '#' since that
  * is all this program is ever interested in.
  */
-char *getline(filep)
+char *get_line(filep)
 	register struct filepointer	*filep;
 {
 	register char	*p,	/* walking pointer */
diff --git a/makedepend/parse.c b/makedepend/parse.c
index f1a45e9..fd8e7b1 100644
--- a/makedepend/parse.c
+++ b/makedepend/parse.c
@@ -83,7 +83,7 @@ find_includes(filep, file, file_red, recursion, failOK)
 	register int	type;
 	boolean recfailOK;
 
-	while (line = getline(filep)) {
+	while (line = get_line(filep)) {
 		switch(type = deftype(line, filep, file_red, file, TRUE)) {
 		case IF:
 		doif:
@@ -205,7 +205,7 @@ gobble(filep, file, file_red)
 	register char	*line;
 	register int	type;
 
-	while (line = getline(filep)) {
+	while (line = get_line(filep)) {
 		switch(type = deftype(line, filep, file_red, file, FALSE)) {
 		case IF:
 		case IFFALSE:
