diff -urN cyrus-imapd-2.4.17/imap/mailbox.c cyrus-imapd-2.4.17.lockfix/imap/mailbox.c
--- cyrus-imapd-2.4.17/imap/mailbox.c	2012-12-01 20:57:54.000000000 +0100
+++ cyrus-imapd-2.4.17.lockfix/imap/mailbox.c	2014-02-25 15:38:11.000000000 +0100
@@ -845,6 +845,9 @@
     listitem = create_listitem(name);
     mailbox = &listitem->m;
 
+    r = mboxlist_lookup(name, &mbentry, NULL);
+    if (r) goto done;
+
     r = mboxname_lock(name, &listitem->l, locktype);
     if (r) {
 	/* locked is not an error - just means we asked for NONBLOCKING */
@@ -853,9 +856,6 @@
 	goto done;
     }
 
-    r = mboxlist_lookup(name, &mbentry, NULL);
-    if (r) goto done;
-
     if (mbentry.mbtype & MBTYPE_MOVING) {
 	r = IMAP_MAILBOX_MOVED;
 	goto done;
