--- imap/backend.c	22 Apr 2008 13:11:17 -0000	1.55
+++ imap/backend.c	19 Sep 2008 02:37:30 -0000
@@ -174,6 +174,65 @@
 #endif /* HAVE_SSL */
 }
 
+static char *intersect_mechlists( char *config, char *server )
+{
+    char *newmechlist = xzmalloc( strlen( config ) + 1 );
+    char *cmech = NULL, *smech = NULL, *s;
+    int count = 0;
+    char csave, ssave;
+
+    do {
+	if ( isalnum( *config ) || *config == '_' || *config == '-' ) {
+	    if ( cmech == NULL ) {
+		cmech = config;
+	    }
+	} else {
+	    if ( cmech != NULL ) {
+		csave = *config;
+		*config = '\0';
+
+		s = server;
+		do {
+		    if ( isalnum( *s ) || *s == '_' || *s == '-' ) {
+			if ( smech == NULL ) {
+			    smech = s;
+			}
+		    } else {
+			if ( smech != NULL ) {
+			    ssave = *s;
+			    *s = '\0';
+
+			    if ( strcasecmp( cmech, smech ) == 0 ) {
+				if ( count > 0 ) {
+				    strcat( newmechlist, " " );
+				}
+				strcat( newmechlist, cmech );
+				count++;
+
+				*s = ssave;
+				smech = NULL;
+				break;
+			    }
+
+			    *s = ssave;
+			    smech = NULL;
+			}
+		    }
+		} while ( *s++ );
+
+		*config = csave;
+		cmech = NULL;
+	    }
+	}
+    } while ( *config++ );
+
+    if ( count == 0 ) {
+	free( newmechlist );
+	return( NULL );
+    }
+    return( newmechlist );
+}
+
 static int backend_authenticate(struct backend *s, struct protocol_t *prot,
 				char **mechlist, const char *userid,
 				sasl_callback_t *cb, const char **status)
@@ -244,8 +303,15 @@
     do {
 	/* If we have a mech_conf, use it */
 	if (mech_conf) {
+	    char *newmechlist = intersect_mechlists( mech_conf, *mechlist );
+
+	    if ( newmechlist == NULL ) {
+		syslog( LOG_INFO, "%s did not offer %s", s->hostname,
+			mech_conf );
+	    }
+
 	    free(*mechlist);
-	    *mechlist = xstrdup(mech_conf);
+	    *mechlist = newmechlist;
 	}
 
 	if (*mechlist) {
