--- imapd.c.old	2011-09-14 10:36:51.000000000 +0200
+++ imapd.c	2011-09-14 10:40:31.000000000 +0200
@@ -3753,7 +3753,7 @@
 	prot_printf(backend_current->out, "%s %s {" SIZE_T_FMT "+}\r\n%s",
 		    tag, cmd, strlen(name), name);
 	if (v->uidvalidity) {
-	    prot_printf(backend_current->out, " (QRESYNC %lu " MODSEQ_FMT,
+	    prot_printf(backend_current->out, " (QRESYNC (%lu " MODSEQ_FMT,
 			v->uidvalidity, v->modseq);
 	    if (v->sequence) {
 		prot_printf(backend_current->out, " %s", v->sequence);
@@ -3762,7 +3762,7 @@
 		prot_printf(backend_current->out, " (%s %s)",
 			    v->match_seq, v->match_uid);
 	    }
-	    prot_printf(backend_current->out, ")");
+	    prot_printf(backend_current->out, "))");
 	}
 	prot_printf(backend_current->out, "\r\n");
 
