--- cyrus-imapd-2.3.15.orig/perl/imap/IMAP.pm	2008-04-04 14:47:08.000000000 +0200
+++ cyrus-imapd-2.3.15/perl/imap/IMAP.pm	2009-10-20 16:54:11.000000000 +0200
@@ -109,13 +109,14 @@
       if (ref($rest[0]) =~ /(^|=)HASH($|\()/) {
 	my %vals = %{shift(@rest)};
 	foreach (keys %vals) {
-	  $res .= $self->_stringize($_) . ' ' .
-	          $self->_stringize($vals{$_}) . $spc;
+	  $res .= $spc .
+	          $self->_stringize($_) . ' ' .
+	          $self->_stringize($vals{$_});
 	  $spc = ' ';
 	}
       } else {
 	foreach (@{shift(@rest)}) {
-	  $res .= $self->_stringize($_) . $spc;
+	  $res .= $spc . $self->_stringize($_);
 	  $spc = ' ';
 	}
       }
