diff -ru cyrus-imapd-2.4.12.orig/cmulocal/afs.m4 cyrus-imapd-2.4.12/cmulocal/afs.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/afs.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/afs.m4	2011-10-08 10:51:18.380556768 +0000
@@ -58,7 +58,7 @@
 AC_REQUIRE([CMU_SOCKETS])
 AC_REQUIRE([CMU_LIBSSL])
 AC_ARG_WITH(AFS,
-	[  --with-afs=PREFIX      Compile with AFS support],
+	[AS_HELP_STRING([--with-afs=PREFIX], [Compile with AFS support])],
 	[if test "X$with_AFS" = "X"; then
 		with_AFS=yes
 	fi])
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/agentx.m4 cyrus-imapd-2.4.12/cmulocal/agentx.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/agentx.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/agentx.m4	2011-10-08 14:45:55.307563688 +0000
@@ -9,7 +9,9 @@
 	dnl CMU AgentX
 	dnl
 	AC_MSG_CHECKING([for AgentX])
-	AC_ARG_WITH(agentx, [  --with-agentx              CMU AgentX libraries located in (val)], AGENTX_DIR="$withval", AGENTX_DIR=no)
+	AC_ARG_WITH(agentx,
+		[AS_HELP_STRING([--with-agentx], [CMU AgentX libraries located in (val)])], 
+		AGENTX_DIR="$withval", AGENTX_DIR=no)
 
 	found_agentx="no"
 
@@ -38,9 +40,7 @@
   	  AC_SUBST(LIB_AGENTX)
 	  AGENTXFLAGS="-I$AGENTX_DIR/include"
           AC_SUBST(AGENTXFLAGS)   
-	  AC_MSG_RESULT([found $AGENTX_DIR/lib${ABILIBDIR}/libagentx.a])	
+	  AC_MSG_RESULT([found $AGENTX_DIR/lib${ABILIBDIR}/libagentx.a])
 	fi
-
-
-
 ])
+
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/arx.m4 cyrus-imapd-2.4.12/cmulocal/arx.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/arx.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/arx.m4	2011-10-08 10:54:58.471556877 +0000
@@ -59,17 +59,17 @@
 AC_DEFUN([CMU_USE_ARX], [
 AC_REQUIRE([CMU_FIND_LIB_SUBDIR])
 AC_ARG_WITH(arx,
-	[  --with-arx=PREFIX      Compile with arx support],
+	[AS_HELP_STRING([--with-arx=PREFIX], [Compile with arx support])],
 	[if test "X$with_arx" = "X"; then
 		with_arx=yes
 	fi])
 AC_ARG_WITH(arx-lib,
-	[  --with-arx-lib=dir     use arx libraries in dir],
+	[AS_HELP_STRING([--with-arx-lib=dir], [use arx libraries in dir])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-arx-lib])
 	fi])
 AC_ARG_WITH(arx-include,
-	[  --with-arx-include=dir use arx headers in dir],
+	[AS_HELP_STRING([--with-arx-include=dir], [use arx headers in dir])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-arx-include])
 	fi])
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/clamav.m4 cyrus-imapd-2.4.12/cmulocal/clamav.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/clamav.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/clamav.m4	2011-10-08 14:49:56.358561668 +0000
@@ -5,7 +5,8 @@
 
 AC_DEFUN([CMU_CLAMAV], [
 AC_REQUIRE([CMU_FIND_LIB_SUBDIR])
-AC_ARG_WITH(clamav,[  --with-clamav=PATH	use ClamAV - PATH to clamav-config (yes)],
+AC_ARG_WITH(clamav,
+	[AS_HELP_STRING([--with-clamav=DIR], [use ClamAV - DIR to clamav-config (yes)])],
 	with_clamav=$withval, with_clamav=yes)
   have_clamav=no
   if test "$with_clamav" != no; then
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/com_err_link.m4 cyrus-imapd-2.4.12/cmulocal/com_err_link.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/com_err_link.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/com_err_link.m4	2011-10-08 14:51:47.217563877 +0000
@@ -62,17 +62,17 @@
 AC_DEFUN([CMU_USE_COMERR], [
 AC_REQUIRE([CMU_FIND_LIB_SUBDIR])
 AC_ARG_WITH(comerr,
-	[  --with-comerr=PREFIX      Compile with com_err support],
+	[AS_HELP_STRING([--with-comerr=PREFIX], [Compile with com_err support])],
 	[if test "X$with_comerr" = "X"; then
 		with_comerr=yes
 	fi])
 AC_ARG_WITH(comerr-lib,
-	[  --with-comerr-lib=dir     use com_err libraries in dir],
+	[AS_HELP_STRING([--with-comerr-lib=DIR], [use com_err libraries in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-comerr-lib])
 	fi])
 AC_ARG_WITH(comerr-include,
-	[  --with-comerr-include=dir use com_err headers in dir],
+	[AS_HELP_STRING([--with-comerr-include=DIR], [use com_err headers in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-comerr-include])
 	fi])
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/db.m4 cyrus-imapd-2.4.12/cmulocal/db.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/db.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/db.m4	2011-10-08 10:59:18.096557005 +0000
@@ -4,7 +4,7 @@
 dnl
 AC_DEFUN([rk_DB],[berkeley_db=db
 AC_ARG_WITH(berkeley-db,
-[  --without-berkeley-db   if you don't want berkeley db],[
+[AS_HELP_STRING([--without-berkeley-db], [if you don't want berkeley db])],[
 if test "$withval" = no; then
 	berkeley_db=""
 fi
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/heimdal.m4 cyrus-imapd-2.4.12/cmulocal/heimdal.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/heimdal.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/heimdal.m4	2011-10-08 14:53:26.646563925 +0000
@@ -102,17 +102,17 @@
 AC_REQUIRE([CMU_SOCKETS])
 AC_REQUIRE([CMU_USE_COMERR])
 AC_ARG_WITH(LIBHEIMDAL,
-	[  --with-libheimdal=PREFIX      Compile with Heimdal support],
+	[AS_HELP_STRING([--with-libheimdal=PREFIX], [Compile with Heimdal support])],
 	[if test "X$with_libheimdal" = "X"; then
 		with_libheimdal=yes
 	fi])
 AC_ARG_WITH(libheimdal-lib,
-	[  --with-libheimdal-lib=dir     use heimdal libraries in dir],
+	[AS_HELP_STRING([--with-libheimdal-lib=DIR], [use heimdal libraries in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-libheimdal-lib])
 	fi])
 AC_ARG_WITH(libheimdal-include,
-	[  --with-libheimdal-include=dir use heimdal headers in dir],
+	[AS_HELP_STRING([--with-libheimdal-include=DIR], [use heimdal headers in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-libheimdal-include])
 	fi])
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/kafs.m4 cyrus-imapd-2.4.12/cmulocal/kafs.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/kafs.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/kafs.m4	2011-10-08 14:54:34.061563960 +0000
@@ -75,17 +75,17 @@
 AC_REQUIRE([CMU_KRB4])
 AC_REQUIRE([CMU_KRB5])
 AC_ARG_WITH(kafs,
-	[  --with-kafs=PREFIX      Compile with Kafs support],
+	[AS_HELP_STRING([--with-kafs=PREFIX], [Compile with Kafs support])],
 	[if test "X$with_kafs" = "X"; then
 		with_kafs=yes
 	fi])
 AC_ARG_WITH(kafs-lib,
-	[  --with-kafs-lib=dir     use kafs libraries in dir],
+	[AS_HELP_STRING([--with-kafs-lib=DIR], [use kafs libraries in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-kafs-lib])
 	fi])
 AC_ARG_WITH(kafs-include,
-	[  --with-kafs-include=dir use kafs headers in dir],
+	[AS_HELP_STRING([--with-kafs-include=DIR], [use kafs headers in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-kafs-include])
 	fi])
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/kerberos_v4.m4 cyrus-imapd-2.4.12/cmulocal/kerberos_v4.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/kerberos_v4.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/kerberos_v4.m4	2011-10-08 14:56:00.980563802 +0000
@@ -147,17 +147,17 @@
 AC_REQUIRE([CMU_SOCKETS])
 AC_REQUIRE([CMU_LIBSSL])
 AC_ARG_WITH(krb4,
-	[  --with-krb4=PREFIX      Compile with Kerberos 4 support],
+	[AS_HELP_STRING([--with-krb4=PREFIX], [Compile with Kerberos 4 support])],
 	[if test "X$with_krb4" = "X"; then
 		with_krb4=yes
 	fi])
 AC_ARG_WITH(krb4-lib,
-	[  --with-krb4-lib=dir     use kerberos 4 libraries in dir],
+	[AS_HELP_STRING([--with-krb4-lib=DIR], [use kerberos 4 libraries in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-krb4-lib])
 	fi])
 AC_ARG_WITH(krb4-include,
-	[  --with-krb4-include=dir use kerberos 4 headers in dir],
+	[AS_HELP_STRING([--with-krb4-include=DIR], [use kerberos 4 headers in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-krb4-include])
 	fi])
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/kerberos_v5.m4 cyrus-imapd-2.4.12/cmulocal/kerberos_v5.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/kerberos_v5.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/kerberos_v5.m4	2011-10-08 15:02:17.670564590 +0000
@@ -62,23 +62,22 @@
 AC_REQUIRE([CMU_SOCKETS])
 AC_REQUIRE([CMU_USE_COMERR])
 AC_ARG_WITH(krb5,
-	[  --with-krb5=PREFIX      Compile with Kerberos 5 support],
+	[AS_HELP_STRING([--with-krb5=PREFIX], [Compile with Kerberos 5 support])],
 	[if test "X$with_krb5" = "X"; then
 		with_krb5=yes
 	fi])
 AC_ARG_WITH(krb5-lib,
-	[  --with-krb5-lib=dir     use kerberos 5 libraries in dir],
+	[AS_HELP_STRING([--with-krb5-lib=DIR], [use kerberos 5 libraries in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-krb5-lib])
 	fi])
 AC_ARG_WITH(krb5-include,
-	[  --with-krb5-include=dir use kerberos 5 headers in dir],
+	[AS_HELP_STRING([--with-krb5-include=DIR], [use kerberos 5 headers in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-krb5-include])
 	fi])
 AC_ARG_WITH(krb5-impl,
-	[  --with-krb5-impl=heimdal use heimdal kerberos 5 libraries
-  --with-krb5-impl=mit     use MIT kerberos 5 libraries],
+	[AS_HELP_STRING([--with-krb5-impl=heimdal/mit], [use heimdal/MIT kerberos 5 libraries])],
 	[if test "$withval" != "heimdal" -a "$withval" != "mit"; then
 		AC_MSG_ERROR([Invalid argument for --with-krb5-impl])
 	fi])
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/libcyrus.m4 cyrus-imapd-2.4.12/cmulocal/libcyrus.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/libcyrus.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/libcyrus.m4	2011-10-08 15:29:58.050564783 +0000
@@ -59,17 +59,17 @@
 AC_REQUIRE([CMU_SASL2])
 AC_REQUIRE([CMU_LIBSSL])
 AC_ARG_WITH(libcyrus,
-	[  --with-libcyrus=PREFIX      Compile with Libcyrus support],
+	[AS_HELP_STRING([--with-libcyrus=PREFIX], [Compile with Libcyrus support])],
 	[if test "X$with_libcyrus" = "X"; then
 		with_libcyrus=yes
 	fi])
 AC_ARG_WITH(libcyrus-lib,
-	[  --with-libcyrus-lib=dir     use libcyrus libraries in dir],
+	[AS_HELP_STRING([--with-libcyrus-lib=DIR], [use libcyrus libraries in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-libcyrus-lib])
 	fi])
 AC_ARG_WITH(libcyrus-include,
-	[  --with-libcyrus-include=dir use libcyrus headers in dir],
+	[AS_HELP_STRING([--with-libcyrus-include=DIR], [use libcyrus headers in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-libcyrus-include])
 	fi])
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/libloguse.m4 cyrus-imapd-2.4.12/cmulocal/libloguse.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/libloguse.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/libloguse.m4	2011-10-08 11:08:56.089557289 +0000
@@ -32,7 +32,7 @@
 AC_REQUIRE([CMU_FIND_LIB_SUBDIR])
 AC_REQUIRE([CMU_SOCKETS])
 AC_ARG_WITH(loguse,
-	[  --with-loguse=PREFIX      Compile with LOGUSE support],
+	[AS_HELP_STRING([--with-loguse=PREFIX], [Compile with LOGUSE support])],
 	[if test "X$with_loguse" = "X"; then
 		with_loguse=yes
 	fi])
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/libnet.m4 cyrus-imapd-2.4.12/cmulocal/libnet.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/libnet.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/libnet.m4	2011-10-08 15:30:54.916565156 +0000
@@ -80,22 +80,22 @@
 AC_DEFUN([CMU_LIBNET], [
 AC_REQUIRE([CMU_FIND_LIB_SUBDIR])
 AC_ARG_WITH(libnet,
-	[  --with-libnet=PREFIX      Compile with LIBNET support],
+	[AS_HELP_STRING([--with-libnet=PREFIX], [Compile with LIBNET support])],
 	[if test "X$with_libnet" = "X"; then
 		with_libnet=yes
 	fi])
 AC_ARG_WITH(libnet-config,
-	[  --with-libnet-config=dir  use libnet config program in dir],
+	[AS_HELP_STRING([--with-libnet-config=DIR], [use libnet config program in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-libnet-config])
 	fi])
 AC_ARG_WITH(libnet-lib,
-	[  --with-libnet-lib=dir     use libnet libraries in dir],
+	[AS_HELP_STRING([--with-libnet-lib=DIR], [use libnet libraries in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-libnet-lib])
 	fi])
 AC_ARG_WITH(libnet-include,
-	[  --with-libnet-include=dir use libnet headers in dir],
+	[AS_HELP_STRING([--with-libnet-include=DIR], [use libnet headers in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-libnet-include])
 	fi])
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/libpcap.m4 cyrus-imapd-2.4.12/cmulocal/libpcap.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/libpcap.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/libpcap.m4	2011-10-08 15:32:06.072565134 +0000
@@ -52,17 +52,17 @@
 AC_DEFUN([CMU_PCAP], [
 AC_REQUIRE([CMU_FIND_LIB_SUBDIR])
 AC_ARG_WITH(pcap,
-	[  --with-pcap=PREFIX      Compile with PCAP support],
+	[AS_HELP_STRING([--with-pcap=PREFIX], [Compile with PCAP support])],
 	[if test "X$with_pcap" = "X"; then
 		with_pcap=yes
 	fi])
 AC_ARG_WITH(pcap-lib,
-	[  --with-pcap-lib=dir     use pcap libraries in dir],
+	[AS_HELP_STRING([--with-pcap-lib=DIR], [use pcap libraries in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-pcap-lib])
 	fi])
 AC_ARG_WITH(pcap-include,
-	[  --with-pcap-include=dir use pcap headers in dir],
+	[AS_HELP_STRING([--with-pcap-include=DIR], [use pcap headers in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-pcap-include])
 	fi])
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/librestrict.m4 cyrus-imapd-2.4.12/cmulocal/librestrict.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/librestrict.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/librestrict.m4	2011-10-08 11:40:57.482558235 +0000
@@ -31,7 +31,7 @@
 AC_DEFUN([CMU_RESTRICT], [
 AC_REQUIRE([CMU_FIND_LIB_SUBDIR])
 AC_ARG_WITH(restrict,
-	[  --with-restrict=PREFIX      Compile with RESTRICT support],
+	[AS_HELP_STRING([--with-restrict=PREFIX], [Compile with RESTRICT support])],
 	[if test "X$with_restrict" = "X"; then
 		with_restrict=yes
 	fi])
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/libssl.m4 cyrus-imapd-2.4.12/cmulocal/libssl.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/libssl.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/libssl.m4	2011-10-08 11:41:55.714558533 +0000
@@ -57,7 +57,7 @@
 AC_REQUIRE([CMU_FIND_LIB_SUBDIR])
 AC_REQUIRE([CMU_SOCKETS])
 AC_ARG_WITH(libssl,
-	[  --with-libssl=PREFIX      Compile with Libssl support],
+	[AS_HELP_STRING([--with-libssl=PREFIX], [Compile with Libssl support])],
 	[if test "X$with_libssl" = "X"; then
 		with_libssl=yes
 	fi])
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/libwrap.m4 cyrus-imapd-2.4.12/cmulocal/libwrap.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/libwrap.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/libwrap.m4	2011-10-08 14:25:47.390561816 +0000
@@ -2,10 +2,10 @@
 dnl $Id: libwrap.m4,v 1.10 2005/04/26 19:14:08 shadow Exp $
 
 AC_DEFUN([CMU_LIBWRAP], [
-AC_REQUIRE([CMU_FIND_LIB_SUBDIR])
+  AC_REQUIRE([CMU_FIND_LIB_SUBDIR])
   AC_REQUIRE([CMU_SOCKETS])
   AC_ARG_WITH(libwrap, 
-              [  --with-libwrap=DIR      use libwrap (rooted in DIR) [yes] ],
+     [AS_HELP_STRING([--with-libwrap=DIR], [use libwrap (rooted in DIR) [yes] ])],
               with_libwrap=$withval, with_libwrap=yes)
   if test "$with_libwrap" != no; then
     if test -d "$with_libwrap"; then
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/libXau.m4 cyrus-imapd-2.4.12/cmulocal/libXau.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/libXau.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/libXau.m4	2011-10-08 15:33:52.089564847 +0000
@@ -57,17 +57,17 @@
 AC_REQUIRE([CMU_FIND_LIB_SUBDIR])
 AC_REQUIRE([CMU_SOCKETS])
 AC_ARG_WITH(Xau,
-	[  --with-Xau=PREFIX      Compile with Xau support],
+	[AS_HELP_STRING([--with-Xau=PREFIX], [Compile with Xau support])],
 	[if test "X$with_Xau" = "X"; then
 		with_Xau=yes
 	fi])
 AC_ARG_WITH(Xau-lib,
-	[  --with-Xau-lib=dir     use Xau libraries in dir],
+	[AS_HELP_STRING([--with-Xau-lib=DIR], [use Xau libraries in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-Xau-lib])
 	fi])
 AC_ARG_WITH(Xau-include,
-	[  --with-Xau-include=dir use Xau headers in dir],
+	[AS_HELP_STRING([--with-Xau-include=DIR], [use Xau headers in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-Xau-include])
 	fi])
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/mips-abi.m4 cyrus-imapd-2.4.12/cmulocal/mips-abi.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/mips-abi.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/mips-abi.m4	2011-10-08 15:35:36.461565959 +0000
@@ -6,7 +6,7 @@
 
 AC_DEFUN([AC_MIPS_ABI], [
 AC_ARG_WITH(mips_abi,
-[  --with-mips-abi=abi     ABI to use for IRIX (32, n32, or 64)])
+    [AS_HELP_STRING([--with-mips-abi=abi], [ABI to use for IRIX (32, n32, or 64)])])
 
 case "$host_os" in
 irix*)
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/nadine.m4 cyrus-imapd-2.4.12/cmulocal/nadine.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/nadine.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/nadine.m4	2011-10-08 15:36:19.049565087 +0000
@@ -62,17 +62,17 @@
 AC_DEFUN([CMU_NADINE], [
 AC_REQUIRE([CMU_SOCKETS])
 AC_ARG_WITH(nadine,
-	[  --with-nadine=PREFIX      Compile with nadine libevent support],
+	[AS_HELP_STRING([--with-nadine=PREFIX], [Compile with nadine libevent support])],
 	[if test "X$with_nadine" = "X"; then
 		with_nadine=yes
 	fi])
 AC_ARG_WITH(nadine-lib,
-	[  --with-nadine-lib=dir     use nadine libraries in dir],
+	[AS_HELP_STRING([--with-nadine-lib=DIR], [use nadine libraries in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-nadine-lib])
 	fi])
 AC_ARG_WITH(nadine-include,
-	[  --with-nadine-include=dir use nadine headers in dir],
+	[AS_HELP_STRING([--with-nadine-include=DIR], [use nadine headers in DIR])],
 	[if test "$withval" = "yes" -o "$withval" = "no"; then
 		AC_MSG_ERROR([No argument for --with-nadine-include])
 	fi])
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/nana.m4 cyrus-imapd-2.4.12/cmulocal/nana.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/nana.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/nana.m4	2011-10-08 14:23:38.439563101 +0000
@@ -4,7 +4,9 @@
 
 AC_DEFUN([CMU_NANA], [
   AC_REQUIRE([AC_PROG_CC])
-  AC_ARG_WITH(nana, [[  --with-nana             use NANA [yes] ]],,with_nana=yes)
+  AC_ARG_WITH(nana,
+	[AS_HELP_STRING([--with-nana], [use NANA [yes]])],,
+	with_nana=yes)
   if test "$GCC" != yes; then
     with_nana=no
   elif test "$with_nana" = yes; then
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/openssl.m4 cyrus-imapd-2.4.12/cmulocal/openssl.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/openssl.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/openssl.m4	2011-10-08 15:37:06.850565238 +0000
@@ -5,7 +5,8 @@
 
 AC_DEFUN([CMU_HAVE_OPENSSL], [
 AC_REQUIRE([CMU_FIND_LIB_SUBDIR])
-AC_ARG_WITH(openssl,[  --with-openssl=PATH     use OpenSSL from PATH],
+AC_ARG_WITH(openssl,
+	[AS_HELP_STRING([--with-openssl=DIR], [use OpenSSL from DIR])],
 	with_openssl=$withval, with_openssl="yes")
 
 	save_CPPFLAGS=$CPPFLAGS
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/sasl.m4 cyrus-imapd-2.4.12/cmulocal/sasl.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/sasl.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/sasl.m4	2011-10-08 15:01:42.032565372 +0000
@@ -48,7 +48,7 @@
 AC_DEFUN([CMU_SASL], [
 AC_REQUIRE([CMU_FIND_LIB_SUBDIR])
 AC_ARG_WITH(sasl,
-            [  --with-sasl=DIR        Compile with libsasl in <DIR>],
+            [AS_HELP_STRING([--with-sasl=DIR], [Compile with libsasl in DIR])],
 	    with_sasl="$withval",
             with_sasl="yes")
 
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/sql.m4 cyrus-imapd-2.4.12/cmulocal/sql.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/sql.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/sql.m4	2011-10-08 11:57:37.851558702 +0000
@@ -43,11 +43,11 @@
 AC_DEFUN([CYRUS_MYSQL_OPTS],
 [
 AC_ARG_WITH(mysql-libdir,
-	[  --with-mysql-libdir=DIR   MySQL lib files are in DIR],
+	[AS_HELP_STRING([--with-mysql-libdir=DIR], [MySQL lib files are in DIR])],
 	with_mysql_lib=$withval,
 	[ test "${with_mysql_lib+set}" = set || with_mysql_lib=none])
 AC_ARG_WITH(mysql-incdir,
-	[  --with-mysql-incdir=DIR   MySQL include files are in DIR],
+	[AS_HELP_STRING([--with-mysql-incdir=DIR], [MySQL include files are in DIR])],
 	with_mysql_inc=$withval,
 	[ test "${with_mysql_inc+set}" = set || with_mysql_inc=none ])
 ])
@@ -117,11 +117,11 @@
 AC_DEFUN([CYRUS_PGSQL_OPTS],
 [
 AC_ARG_WITH(pgsql-libdir,
-	[  --with-pgsql-libdir=DIR   Pgsql lib files are in DIR],
+	[AS_HELP_STRING([--with-pgsql-libdir=DIR], [Pgsql lib files are in DIR])],
 	with_pgsql_lib=$withval,
 	[ test "${with_pgsql_lib+set}" = set || with_pgsql_lib=none])
 AC_ARG_WITH(pgsql-incdir,
-	[  --with-pgsql-incdir=DIR   Pgsql include files are in DIR],
+	[AS_HELP_STRING([--with-pgsql-incdir=DIR], [Pgsql include files are in DIR])],
 	with_pgsql_inc=$withval,
 	[ test "${with_pgsql_inc+set}" = set || with_pgsql_inc=none ])
 ])
@@ -191,11 +191,11 @@
 AC_DEFUN([CYRUS_SQLITE_OPTS],
 [
 AC_ARG_WITH(sqlite-libdir,
-	[  --with-sqlite-libdir=DIR   SQLite lib files are in DIR],
+	[AS_HELP_STRING([--with-sqlite-libdir=DIR], [SQLite lib files are in DIR])],
 	with_sqlite_lib=$withval,
 	[ test "${with_sqlite_lib+set}" = set || with_sqlite_lib=none])
 AC_ARG_WITH(sqlite-incdir,
-	[  --with-sqlite-incdir=DIR   SQLite include files are in DIR],
+	[AS_HELP_STRING([--with-sqlite-incdir=DIR], [SQLite include files are in DIR])],
 	with_sqlite_inc=$withval,
 	[ test "${with_sqlite_inc+set}" = set || with_sqlite_inc=none ])
 ])
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/tcl.m4 cyrus-imapd-2.4.12/cmulocal/tcl.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/tcl.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/tcl.m4	2011-10-08 15:41:39.719565349 +0000
@@ -45,14 +45,15 @@
 dnl
 
 dnl Check for some information from the user on what the world looks like
-AC_ARG_WITH(tclconfig,[  --with-tclconfig=PATH   use tclConfig.sh from PATH
-                          (configure gets Tcl configuration from here)],
+AC_ARG_WITH(tclconfig,
+	[AS_HELP_STRING([--with-tclconfig=DIR], [use tclConfig.sh from DIR (configure gets Tcl configuration from here)])], 
         dnl trim tclConfig.sh off the end so we can add it back on later.
 	TclLibBase=`echo ${withval} | sed s/tclConfig.sh\$//`)
-AC_ARG_WITH(tcl,      [  --with-tcl=PATH         use Tcl from PATH],
+AC_ARG_WITH(tcl,
+        [AS_HELP_STRING([--with-tcl=DIR], [use Tcl from DIR])],
 	TclLibBase="${withval}/lib")
-AC_ARG_WITH(tclsh,    [  --with-tclsh=TCLSH      use TCLSH as the tclsh program
-                          (let configure find Tcl using this program)],
+AC_ARG_WITH(tclsh,
+        [AS_HELP_STRING([--with-tclsh=TCLSH], [use TCLSH as the tclsh program (let configure find Tcl using this program)])],
 	TCLSH="${withval}")
 
 if test "$TCLSH" = "no" -o "$with_tclconfig" = "no" ; then
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/ucdsnmp.m4 cyrus-imapd-2.4.12/cmulocal/ucdsnmp.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/ucdsnmp.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/ucdsnmp.m4	2011-10-08 14:27:36.670563146 +0000
@@ -5,7 +5,7 @@
 AC_REQUIRE([CMU_FIND_LIB_SUBDIR])
   AC_REQUIRE([CMU_SOCKETS])
   AC_ARG_WITH(snmp, 
-              [  --with-snmp=DIR         use ucd|net snmp (rooted in DIR) [yes] ],
+              [AS_HELP_STRING([--with-snmp=DIR], [use ucd|net snmp (rooted in DIR) [yes]])],
               with_snmp=$withval, with_snmp=yes)
 
   dnl
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/zephyr.m4 cyrus-imapd-2.4.12/cmulocal/zephyr.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/zephyr.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/zephyr.m4	2011-10-08 15:43:41.294565323 +0000
@@ -61,17 +61,17 @@
 AC_REQUIRE([CMU_SOCKETS])
 AC_REQUIRE([CMU_KRB4])
 AC_ARG_WITH(zephyr,
-        [  --with-zephyr=PREFIX      Compile with Zephyr support],
+        [AS_HELP_STRING([--with-zephyr=PREFIX],[Compile with Zephyr support])],
         [if test "X$with_zephyr" = "X"; then
                 with_zephyr=yes
         fi])
 AC_ARG_WITH(zephyr-lib,
-        [  --with-zephyr-lib=dir     use zephyr libraries in dir],
+        [AS_HELP_STRING([--with-zephyr-lib=DIR], [use zephyr libraries in DIR])],
         [if test "$withval" = "yes" -o "$withval" = "no"; then
                 AC_MSG_ERROR([No argument for --with-zephyr-lib])
         fi])
 AC_ARG_WITH(zephyr-include,
-        [  --with-zephyr-include=dir use zephyr headers in dir],
+        [AS_HELP_STRING([--with-zephyr-include=DIR], [use zephyr headers in DIR])],
         [if test "$withval" = "yes" -o "$withval" = "no"; then
                 AC_MSG_ERROR([No argument for --with-zephyr-include])
         fi])
diff -ru cyrus-imapd-2.4.12.orig/cmulocal/zlib.m4 cyrus-imapd-2.4.12/cmulocal/zlib.m4
--- cyrus-imapd-2.4.12.orig/cmulocal/zlib.m4	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/cmulocal/zlib.m4	2011-10-08 15:44:07.868565850 +0000
@@ -5,7 +5,7 @@
 
 AC_DEFUN([CMU_HAVE_ZLIB], [
 AC_REQUIRE([CMU_FIND_LIB_SUBDIR])
-AC_ARG_WITH(zlib,[  --with-zlib=PATH        use zlib from PATH],
+AC_ARG_WITH(zlib, [AS_HELP_STRING([--with-zlib=DIR],[use zlib from DIR])],
 	with_zlib=$withval, with_zlib="yes")
 
 	save_CPPFLAGS=$CPPFLAGS
diff -ru cyrus-imapd-2.4.12.orig/configure.in cyrus-imapd-2.4.12/configure.in
--- cyrus-imapd-2.4.12.orig/configure.in	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/configure.in	2011-10-08 16:01:46.040566014 +0000
@@ -46,15 +46,19 @@
 dnl
 
 
-AC_INIT(imap/imapd.c)
-AC_PREREQ([2.54])
+AC_INIT([cyrus-imapd], [2.4.12], [http://bugzilla.cyrusimap.org/index.cgi],, [http://www.cyrusimap.org/])
+AC_CONFIG_SRCDIR([imap/imapd.c])
+AC_PREREQ([2.68])
 AC_CONFIG_HEADER(config.h)
 AC_CANONICAL_SYSTEM
 
+AC_DEFINE_UNQUOTED([_CYRUS_VERSION], ["v$PACKAGE_VERSION"], [Exposes the version of cyrus-imap to config.h(.in)])
 dnl Useful hook for distributions
-AC_ARG_WITH(extraident,[  --with-extraident=STRING   use STRING as extra version information],
-	[AC_DEFINE_UNQUOTED(EXTRA_IDENT,"$withval", [Extra version information for imap/version.h])])
+AC_ARG_WITH(extraident,
+	[AS_HELP_STRING([--with-extraident=STRING], [use STRING as extra version information])],
+	[AC_DEFINE_UNQUOTED(EXTRA_IDENT,"$withval", [Extra version information for imap/version.c and notifyd/version.h])])
 
+AC_CONFIG_LIBOBJ_DIR([lib])
 AC_CHECK_PROG(MAKEDEPEND,makedepend,makedepend,[`cd ${srcdir};pwd`/tools/not-mkdep])
 if test "$MAKEDEPEND" != "makedepend"; then
         AC_MSG_WARN([Makedepend is not installed on this system.  You should compile and install the version from the makedepend subdirectory.])
@@ -63,19 +67,23 @@
 AC_ARG_WITH(login,,AC_ERROR([--with-login is no longer supported.
 Configure SASL appropriately instead.]))
 
-AC_ARG_WITH(cyrus-prefix,[  --with-cyrus-prefix=DIR    use DIR as cyrus server install directory],
-	cyrus_prefix="$withval",cyrus_prefix="/usr/cyrus")
+AC_ARG_WITH(cyrus-prefix,
+	[AS_HELP_STRING([--with-cyrus-prefix=DIR], [use DIR as cyrus server install directory [PREFIX/cyrus]])],
+	cyrus_prefix="$withval",cyrus_prefix="$prefix/cyrus")
 AC_SUBST(cyrus_prefix)
 AC_DEFINE_UNQUOTED(CYRUS_PATH,"$cyrus_prefix",[Where will we be installed?])
-AC_ARG_WITH(service-path,[  --with-service-path=DIR    use DIR as service install directory],
+AC_ARG_WITH(service-path,
+	[AS_HELP_STRING([--with-service-path=DIR], [use DIR as service install directory [PREFIX/cyrus/bin]])],
 	service_path="$withval",service_path="$cyrus_prefix/bin")
 AC_SUBST(service_path)
 AC_DEFINE_UNQUOTED(SERVICE_PATH,"$service_path",[Directory to use for service binaries])
-AC_ARG_WITH(cyrus-user,[  --with-cyrus-user=USERID   use USERID cyrus userid],
+AC_ARG_WITH(cyrus-user,
+	[AS_HELP_STRING([--with-cyrus-user=USERID], [use USERID cyrus userid])],
 	cyrus_user="$withval",cyrus_user="cyrus")
 AC_SUBST(cyrus_user)
 AC_DEFINE_UNQUOTED(CYRUS_USER, "$cyrus_user",[What user will we run as?])
-AC_ARG_WITH(cyrus-group,[  --with-cyrus-group=GROUPID use GROUPID cyrus group],
+AC_ARG_WITH(cyrus-group,
+	[AS_HELP_STRING([--with-cyrus-group=GROUPID], [use GROUPID cyrus group])],
 	cyrus_group="$withval",cyrus_group="mail")
 AC_SUBST(cyrus_group)
 
@@ -121,9 +129,9 @@
 dnl check for -R, etc. switch
 CMU_GUESS_RUNPATH_SWITCH
 
-AC_CHECK_HEADERS(unistd.h sys/select.h sys/param.h stdarg.h)
-AC_REPLACE_FUNCS(memmove strcasecmp ftruncate strerror)
-AC_CHECK_FUNCS(strlcat strlcpy getgrouplist)
+AC_CHECK_HEADERS(stdarg.h unistd.h sys/select.h sys/param.h)
+AC_REPLACE_FUNCS(ftruncate memmove strcasecmp strerror)
+AC_CHECK_FUNCS(getgrouplist strlcat strlcpy)
 AC_HEADER_DIRENT
 
 dnl check whether to use getpassphrase or getpass
@@ -178,7 +186,8 @@
 
 dnl Berkeley DB Detection
 
-AC_ARG_WITH(bdb, [  --with-bdb=DIR          use Berkeley DB (in DIR) [[yes]] ],
+AC_ARG_WITH(bdb,
+	[AS_HELP_STRING([--with-bdb=DIR], [use Berkeley DB (in DIR) [yes]])],
 	with_bdb=$withval, with_bdb="yes")
 
 dnl support old-style
@@ -193,7 +202,6 @@
 	with_bdb_lib=none
 	with_bdb_inc=none
 	;;
-  
   *)
 	use_berkeley="yes"
 	with_bdb_lib="$with_bdb/lib"
@@ -233,8 +241,9 @@
 HAVE_SQL=0
 
 dnl MySQL Detection
-AC_ARG_WITH(mysql,  [  --with-mysql=DIR          use MySQL (in DIR) [[no]] ],
-  with_mysql=$withval, with_mysql="no")
+AC_ARG_WITH(mysql,
+	[AS_HELP_STRING([--with-mysql=DIR], [use MySQL (in DIR) [no]])],
+	with_mysql=$withval, with_mysql="no")
 
 case "$with_mysql" in
   no)
@@ -245,7 +254,6 @@
 	with_mysql_lib=none
 	with_mysql_inc=none
 	;;
-  
   *)
 	use_mysql="yes"
 	with_mysql_lib="$with_mysql/lib"
@@ -277,8 +285,9 @@
 
 
 dnl PgSQL Detection
-AC_ARG_WITH(pgsql,  [  --with-pgsql=DIR          use PostgreSQL (in DIR) [[no]] ],
-  with_pgsql=$withval, with_pgsql="no")
+AC_ARG_WITH(pgsql,
+	[AS_HELP_STRING([--with-pgsql=DIR], [use PostgreSQL (in DIR) [no]])],
+	with_pgsql=$withval, with_pgsql="no")
 
 case "$with_pgsql" in
   no)
@@ -289,7 +298,6 @@
 	with_pgsql_lib=none
 	with_pgsql_inc=none
 	;;
-  
   *)
 	use_pgsql="yes"
 	with_pgsql_lib="$with_pgsql/lib"
@@ -321,8 +329,9 @@
 
 
 dnl SQLite Detection
-AC_ARG_WITH(sqlite,  [  --with-sqlite=DIR          use SQLite (in DIR) [[no]] ],
-  with_sqlite=$withval, with_sqlite=no)
+AC_ARG_WITH(sqlite,
+	[AS_HELP_STRING([--with-sqlite=DIR], [use SQLite (in DIR) [no]])],
+	with_sqlite=$withval, with_sqlite=no)
 
 case "$with_sqlite" in
   no)
@@ -333,7 +342,6 @@
 	with_sqlite_lib=none
 	with_sqlite_inc=none
 	;;
-  
   *)
 	use_sqlite="yes"
 	with_sqlite_lib="$with_sqlite/lib"
@@ -375,14 +383,14 @@
 SIEVE_SUBDIRS=""
 sievedir="sieve"
 AC_ARG_ENABLE(sieve,
-        [  --disable-sieve         disable Sieve support],
+        [AS_HELP_STRING([--disable-sieve], [disable Sieve support])],
 	if test "$enableval" = no; then
 		sievedir="no"
 	fi)
 
 testpcre="yes"
 AC_ARG_ENABLE(pcre,
-		[  --disable-pcre      disable PCRE library],
+	[AS_HELP_STRING([--disable-pcre], [disable PCRE library])],
 	if test "$enableval" = no; then
 		testpcre="no"
 	fi)
@@ -602,7 +610,7 @@
 
 AC_SUBST(WITH_MAP)
 AC_ARG_WITH(lock,
-  [  --with-lock=METHOD      force use of METHOD for locking (flock or fcntl)],
+  [AS_HELP_STRING([--with-lock=METHOD], [force use of METHOD for locking (flock or fcntl)])],
   WITH_LOCK="$withval", [
   AC_CHECK_FUNC(fcntl,WITH_LOCK="fcntl",[
 		AC_CHECK_FUNC(flock,WITH_LOCK="flock",[
@@ -646,19 +654,24 @@
 
 # ok, we still look for this stuff because of checking groups, but
 # all authentication goes through SASL
-AC_ARG_ENABLE([afs],[  --enable-afs            Enable AFS with ptloader])
+AC_ARG_ENABLE([afs],
+	[AS_HELP_STRING([--enable-afs], [Enable AFS with ptloader])])
 
-AC_ARG_WITH(afs-libdir,[  --with-afs-libdir=PATH  use AFS libraries from PATH [[/usr/lib]]],
-	    afs_libdir="${withval}", afs_libdir="/usr/lib")
+AC_ARG_WITH(afs-libdir,
+	[AS_HELP_STRING([--with-afs-libdir=DIR], [use AFS libraries from DIR [/usr/lib]])],
+	afs_libdir="${withval}", afs_libdir="/usr/lib")
  
-AC_ARG_WITH(afs-incdir,[  --with-afs-incdir=PATH  use AFS headers from PATH [[/usr/include]]],
-            afs_incdir="${withval}", afs_incdir="/usr/include")
-
-AC_ARG_WITH(ldap, [  --with-ldap=DIR         use LDAP (in DIR) (experimental) [/usr/local] ],
-	    with_ldap="${withval}", with_ldap="no")
+AC_ARG_WITH(afs-incdir,
+	[AS_HELP_STRING([--with-afs-incdir=DIR], [use AFS headers from DIR [/usr/include]])],
+	afs_incdir="${withval}", afs_incdir="/usr/include")
+
+AC_ARG_WITH(ldap,
+	[AS_HELP_STRING([--with-ldap=DIR], [use LDAP (in DIR) (experimental) [/usr/local]])],
+	with_ldap="${withval}", with_ldap="no")
   
 dnl select mode of afspts
-AC_ARG_ENABLE(krb5afspts,[  --enable-krb5afspts     compile afskrb PTS module with krb5 support])
+AC_ARG_ENABLE(krb5afspts,
+	[AS_HELP_STRING([--enable-krb5afspts], [compile afskrb PTS module with krb5 support])])
 
 if test "x$enable_afs" = "xyes"; then
 # krb5afspts is only valid if AFS is enabled, so check inside here.
@@ -767,7 +780,7 @@
 
 SERVER_SUBDIRS="master imap"
 AC_ARG_ENABLE(server,
-	[  --disable-server        disable compiling servers],
+	[AS_HELP_STRING([--disable-server], [disable compiling servers])],
 	if test "$enableval" = no; then
         	SERVER_SUBDIRS=""
 	fi)
@@ -776,14 +789,16 @@
 
 dnl this is the new simple check for kerberos; since the person had to
 dnl compile SASL, we might as well use the same checks.
-AC_ARG_WITH(krb,[  --with-krb=PATH         use Kerberos from PATH],
+AC_ARG_WITH(krb,
+	[AS_HELP_STRING([--with-krb=DIR], [use Kerberos from DIR])],
 	with_krb="$withval", with_krb="no")
 
-AC_ARG_WITH(krbimpl,[  --with-krbimpl=\[kth|mit\]         assume Kerberos 4 from KTH or MIT],
+AC_ARG_WITH(krbimpl,
+	[AS_HELP_STRING([--with-krbimpl=kth|mit], [assume Kerberos 4 from KTH or MIT])],
 	with_krbimpl="$withval", with_krbimpl="kth")
 
 AC_ARG_ENABLE(statickrb,
-		[  --enable-statickrb      link Kerberos statically],
+	[AS_HELP_STRING([--enable-statickrb], [link Kerberos statically])],
 	with_statickrb="yes", with_statickrb="no")
 
 dnl In order to compile kerberos4, we need libkrb and libdes.
@@ -797,8 +812,9 @@
 
 if test "$with_krb" != "no"; then
 dnl Do we need DES for kerberos?
-AC_ARG_WITH(krbdes,[  --with-krbdes           use Kerberos DES implementation [[yes]]],
-      with_krbdes="$withval", with_krbdes="yes")
+AC_ARG_WITH(krbdes,
+	[AS_HELP_STRING([--with-krbdes], [use Kerberos DES implementation [yes]])],
+	with_krbdes="$withval", with_krbdes="yes")
 if test "$with_krbdes" = "yes"; then
   AC_CHECK_LIB(des,des_ecb_encrypt,
       if test "$with_statickrb" = "yes"; then
@@ -863,7 +879,8 @@
 dnl Test for OpenSSL
 dnl
 IMAP_PROGS=""
-AC_ARG_WITH(openssl,[  --with-openssl=PATH     use OpenSSL from PATH],
+AC_ARG_WITH(openssl,
+	[AS_HELP_STRING([--with-openssl=DIR], [use OpenSSL from DIR])],
 	with_openssl="${withval}")
 
 OPENSSL_INC=
@@ -913,8 +930,7 @@
 dnl Allow for setting EGD socket file on systems without /dev/*random.
 dnl
 AC_ARG_WITH(egd-socket,
-      [  --with-egd-socket=FILE  Entropy Gathering Daemon socket pathname
-                          for systems without /dev/urandom],
+      [AC_HELP_STRING([--with-egd-socket=FILE], [Entropy Gathering Daemon socket pathname for systems without /dev/urandom])],
               [ EGD_SOCKET="$withval" ]
       )
 if test -n "$EGD_SOCKET" ; then
@@ -932,7 +948,8 @@
 dnl
 dnl Test for Zephyr
 dnl
-AC_ARG_WITH(zephyr,[  --with-zephyr[=PATH]    enable Zephyr notification (installed on PATH)],
+AC_ARG_WITH(zephyr,
+	[AS_HELP_STRING([--with-zephyr=DIR], [enable Zephyr notification (installed on DIR)])],
 	with_zephyr="${withval}")
 if test -z "$with_zephyr"; then
      if test -f /usr/local/lib/libzephyr.a; then
@@ -963,7 +980,8 @@
 dnl
 dnl Set pidfile location
 dnl
-AC_ARG_WITH(pidfile,[  --with-pidfile[=PATH]   pidfile in PATH (/var/run/cyrus-master.pid)],
+AC_ARG_WITH(pidfile,
+	[AS_HELP_STRING([--with-pidfile=DIR], [pidfile in DIR [/var/run/cyrus-master.pid]])],
 	[MASTERPIDFILE="$withval"],
 	[MASTERPIDFILE="/var/run/cyrus-master.pid"])
 MASTERPIDFILE="\"$MASTERPIDFILE\""
@@ -973,7 +991,7 @@
 dnl see if we're compiling with IMAP idled support
 dnl
 AC_ARG_ENABLE(idled,
-        [  --enable-idled          enable IMAP idled support],
+        [AS_HELP_STRING([--enable-idled], [enable IMAP idled support])],
         if test "$enable_val" != no; then
 	  IMAP_PROGS="$IMAP_PROGS idled"
         fi) 
@@ -983,7 +1001,7 @@
 dnl
 ENABLE_NNTP=no
 AC_ARG_ENABLE(nntp,
-        [  --enable-nntp           enable NNTP support],
+        [AS_HELP_STRING([--enable-nntp], [enable NNTP support])],
 	ENABLE_NNTP=$enableval
         if test "$ENABLE_NNTP" != no; then
 	  IMAP_PROGS="$IMAP_PROGS nntpd fetchnews"
@@ -994,7 +1012,7 @@
 dnl
 ENABLE_MURDER=no
 AC_ARG_ENABLE(murder,
-        [  --enable-murder         enable IMAP Murder support],
+        [AS_HELP_STRING([--enable-murder], [enable IMAP Murder support])],
 	ENABLE_MURDER=$enableval)
 
 if test "$ENABLE_MURDER" != no; then
@@ -1008,7 +1026,7 @@
 dnl
 ENABLE_REPLICATION=no
 AC_ARG_ENABLE(replication,
-        [  --enable-replication    enable replication support (experimental)],
+        [AS_HELP_STRING([--enable-replication], [enable replication support (experimental)])],
 	ENABLE_REPLICATION=$enableval
         if test "$ENABLE_REPLICATION" != no; then
 	  IMAP_PROGS="$IMAP_PROGS sync_client sync_server sync_reset"
@@ -1022,8 +1040,7 @@
 dnl deficient installs of com_err, unfortunately, which leave out compile_et)
 dnl
 AC_ARG_WITH(com_err,
-[  --with-com_err=PATH     use com_err from path -- includes in PATH/include,
-                          libs in PATH/lib, and compile_et in PATH/bin])
+	[AS_HELP_STRING([--with-com_err=DIR], [use com_err from path -- includes in DIR/include, libs in DIR/lib, and compile_et in DIR/bin])])
 if test -z "$with_com_err"; then
 	# no value supplied
 	AC_CHECK_LIB(com_err, com_err, [
@@ -1107,7 +1124,8 @@
 
 AC_MSG_CHECKING(which syslog facility to use)
 SYSLOG_FACILITY=LOG_LOCAL6
-AC_ARG_WITH(syslogfacility,[  --with-syslogfacility=FACILITY set the syslog facility to use (default LOCAL6)],
+AC_ARG_WITH(syslogfacility,
+	[AS_HELP_STRING([--with-syslogfacility=FACILITY], [set the syslog facility to use (default LOCAL6)])],
                [ if test "$withval" != "yes" -a "$withval" != "no" ; then
                        SYSLOG_FACILITY=LOG_$withval
                fi; ])
@@ -1117,7 +1135,7 @@
 dnl Have to check getdtabalesize after adding ossup, as some ossups define it
 AC_REPLACE_FUNCS(getdtablesize)
 AC_ARG_ENABLE(cmulocal,
-	[  --enable-cmulocal       enable CMU-specific local support],
+	[AS_HELP_STRING([--enable-cmulocal], [enable CMU-specific local support])],
 	if test "$enableval" = yes; then
 		EXTRA_SUBDIRS="${EXTRA_SUBDIRS} netnews depot"
 		EXTRA_OUTPUT="${EXTRA_OUTPUT} depot/Makefile"
@@ -1125,7 +1143,7 @@
 
 AC_MSG_CHECKING(to use old sieve service name)
 AC_ARG_ENABLE(oldsievename,
-	[  --enable-oldsievename   enable the use of 'imap' as the sieve service name],
+	[AS_HELP_STRING([--enable-oldsievename], [enable the use of 'imap' as the sieve service name])],
 	if test "$enableval" = yes; then
 		AC_MSG_RESULT(yes)
 		AC_DEFINE(OLD_SIEVE_SERVICE_NAME,[],[Use "imap" as sieve service name?])
@@ -1135,8 +1153,7 @@
 	AC_MSG_RESULT(no))
 
 AC_ARG_ENABLE(netscapehack,
-	[  --enable-netscapehack   enable Netscape hack for the menu option
-                          in Communicator to Administrate Mail],
+	[AS_HELP_STRING([--enable-netscapehack], [enable Netscape hack for the menu option in Communicator to Administrate Mail])],
 	if test "$enableval" = yes; then
 		AC_DEFINE(ENABLE_X_NETSCAPE_HACK,[],[Enable Netscape Menu Option Hack?])
 	fi)
@@ -1145,7 +1162,8 @@
 CMU_SASL2_REQUIRE_VER(2,1,7)
 CMU_SASL2_CHECKAPOP_REQUIRED
 
-AC_ARG_WITH(perl,       [  --with-perl=PERL        use PERL for perl],
+AC_ARG_WITH(perl,
+	[AS_HELP_STRING([--with-perl=PERL], [use PERL for perl])],
 	with_perl="$withval", with_perl="perl")
 
 if test "${with_perl}" = yes; then
@@ -1233,11 +1251,9 @@
   esac
 done
 
-IMAP_COM_ERR_LIBS="${COM_ERR_LIBS}"
 IMAP_LIBS="${LIB_SASL} ${LIBS} ${SQL_LIBS}"
 
 AC_SUBST(LIB_RT)
-AC_SUBST(IMAP_COM_ERR_LIBS)
 AC_SUBST(IMAP_LIBS)
 
 dnl AC_OUTPUT_COMMANDS([
@@ -1465,4 +1481,5 @@
 ])
 
 dnl make sure that Makefile is the last thing output
-AC_OUTPUT(man/Makefile master/Makefile lib/Makefile imap/Makefile imtest/Makefile netnews/Makefile perl/Makefile $EXTRA_OUTPUT Makefile)
+AC_CONFIG_FILES([man/Makefile master/Makefile lib/Makefile imap/Makefile imtest/Makefile netnews/Makefile perl/Makefile $EXTRA_OUTPUT Makefile])
+AC_OUTPUT
diff -ru cyrus-imapd-2.4.12.orig/contrib/drac_auth.patch cyrus-imapd-2.4.12/contrib/drac_auth.patch
--- cyrus-imapd-2.4.12.orig/contrib/drac_auth.patch	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/contrib/drac_auth.patch	2011-10-08 12:21:13.392559427 +0000
@@ -94,10 +94,10 @@
 +++ b/imap/Makefile.in
 @@ -66,6 +66,7 @@ SIEVE_OBJS = @SIEVE_OBJS@
  SIEVE_LIBS = @SIEVE_LIBS@
- IMAP_COM_ERR_LIBS = @IMAP_COM_ERR_LIBS@
+ COM_ERR_LIBS = @COM_ERR_LIBS@
  LIB_WRAP = @LIB_WRAP@
 +DRAC_LIBS = @DRACLIBS@
- LIBS = $(IMAP_LIBS) $(IMAP_COM_ERR_LIBS)
+ LIBS = $(IMAP_LIBS) $(COM_ERR_LIBS)
  DEPLIBS = ../lib/libcyrus.a ../lib/libcyrus_min.a @DEPLIBS@
  
 @@ -203,17 +204,17 @@ lmtpd.pure: lmtpd.o proxy.o $(LMTPOBJS) $(SIEVE_OBJS) \
diff -ru cyrus-imapd-2.4.12.orig/imap/Makefile.in cyrus-imapd-2.4.12/imap/Makefile.in
--- cyrus-imapd-2.4.12.orig/imap/Makefile.in	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/imap/Makefile.in	2011-10-08 12:21:13.427559424 +0000
@@ -63,9 +63,9 @@
 IMAP_LIBS = @IMAP_LIBS@ @LIB_RT@
 SIEVE_OBJS = @SIEVE_OBJS@
 SIEVE_LIBS = @SIEVE_LIBS@
-IMAP_COM_ERR_LIBS = @IMAP_COM_ERR_LIBS@
+COM_ERR_LIBS = @COM_ERR_LIBS@
 LIB_WRAP = @LIB_WRAP@
-LIBS = $(IMAP_LIBS) $(IMAP_COM_ERR_LIBS)
+LIBS = $(IMAP_LIBS) $(COM_ERR_LIBS)
 DEPLIBS = ../lib/libcyrus.a ../lib/libcyrus_min.a @DEPLIBS@
 
 CFLAGS = @CFLAGS@
@@ -152,7 +152,7 @@
 
 ### Built Source Files
 
-version.o: ../xversion.h
+version.o: ../xversion.h ../config.h
 
 pushstats.c: pushstats.snmp $(srcdir)/../snmp/snmpgen
 	$(srcdir)/../snmp/snmpgen $(srcdir)/pushstats.snmp
Only in cyrus-imapd-2.4.12/lib: imapopts.c
Only in cyrus-imapd-2.4.12/lib: imapopts.h
diff -ru cyrus-imapd-2.4.12.orig/Makefile.in cyrus-imapd-2.4.12/Makefile.in
--- cyrus-imapd-2.4.12.orig/Makefile.in	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/Makefile.in	2011-10-08 12:21:13.434559427 +0000
@@ -63,9 +63,9 @@
 LEX = @LEX@
 COMPILE_ET = @COMPILE_ET@
 
-PACKAGE = cyrus-imapd
-VERSION = 2.4.12
-GIT_VERSION = $(VERSION).git$(shell date +'%Y%m%d%H%M')
+PACKAGE_NAME = @PACKAGE_NAME@
+PACKAGE_VERSION = @PACKAGE_VERSION@
+GIT_VERSION = $(PACKAGE_VERSION).git$(shell date +'%Y%m%d%H%M')
 
 all:: xversion
 	@for d in  $(SUBDIRS); \
@@ -77,15 +77,14 @@
 snapshot::
 	@echo "creating snapshot $(GIT_VERSION)"
 	@if [ \! -d snapshot ]; then mkdir snapshot ; fi
-	@git archive --format=tar --prefix=$(PACKAGE)-$(GIT_VERSION)/ HEAD | tar -C snapshot/ -x -f -
-	@echo "/* $(PACKAGE) $(GIT_VERSION) */" > snapshot/$(PACKAGE)-$(GIT_VERSION)/xversion.h
-	@echo "#define _CYRUS_VERSION \"git$(VERSION)+`git log --pretty=oneline $(PACKAGE)-$(VERSION)..HEAD | wc -l`\"" >> snapshot/$(PACKAGE)-$(GIT_VERSION)/xversion.h
-	@echo "#define CYRUS_GITVERSION \"`git log --abbrev=8 --pretty="%h %ci" -1 | cut -c 1-18`\"" >> snapshot/$(PACKAGE)-$(GIT_VERSION)/xversion.h
-	@touch snapshot/$(PACKAGE)-$(GIT_VERSION)/xversion
-	@cd snapshot/$(PACKAGE)-$(GIT_VERSION) ; sh SMakefile
+	@git archive --format=tar --prefix=$(PACKAGE_NAME)-$(GIT_VERSION)/ HEAD | tar -C snapshot/ -x -f -
+	@echo "/* $(PACKAGE_NAME) $(GIT_VERSION) */" > snapshot/$(PACKAGE_NAME)-$(GIT_VERSION)/xversion.h
+	@echo "#define CYRUS_GITVERSION \"`git log --abbrev=8 --pretty="%h %ci" -1 | cut -c 1-18`\"" >> snapshot/$(PACKAGE_NAME)-$(GIT_VERSION)/xversion.h
+	@touch snapshot/$(PACKAGE_NAME)-$(GIT_VERSION)/xversion
+	@cd snapshot/$(PACKAGE_NAME)-$(GIT_VERSION) ; sh SMakefile
 	@for d in $(SUBDIRS) $(DISTSUBDIRS); \
 	do \
-	(cd snapshot/$(PACKAGE)-$(GIT_VERSION)/$$d; \
+	(cd snapshot/$(PACKAGE_NAME)-$(GIT_VERSION)/$$d; \
 	if [ -f Makefile.dist ]; then \
 	echo "### Making" snapshot "in" `pwd`; \
 	$(MAKE) $(MFLAGS) YACC='$(YACC)' LEX='$(LEX)' \
@@ -93,15 +92,15 @@
 	fi) || exit 1; \
 	done
 	@echo "creating tarfile"
-	tar -C snapshot -c -f - $(PACKAGE)-$(GIT_VERSION) | gzip -9 > $(PACKAGE)-$(GIT_VERSION).tar.gz
+	tar -C snapshot -c -f - $(PACKAGE_NAME)-$(GIT_VERSION) | gzip -9 > $(PACKAGE_NAME)-$(GIT_VERSION).tar.gz
 	@rm -rf snapshot
 
 dist:: distgit
 	@echo "creating configure"
-	@cd dist/cyrus-imapd-$(VERSION) ; sh SMakefile
+	@cd dist/cyrus-imapd-$(PACKAGE_VERSION) ; sh SMakefile
 	@for d in $(SUBDIRS) $(DISTSUBDIRS); \
 	do \
-	(cd dist/cyrus-imapd-$(VERSION)/$$d; \
+	(cd dist/cyrus-imapd-$(PACKAGE_VERSION)/$$d; \
 	if [ -f Makefile.dist ]; then \
 	echo "### Making" dist "in" `pwd`; \
 	$(MAKE) $(MFLAGS) YACC='$(YACC)' LEX='$(LEX)' \
@@ -109,24 +108,22 @@
 	fi) || exit 1; \
 	done
 	@echo "creating tarfile"
-	@cd dist ; tar cf - cyrus-imapd-$(VERSION) | \
-	gzip -9 > ../cyrus-imapd-$(VERSION).tar.gz
+	@cd dist ; tar cf - cyrus-imapd-$(PACKAGE_VERSION) | \
+	gzip -9 > ../cyrus-imapd-$(PACKAGE_VERSION).tar.gz
 	rm -f distgit
 	rm -rf dist
 
 distgit:
 	if [ \! -d dist ]; then mkdir dist ; fi
-	@echo "checking out the distribution from tag $(PACKAGE)-$(VERSION)"
-	git archive --format=tar --prefix=$(PACKAGE)-$(VERSION)/ $(PACKAGE)-$(VERSION) | tar -C dist -x -f -
-	echo "/* Release $(PACKAGE)-$(VERSION) */" > dist/$(PACKAGE)-$(VERSION)/xversion.h
-	echo "#define _CYRUS_VERSION \"v$(VERSION)\"" >> dist/$(PACKAGE)-$(VERSION)/xversion.h
-	echo "#define CYRUS_GITVERSION \"`git log --abbrev=8 --pretty=\"%h %ci\" -1 | cut -d\" \" -f1,2`\"" >> dist/$(PACKAGE)-$(VERSION)/xversion.h
-	touch dist/$(PACKAGE)-$(VERSION)/xversion
+	@echo "checking out the distribution from tag $(PACKAGE_NAME)-$(PACKAGE_VERSION)"
+	git archive --format=tar --prefix=$(PACKAGE_NAME)-$(PACKAGE_VERSION)/ $(PACKAGE_NAME)-$(PACKAGE_VERSION) | tar -C dist -x -f -
+	echo "/* Release $(PACKAGE_NAME)-$(PACKAGE_VERSION) */" > dist/$(PACKAGE_NAME)-$(PACKAGE_VERSION)/xversion.h
+	echo "#define CYRUS_GITVERSION \"`git log --abbrev=8 --pretty=\"%h %ci\" -1 | cut -d\" \" -f1,2`\"" >> dist/$(PACKAGE_NAME)-$(PACKAGE_VERSION)/xversion.h
+	touch dist/$(PACKAGE_NAME)-$(PACKAGE_VERSION)/xversion
 	touch distgit
 
 xversion xversion.h:
 	@echo "/* Development version */" > xversion.h.tmp
-	@echo "#define _CYRUS_VERSION \"git$(VERSION)+`git log --pretty=oneline $(PACKAGE)-$(VERSION)..HEAD | wc -l`\"" >> xversion.h.tmp
 	@echo "#define CYRUS_GITVERSION \"`git log --abbrev=8 --pretty=\"%h %ci\" -1 | cut -d\" \" -f1,2`\"" >> xversion.h.tmp
 	@cmp -s xversion.h.tmp xversion.h || mv xversion.h.tmp xversion.h
 	@rm -f xversion.h.tmp
Only in cyrus-imapd-2.4.12/man: imapd.conf.5
Only in cyrus-imapd-2.4.12/man: sieveshell.1
diff -ru cyrus-imapd-2.4.12.orig/master/cyrusMasterMIB.c cyrus-imapd-2.4.12/master/cyrusMasterMIB.c
--- cyrus-imapd-2.4.12.orig/master/cyrusMasterMIB.c	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/master/cyrusMasterMIB.c	2011-10-08 12:21:13.440559427 +0000
@@ -1,3 +1,4 @@
+/* This should always be included first before anything else */
 #include <config.h>
 
 #if defined(HAVE_UCDSNMP) || defined(HAVE_NETSNMP)
@@ -5,9 +6,6 @@
 /* This file was generated by mib2c and is intended for use as a mib module
    for the ucd-snmp snmpd agent. */
 
-/* This should always be included first before anything else */
-#include <config.h>
-
 #include "xmalloc.h"
 #include "xstrlcpy.h"
 #include "xstrlcat.h"
diff -ru cyrus-imapd-2.4.12.orig/master/Makefile.in cyrus-imapd-2.4.12/master/Makefile.in
--- cyrus-imapd-2.4.12.orig/master/Makefile.in	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/master/Makefile.in	2011-10-08 13:20:19.923561173 +0000
@@ -53,12 +53,12 @@
 CYRUS_GROUP=@cyrus_group@
 
 DEFS = @DEFS@ @LOCALDEFS@
-CPPFLAGS = -I.. -I$(srcdir)/../lib -I$(srcdir)/../imap @CPPFLAGS@ @COM_ERR_CPPFLAGS@
+CPPFLAGS = -I.. -I$(srcdir)/../lib -I$(srcdir)/../imap @CPPFLAGS@
 DEPLIBS = @DEPLIBS@
 
 CFLAGS = @CFLAGS@
-LDFLAGS = @LDFLAGS@ @COM_ERR_LDFLAGS@
-LIBS = ../lib/libcyrus_min.a @LIB_UCDSNMP@ @LIBS@ @COM_ERR_LIBS@
+LDFLAGS = @LDFLAGS@
+LIBS = ../lib/libcyrus_min.a @LIB_UCDSNMP@ @LIBS@
 
 SHELL = /bin/sh
 MAKEDEPEND = @MAKEDEPEND@
diff -ru cyrus-imapd-2.4.12.orig/master/masterconf.c cyrus-imapd-2.4.12/master/masterconf.c
--- cyrus-imapd-2.4.12.orig/master/masterconf.c	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/master/masterconf.c	2011-10-08 12:21:13.446559427 +0000
@@ -54,7 +54,8 @@
 #include <sys/stat.h>
 #include <sysexits.h>
 
-#include "libconfig.c"
+#include "lib/util.h"
+#include "lib/libconfig.h"
 
 #if HAVE_UNISTD_H
 # include <unistd.h>
diff -ru cyrus-imapd-2.4.12.orig/notifyd/Makefile.in cyrus-imapd-2.4.12/notifyd/Makefile.in
--- cyrus-imapd-2.4.12.orig/notifyd/Makefile.in	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/notifyd/Makefile.in	2011-10-08 12:21:13.452559427 +0000
@@ -71,9 +71,9 @@
 SERVICE=../master/service.o
 
 IMAP_LIBS = @IMAP_LIBS@ @LIB_RT@
-IMAP_COM_ERR_LIBS = @IMAP_COM_ERR_LIBS@
+COM_ERR_LIBS = @COM_ERR_LIBS@
 LIB_WRAP = @LIB_WRAP@
-LIBS = @ZEPHYR_LIBS@ @LIBS@ $(IMAP_COM_ERR_LIBS)
+LIBS = @ZEPHYR_LIBS@ @LIBS@ $(COM_ERR_LIBS)
 DEPLIBS=../imap/mutex_fake.o ../imap/libimap.a ../lib/libcyrus.a ../lib/libcyrus_min.a @DEPLIBS@
 
 PURIFY=/usr/local/bin/purify
diff -ru cyrus-imapd-2.4.12.orig/sieve/Makefile.in cyrus-imapd-2.4.12/sieve/Makefile.in
--- cyrus-imapd-2.4.12.orig/sieve/Makefile.in	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/sieve/Makefile.in	2011-10-08 12:21:13.459559427 +0000
@@ -66,9 +66,9 @@
 MAKEDEPEND = @MAKEDEPEND@
 
 IMAP_LIBS = @IMAP_LIBS@
-IMAP_COM_ERR_LIBS = @IMAP_COM_ERR_LIBS@
+COM_ERR_LIBS = @COM_ERR_LIBS@
 IMAP_LIBS = @IMAP_LIBS@ @LIB_RT@
-LIBS = libsieve.a $(IMAP_LIBS) $(IMAP_COM_ERR_LIBS)
+LIBS = libsieve.a $(IMAP_LIBS) $(COM_ERR_LIBS)
 DEPLIBS=../lib/libcyrus.a ../lib/libcyrus_min.a @DEPLIBS@
 
 BUILT_SOURCES = addr.c addr.h addr-lex.c \
diff -ru cyrus-imapd-2.4.12.orig/timsieved/Makefile.in cyrus-imapd-2.4.12/timsieved/Makefile.in
--- cyrus-imapd-2.4.12.orig/timsieved/Makefile.in	2011-10-04 19:53:03.000000000 +0000
+++ cyrus-imapd-2.4.12/timsieved/Makefile.in	2011-10-08 12:21:13.464559427 +0000
@@ -71,9 +71,9 @@
 SERVICE=../master/service.o
 
 IMAP_LIBS = @IMAP_LIBS@ @LIB_RT@
-IMAP_COM_ERR_LIBS = @IMAP_COM_ERR_LIBS@
+COM_ERR_LIBS = @COM_ERR_LIBS@
 LIB_WRAP = @LIB_WRAP@
-LIBS = $(IMAP_COM_ERR_LIBS)
+LIBS = $(COM_ERR_LIBS)
 DEPLIBS=../sieve/libsieve.a ../imap/mutex_fake.o ../imap/libimap.a ../lib/libcyrus.a ../lib/libcyrus_min.a @DEPLIBS@
 
 PURIFY=/usr/local/bin/purify
