--- install-admin-mb.html	2006-11-30 12:11:16.000000000 -0500
+++ install-admin-mb.html.fixed	2008-01-03 23:40:50.000000000 -0500
@@ -36,9 +36,9 @@
 <p><b>Note:</b>If you run "<tt>cyradm</tt>" on a system not using
 Kerberos for authentication, you will be prompted for your user name
 and password before you can issue any "<tt>cyradm</tt>" commands.  Use
-the "<tt>-u</tt>" option to specify a particular user.
+both the "<tt>-u</tt>" and "<tt>-a</tt>" options to specify a particular user.
 
-<p>The mailbox naming convention requires that the primary mailbox (inbox)
+<p>The mailbox naming convention requires that the primary mailbox (INBOX)
 for anyone must be named "<tt>user.<var>&lt;userid&gt;</var></tt>".
 To create a mailbox, type:
 
@@ -58,7 +58,7 @@
 <kbd>   setquota user.smith 10000
 </kbd></pre>
 
-Once the inbox is created, users can create their own additional mailboxes 
+Once the INBOX is created, users can create their own additional mailboxes 
 from a mail program.  If Smith created a work mailbox and a play mailbox, 
 the full names of the mailboxes would be:
 
@@ -85,5 +85,3 @@
 <P><HR>
 last modified: $Date: 2006/11/30 17:11:16 $
 </BODY></HTML>
-
-
