--- cyrus-imapd-2.4.6/imap/lmtpengine.c.orig	2010-12-20 14:15:49.000000000 +0100
+++ cyrus-imapd-2.4.6/imap/lmtpengine.c	2011-02-17 13:27:49.000000000 +0100
@@ -462,6 +462,7 @@
 {
     char *p, *ret;
     int len;
+    int munge8bit = config_getswitch(IMAPOPT_MUNGE8BIT);
 
     p = s;
 
@@ -500,6 +501,12 @@
 		if (!*++p) return 0;
 	    }
 	    else {
+		if (*p & 128 && munge8bit) {
+		    /* this prevents us from becoming a backscatter
+		       source because popular MTAs allow 8bit in
+		       local-part of adresses. */
+		    *p = 'X';
+		}
 		if (*p <= ' ' || (*p & 128) ||
 		    strchr("<>()[]\\,;:\"", *p)) return 0;
 	    }
