--- sync_client.c.orig	Mon Feb 13 11:57:42 2006
+++ sync_client.c	Mon Feb 13 11:58:07 2006
@@ -672,6 +672,11 @@
 
 static int user_addsub(char *user, char *name)
 {
+    char *p;
+
+    if ((p = strchr(name, '!')) && strchr(p + 1, '!'))
+	name = p + 1;
+
     if (verbose) 
         printf("ADDSUB %s %s\n", user, name);
 
@@ -690,6 +695,11 @@
 
 static int user_delsub(char *user, char *name)
 {
+    char *p;
+
+    if ((p = strchr(name, '!')) && strchr(p + 1, '!'))
+	name = p + 1;
+
     if (verbose) 
         printf("DELSUB %s %s\n", user, name);
 
