diff -Naur cyrus-sasl-2.1.17-old/saslauthd/lak.c cyrus-sasl-2.1.17-new/saslauthd/lak.c
--- cyrus-sasl-2.1.17-old/saslauthd/lak.c	2003-10-30 09:21:47.000000000 -0800
+++ cyrus-sasl-2.1.17-new/saslauthd/lak.c	2004-01-13 09:07:11.863220960 -0800
@@ -1264,7 +1264,7 @@
 	free(group);
 	if (rc != LDAP_COMPARE_TRUE) {
 		syslog(LOG_WARNING|LOG_AUTH, "ldap_compare_s() for group failed.");
-		rc = LAK_FAIL;
+		return LAK_FAIL;
 	}
 
 	return LAK_OK;
