diff -r -u cyrus-imapd-2.3.13.orig/imap/append.c cyrus-imapd-2.3.13/imap/append.c
--- cyrus-imapd-2.3.13.orig/imap/append.c	2008-09-23 18:19:42.000000000 +0200
+++ cyrus-imapd-2.3.13/imap/append.c	2008-12-18 11:06:49.000000000 +0100
@@ -217,7 +217,6 @@
 	    as->m.quota.used + quotacheck > 
 	    ((uquota_t) as->m.quota.limit * QUOTA_UNITS)) {
 	    quota_abort(&as->tid);
-	    mailbox_close(&as->m);
 	    r = IMAP_QUOTA_EXCEEDED;
 	}
     }
