--- ldap.c.orig	2007-06-22 10:05:56.000000000 -0700
+++ ldap.c	2007-06-22 10:07:04.000000000 -0700
@@ -1066,9 +1066,9 @@
 
 	    int j;
 	    strcpy((*newstate)->groups[i].id, "group:");
-	    for(j =0; j < strlen(vals[i]); j++) {
-	      if(isupper(vals[i][j]))
-		vals[i][j]=tolower(vals[i][j]);
+	    for(j =0; j < strlen(vals[0]); j++) {
+	      if(isupper(vals[0][j]))
+		vals[0][j]=tolower(vals[0][j]);
 	    }
 
             strlcat((*newstate)->groups[i].id, vals[0], 
