diff -urN cyrus-imapd-2.4.17/imap/tls.c cyrus-imapd-2.4.17.clientcerts/imap/tls.c
--- cyrus-imapd-2.4.17/imap/tls.c	2012-12-01 20:57:54.000000000 +0100
+++ cyrus-imapd-2.4.17.clientcerts/imap/tls.c	2013-12-09 18:23:27.000000000 +0100
@@ -729,7 +729,8 @@
     if ((!SSL_CTX_load_verify_locations(s_ctx, CAfile, CApath)) ||
 	(!SSL_CTX_set_default_verify_paths(s_ctx))) {
 	/* just a warning since this is only necessary for client auth */
-	syslog(LOG_NOTICE,"TLS server engine: cannot load CA data");	
+	syslog(LOG_NOTICE,"TLS server engine: cannot load CA data. disabling client certs");
+        askcert = 0;
     }
 
     s_cert_file = config_getstring(IMAPOPT_TLS_CERT_FILE);
