diff -uNr cyrus-imapd-2.3.16.orig/imap/imap_proxy.c cyrus-imapd-2.3.16/imap/imap_proxy.c
--- cyrus-imapd-2.3.16.orig/imap/imap_proxy.c	2009-12-02 04:09:43.000000000 +0200
+++ cyrus-imapd-2.3.16/imap/imap_proxy.c	2010-05-21 11:01:10.000000000 +0300
@@ -1457,11 +1457,15 @@
 			continue;
 		    }
 
+		    if (!isdigit(c)) {
+			fatal("Invalid response from backend", EC_PROTOCOL);
+		    }
+
 		    /* read uidvalidity */
-		    while (isdigit(c = prot_getc(be->in))) {
+		    do {
 			avail *= 10;
 			avail += c - '0';
-		    }
+		    } while (isdigit(c = prot_getc(be->in)));
 		    if (c != '\"') { c = EOF; break; }
 		    eatline(be->in, c); /* we don't care about the rest of the line */
 		}
