diff -C 3 -r ../cyrus-imapd-2.2.12.clean/imap/mboxname.c ./imap/mboxname.c
*** ../cyrus-imapd-2.2.12.clean/imap/mboxname.c	2005-02-14 09:39:57 +0300
--- ./imap/mboxname.c	2005-10-10 19:52:47 +0400
***************
*** 96,101 ****
--- 96,108 ----
  #define CHARMOD64(c)  (index_mod64[(unsigned char)(c)])
  
  
+ static int mboxname_check_hierseps(struct namespace *namespace, const char *name, int namelen)
+ {
+     if(namespace->hier_sep == '/' && memchr(name, DOTCHAR, namelen))
+ 	return -1;
+     return 0;
+ }
+ 
  /*
   * Convert the external mailbox 'name' to an internal name.
   * If 'userid' is non-null, it is the name of the current user.
***************
*** 116,121 ****
--- 123,131 ----
      userlen = userid ? strlen(userid) : 0;
      namelen = strlen(name);
  
+     if(mboxname_check_hierseps(namespace, name, namelen))
+ 	return IMAP_MAILBOX_BADNAME;
+ 
      if (config_virtdomains) {
  	if (userid && (cp = strrchr(userid, '@'))) {
  	    /* user logged in as user@domain */
***************
*** 206,211 ****
--- 216,224 ----
      userlen = userid ? strlen(userid) : 0;
      namelen = strlen(name);
  
+     if(mboxname_check_hierseps(namespace, name, namelen))
+ 	return IMAP_MAILBOX_BADNAME;
+ 
      if (config_virtdomains) {
  	if (userid && (cp = strchr(userid, '@'))) {
  	    /* user logged in as user@domain */
