--- imap/imapd.c	8 Oct 2008 15:47:06 -0000	1.553
+++ imap/imapd.c	2 Dec 2008 02:32:15 -0000
@@ -8647,6 +8687,14 @@
 	    r = mboxlist_findall(NULL, buf, 1, imapd_userid,
 				 imapd_authstate, xfer_user_cb,
 				 &rock);
+
+	    if ( !r && mboxlist_delayed_delete_isenabled()) {
+		snprintf(buf, sizeof(buf), "%s.%s.*",
+			config_getstring(IMAPOPT_DELETEDPREFIX), mailboxname);
+		r = mboxlist_findall(NULL, buf, 1, imapd_userid,
+				     imapd_authstate, xfer_user_cb,
+				     &rock);
+	    }
 	}
 
 	/* xxx how do you back out if one of the above moves fails? */
