diff -r -u cyrus-imapd-2.3.16.orig/imap/mbdump.c cyrus-imapd-2.3.16/imap/mbdump.c
--- cyrus-imapd-2.3.16.orig/imap/mbdump.c	2009-08-28 15:53:25.000000000 +0200
+++ cyrus-imapd-2.3.16/imap/mbdump.c	2010-01-13 17:11:44.000000000 +0100
@@ -479,6 +479,7 @@
     struct buf file, data;
     char c;
     uquota_t quotaused = 0;
+    int quotalimit = -1;
     int r = 0;
     int curfile = -1;
     const char *userid = NULL;
@@ -534,9 +535,9 @@
     if(!strcmp(data.s, "NIL")) {
 	/* Remove any existing quotaroot */
 	mboxlist_unsetquota(mbname);
-    } else if(imparse_isnumber(data.s)) {
+    } else if(sscanf(data.s, "%d", &quotalimit) == 1) {
 	/* Set a Quota */ 
-	mboxlist_setquota(mbname, atoi(data.s), 0);
+	mboxlist_setquota(mbname, quotalimit, 0);
     } else {
 	/* Huh? */
 	freebuf(&data);
