diff -ruN cyrus-sasl-2.1.24.orig/plugins/ldapdb.c cyrus-sasl-2.1.24/plugins/ldapdb.c
--- cyrus-sasl-2.1.24.orig/plugins/ldapdb.c	2009-05-07 10:21:15.000000000 -0500
+++ cyrus-sasl-2.1.24/plugins/ldapdb.c	2010-01-10 01:06:11.000000000 -0600
@@ -249,7 +249,6 @@
 	    result = SASL_UNAVAIL;
 	    break;
 
-	case LDAP_PROXY_AUTHZ_FAILURE:
 	case LDAP_INAPPROPRIATE_AUTH:
 	case LDAP_INVALID_CREDENTIALS:
 	case LDAP_INSUFFICIENT_ACCESS:
diff -ruN cyrus-sasl-2.1.24.orig/plugins/Makefile.am cyrus-sasl-2.1.24/plugins/Makefile.am
--- cyrus-sasl-2.1.24.orig/plugins/Makefile.am	2009-08-13 08:37:30.000000000 -0500
+++ cyrus-sasl-2.1.24/plugins/Makefile.am	2010-01-10 01:06:11.000000000 -0600
@@ -119,7 +119,7 @@
 
 libldapdb_la_SOURCES = ldapdb.c ldapdb_init.c $(common_sources)
 libldapdb_la_DEPENDENCIES = $(COMPAT_OBJS)
-libldapdb_la_LIBADD = $(COMPAT_OBJS)
+libldapdb_la_LIBADD = $(LIB_LDAP) $(COMPAT_OBJS)
 
 libsql_la_SOURCES = sql.c sql_init.c $(common_sources)
 libsql_la_LDFLAGS = $(LIB_MYSQL) $(LIB_PGSQL) $(LIB_SQLITE) $(AM_LDFLAGS)
diff -ruN cyrus-sasl-2.1.24.orig/plugins/makeinit.sh cyrus-sasl-2.1.24/plugins/makeinit.sh
--- cyrus-sasl-2.1.24.orig/plugins/makeinit.sh	2009-05-07 10:21:15.000000000 -0500
+++ cyrus-sasl-2.1.24/plugins/makeinit.sh	2010-01-10 01:06:49.000000000 -0600
@@ -90,4 +90,4 @@
 done
 
 # ldapdb is also a canon_user plugin
-echo "SASL_CANONUSER_PLUG_INIT( ldap )" >> ldapdb_init.c
+echo "SASL_CANONUSER_PLUG_INIT( ldapdb )" >> ldapdb_init.c
