From 63a3704d24af7f14d14805462ef6e39a078232e4 Mon Sep 17 00:00:00 2001
From: "Jeroen van Meeuwen (Kolab Systems)" <vanmeeuwen@kolabsys.com>
Date: Sun, 15 Apr 2012 14:21:15 +0100
Subject: [PATCH 4/4] Correct imap/cli_fatal.o -> imap/cli_fatal.c (#3676)

---
 Makefile.am |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index 263bc56..8fd1367 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -532,9 +532,9 @@ perl/sieve/managesieve/Makefile: perl/sieve/managesieve/Makefile.PL
 			   SASL_LIB="$(SASL_LIB)" SASL_INC="$(SASL_INC)" CC="$(CC)" \
 			     $(PERL) ../../../${top_srcdir}/perl/sieve/managesieve/Makefile.PL PREFIX=$(prefix); \
 			fi;
-ptclient_ptdump_SOURCES = imap/cli_fatal.o imap/mutex_fake.o ptclient/ptdump.c
+ptclient_ptdump_SOURCES = imap/cli_fatal.c imap/mutex_fake.o ptclient/ptdump.c
 ptclient_ptdump_LDADD = imap/libimap.a @COM_ERR_LIBS@ $(netnews_remotepurge_LDADD)
-ptclient_ptexpire_SOURCES = imap/cli_fatal.o imap/mutex_fake.o ptclient/ptexpire.c
+ptclient_ptexpire_SOURCES = imap/cli_fatal.c imap/mutex_fake.o ptclient/ptexpire.c
 ptclient_ptexpire_LDADD = $(ptclient_ptdump_LDADD)
 
 #ptclient_ptextract_SOURCES = imap/cli_fatal.o imap/mutex_fake.o ptclient/ptextrac.c
-- 
1.7.7.6

