--- cyrus-imapd-2.4.12/imap/mailbox.c	2011-10-04 20:53:03.000000000 +0100
+++ cyrus-imapd/imap/mailbox.c	2011-11-08 17:33:53.000000000 +0000
@@ -2459,8 +2459,10 @@
     }
 
     /* ensure everything is committed to disk */
-    if (fsync(repack->newindex_fd) || fsync(repack->newcache_fd))
+    if (fsync(repack->newindex_fd) || fsync(repack->newcache_fd)) {
+	r = IMAP_IOERROR;
 	goto fail;
+    }
 
     close(repack->newcache_fd);
     repack->newcache_fd = -1;
