From jrhett@isite.net Thu Oct 10 16:19:19 2002
Date: Fri, 4 Oct 2002 14:38:41 -0700
From: Joe Rhett <jrhett@isite.net>
To: cyrus-bugs@andrew.cmu.edu
Subject: Re: patches for 2_2 .. something I'd really like to see.

Here's the corrected and test patch.

diff -c -r1.56.2.1 Makefile.in
*** Makefile.in 2002/08/28 19:22:44     1.56.2.1
--- Makefile.in 2002/10/04 21:15:35
***************
*** 46,51 ****
--- 46,54 ----
        imtest @SNMP_SUBDIRS@ @EXTRA_SUBDIRS@
  DISTSUBDIRS = doc
  
+ TOOLS = config2header config2man dohash mkimap mupdate-loadgen.pl \
+       not-mkdep rehash translatesieve undohash upgradesieve
+ 
  srcdir = @srcdir@
  VPATH = @srcdir@
  
***************
*** 103,108 ****
--- 106,119 ----
        do \
                (cd $$d; echo "### Making" install "in" `pwd`;  \
                        $(MAKE) $(MFLAGS) DESTDIR=$(DESTDIR) install) ||
exit 1; \
+       done
+ 
+ install-tools::
+       @echo "### Installing tools in $(DESTDIR)$(prefix)/tools"
+       $(srcdir)/install-sh -d ${DESTDIR}$(prefix)/tools
+       for f in $(TOOLS); \
+       do \
+               $(srcdir)/install-sh -c -m 755 tools/$$f $(DESTDIR)$(prefix)/tools || exit 1; \
        done
  
  clean::

On Fri, Oct 04, 2002 at 10:04:10AM -0700, Joe Rhett wrote:
> I know - I found that later but was tired of writing "Sorry, I'm a moron"
> messages and quit for the night.
> 
> On Fri, Oct 04, 2002 at 10:53:16AM -0400, Rob Siemborski wrote:
> > This patch doesn't work, from what I can tell.
> > 
> > First, the path to install-sh in the second line is wrong (you are still
> > in the top level directory there).
> > 
> > Even fixing that, it only seems to move the first file to the tools
> > directory.
> > 
> > Since this doesn't affect the normal build process, I'm willing to commit
> > it if you give me something that works ;)
> > 
> > -Rob
> > 
> > On Thu, 3 Oct 2002, Joe Rhett wrote:
> > 
> > > Okay, so this is another stupid patch.  But you don't always have the
> > > source on the server (sealed server?) and I prefer to have everything in
> > > one installation directory.  It can't hurt to have it there:
> > >
> > > *** Makefile.in-orig    Thu Oct  3 22:19:06 2002
> > > --- Makefile.in Thu Oct  3 22:26:58 2002
> > > ***************
> > > *** 105,110 ****
> > > --- 105,114 ----
> > >                         $(MAKE) $(MFLAGS) DESTDIR=$(DESTDIR) install) ||
> > > exit 1; \
> > >         done
> > >
> > > + install-tools:
> > > +       $(srcdir)/install-sh -d ${DESTDIR}$(prefix)/tools
> > > +       $(srcdir)/../install-sh -m 755 tools/[a-z]*
> > > $(DESTDIR)$(prefix)/tools
> > > +
> > >   clean::
> > >         @-for d in  $(SUBDIRS); \
> > >         do \
> > >
> > > --
> > > Joe Rhett                                                      Chief Geek
> > > JRhett@ISite.Net                                      ISite Services, Inc.
> > >
> > >
> > 
> > -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
> > Rob Siemborski * Andrew Systems Group * Cyert Hall 207 * 412-268-7456
> > Research Systems Programmer * /usr/contributed Gatekeeper
> > 
> 
> -- 
> Joe Rhett                                                      Chief Geek
> JRhett@ISite.Net                                      ISite Services, Inc.

-- 
Joe Rhett                                                      Chief Geek
JRhett@ISite.Net                                      ISite Services, Inc.

