Index: imap/imapd.c
===================================================================
RCS file: /cvs/src/cyrus/imap/imapd.c,v
retrieving revision 1.502.2.3
diff -u -r1.502.2.3 imapd.c
--- imap/imapd.c	13 Dec 2006 20:30:04 -0000	1.502.2.3
+++ imap/imapd.c	20 Dec 2006 17:08:25 -0000
@@ -10461,12 +10461,12 @@
 {
     if (imapd_compress_done) {
 	prot_printf(imapd_out,
-		    "%s BAD [COMPRESSIONACTVE] DEFLATE active via COMPRESS\r\n",
+		    "%s BAD [COMPRESSIONACTIVE] DEFLATE active via COMPRESS\r\n",
 		    tag);
     }
     else if (imapd_tls_comp) {
 	prot_printf(imapd_out,
-		    "%s NO [COMPRESSIONACTVE] %s active via TLS\r\n",
+		    "%s NO [COMPRESSIONACTIVE] %s active via TLS\r\n",
 		    tag, SSL_COMP_get_name(imapd_tls_comp));
     }
     else if (strcasecmp(alg, "DEFLATE")) {
