diff -urN cyrus22-imapd-2.2.12~/imap/mupdate-client.c cyrus22-imapd-2.2.12/imap/mupdate-client.c
--- cyrus22-imapd-2.2.12~/imap/mupdate-client.c	2004-08-25 17:35:41.000000000 +0200
+++ cyrus22-imapd-2.2.12/imap/mupdate-client.c	2005-06-22 08:00:58.383544296 +0200
@@ -73,7 +73,6 @@
 #include "imparse.h"
 #include "iptostring.h"
 #include "mupdate.h"
-#include "mupdate_err.h"
 #include "prot.h"
 #include "protocol.h"
 #include "xmalloc.h"
diff -urN cyrus22-imapd-2.2.12~/imap/mupdate-slave.c cyrus22-imapd-2.2.12/imap/mupdate-slave.c
--- cyrus22-imapd-2.2.12~/imap/mupdate-slave.c	2004-03-08 20:23:04.000000000 +0100
+++ cyrus22-imapd-2.2.12/imap/mupdate-slave.c	2005-06-22 08:00:52.971367072 +0200
@@ -73,7 +73,6 @@
 #include "imparse.h"
 #include "iptostring.h"
 #include "mupdate.h"
-#include "mupdate_err.h"
 #include "exitcodes.h"
 
 /* Returns file descriptor of kick socket (or does not return) */
