diff -u -r cyrus-trunk.original/imap/reconstruct.c cyrus-trunk/imap/reconstruct.c
--- cyrus-trunk.original/imap/reconstruct.c	2010-05-25 15:56:17.000000000 +0200
+++ cyrus-trunk/imap/reconstruct.c	2010-05-25 15:57:10.000000000 +0200
@@ -341,8 +341,10 @@
 					    fflag ? &head : NULL);
 	if (rflag) {
 	    /* build a pattern for submailboxes */
-	    char *p = strchr(buf, '@');
-	    if (p) *p = '\0';
+	    if (domain) {
+		char *p = strchr(buf, '@');
+		if (p) *p = '\0';
+	    }
 	    strlcat(buf, ".*", sizeof(buf));
 
 	    /* append the domain */
