--- a/lib/charset.c	2012-06-15 10:00:18.620038014 +1000
+++ b/lib/charset.c	2012-06-15 10:00:41.210037994 +1000
@@ -723,7 +723,7 @@
     s->substr = (unsigned char *)substr;
 
     /* allocate tracking space and initialise to "no match" */
-    s->starts = xmalloc(s->max_start * sizeof(int));
+    s->starts = xmalloc(s->max_start * sizeof(size_t));
     for (i = 0; i < s->max_start; i++) {
 	s->starts[i] = -1;
     }
