--- backend.c-	2006-07-20 14:46:59.000000000 -0400
+++ backend.c	2006-07-20 14:47:50.000000000 -0400
@@ -276,7 +276,7 @@
     int sock = -1;
     int r;
     int err = -1;
-    struct addrinfo hints, *res0 = NULL, *res1 = NULL, *res;
+    struct addrinfo hints, *res0 = NULL, *res;
     struct sockaddr_un sunsock;
     char buf[2048], *mechlist = NULL;
     struct sigaction action;
@@ -322,12 +322,14 @@
 	    if (!ret_backend) free(ret);
 	    return NULL;
 	}
+#ifdef notdef
         /* Get addrinfo struct for local interface. */
         err = getaddrinfo(config_servername, NULL, &hints, &res1);
         if(err) {
             syslog(LOG_ERR, "getaddrinfo(%s) failed: %s",
                    config_servername, gai_strerror(err));
         }
+#endif // notdef
     }
 
     /* Setup timeout */
@@ -344,6 +346,7 @@
 	sock = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
 	if (sock < 0)
 	    continue;
+#ifdef notdef
         /* Bind to local interface. */
         if (!err) {
             if (bind(sock, res1->ai_addr, res1->ai_addrlen) < 0) {
@@ -353,6 +356,7 @@
             }
             freeaddrinfo(res1);
         }
+#endif // notdef
 	alarm(config_getint(IMAPOPT_CLIENT_TIMEOUT));
 	if (connect(sock, res->ai_addr, res->ai_addrlen) >= 0)
 	    break;
