diff -r -u cyrus-imapd-2.4.5.vanilla//imap/imapd.c cyrus-imapd-2.4.5//imap/imapd.c
--- cyrus-imapd-2.4.5.vanilla//imap/imapd.c	2010-11-29 14:28:06.000000000 +0100
+++ cyrus-imapd-2.4.5//imap/imapd.c	2010-12-06 15:40:59.952977001 +0100
@@ -6616,9 +6616,16 @@
 	    imapd_check(s, 0);
 
 	    if (!r) {
-		prot_printf(s->out, "%s Setquota {" SIZE_T_FMT "+}\r\n%s"
-			    " (Storage %d)\r\n",
-			    tag, strlen(quotaroot), quotaroot, newquota);
+		if (newquota == -1) {
+		    prot_printf(s->out, "%s Setquota {" SIZE_T_FMT "+}\r\n%s"
+				" ()\r\n",
+				tag, strlen(quotaroot), quotaroot);
+		}
+		else {
+		    prot_printf(s->out, "%s Setquota {" SIZE_T_FMT "+}\r\n%s"
+				" (Storage %d)\r\n",
+				tag, strlen(quotaroot), quotaroot, newquota);
+		}
 		pipe_including_tag(s, tag, 0);
 	    }
 	}
